---
title: "DO Multi-Tissue Gene Expression"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
bibliography: hdm.bib 
---

```{r param}
rm(list = ls()) #clear out the environment
min.lod <- 8
remove.pseudogenes = TRUE #if TRUE, removes pseudogenes and predicted genes 
tissue.names <- c("Adipose", "Islet", "Liver", "SkeletalMuscle")
tissue.cols <- c("orange", "#8dd3c7", "tan", "brown") #colors that are not related to the DO/CC colors

#tissue.names <- c("Adipose", "Heart", "Islet", "Liver", "SkeletalMuscle")
#tissue.cols <- c("orange", "brown", "#8dd3c7", "tan", "pink") #colors that are not related to the DO/CC colors

overwrite.results <- FALSE
```

## Introduction
This workflow explores data from the Attie 500 data set.
Here we re-create and expand on observations made by 
Isabela Gerdes-Gyuricza when she was in Gary Churchill's 
lab. Her workflow followed the workflow of an early GTEx 
paper. 

## Data

Data were generated through a collaboration between the Attie
lab at the University of Wisconsin, and The Jackson Laboratory
Detailed descriptions of mouse handling, phenotype gathering,
and initial data processing can be found elsewhere. 

Here we focus exclusively on the transcriptomic data from
five tissues: adipose, heart, islet, liver, and skeletal 
muscle.

I downloaded the following data from the 
[DO QTL viewer](https://churchilllab.jax.org/qtlviewer/attie/DO500HFD)
On March 20, 2023.

1. Adipose.RDS - adipose tissue transcriptome
2. Islet.RDS - pancreatic islet transcriptome
3. Liver.RDS - liver transcriptome
4. QTLViewer_Geno_V10.Rdata - Genotypes for mice
5. SkeletalMuscle.RDS - skeletal muscle transcriptome

The initial results from Isabela are in 
Results > Isabelas_Results. They are also stored in a Dropbox 
folder that Gary shared with me. 

There is a transcriptome from the heart, but we are not 
using those data for this paper. 

### Transcriptome Decomposition

The following plot shows the PCA for the transcriptomes
collected from the five different tissues. Each dot is
a sample, and the samples are colored by their tissue
of origin. 

Before performing the PCA, we regressed out covariates
from each transcriptome. Covariates included sex and 
generation.

```{r here, message = FALSE, warning = FALSE}
library(here)
results.dir <- here("Results", "Transcriptomes")
if(!file.exists(results.dir)){dir.create(results.dir)}
data.results.dir <- here("Results", "Data") #a results folder that holds parsed data that we don't want to mix in with downloaded data
if(!file.exists(data.results.dir)){dir.create(data.results.dir)}
```

```{r load_code, message = FALSE, warning = FALSE}
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}

all.packages <- c("igraph", "pheatmap", "gprofiler2", "qtl2", "stringr",
  "wordcloud", "wordcloud2", "hexbin", "regress", "propagate")
load_libraries(all.packages, personal.library = TRUE)
```

```{r read_transcriptomes}
tissue.data <- lapply(tissue.names, 
  function(x) readRDS(here("Data", "DO", paste0(x, ".RDS"))))
names(tissue.data) <- tissue.names
load(here("Data", "DO", "QTLViewer_Geno_V10.Rdata")) #ensembl.version, genoprobs, K, map, markers

#write out gene information for later use
gene.tables <- lapply(tissue.data, function(x) x$annot.mrna)
saveRDS(gene.tables, file.path(data.results.dir, "Gene_Tables.RDS"))
```

```{r setup}
#set up some objects that we will use throughout the workflow
tissue.pairs <- pair.matrix(1:length(tissue.names))
pair.names <- apply(tissue.pairs, 1, function(x) paste(tissue.names[x], collapse = "-"))

#make a matrix of all binary combinations of tissues
#to compare to.
combo.mat <- t(sapply(2:(2^length(tissue.names))-1, 
  function(x) number2binary(x, length(tissue.names))))
colnames(combo.mat) <- tissue.names
```

```{r remove_pseudogenes}
if(remove.pseudogenes){
  description.files <- file.path(data.results.dir, paste0("Gene_Description_", tissue.names, ".RDS"))
  if(any(!file.exists(description.files))){

    require("biomaRt")

    all.var <- ls()

    lib.loaded <- as.logical(length(which(all.var == "mus")))
    if(!lib.loaded){
        mus <- useEnsembl(biomart="ensembl", dataset="mmusculus_gene_ensembl") 
        }
  }

  for(tx in 1:length(tissue.data)){
    if(!file.exists(description.files[tx]) || overwrite.results){
      one.gene.info <- getBM(attributes = c("ensembl_gene_id", "description"), 
        filters = "ensembl_gene_id", values = tissue.data[[tx]]$annot.mrna$gene.id,
        mart = mus)
      saveRDS(one.gene.info, description.files[tx])
    }else{
      one.gene.info <- readRDS(description.files[tx])
    }

    pred.idx <- grep("predicted", one.gene.info[,"description"], ignore.case = TRUE)
    pseudo.idx <- grep("pseudogene", one.gene.info[,"description"], ignore.case = TRUE)
    remove.idx <- Reduce("union", list(pred.idx, pseudo.idx))

    remove.gene.id <- one.gene.info[remove.idx, "ensembl_gene_id"]
    keep.gene.id <- setdiff(colnames(tissue.data[[tx]]$data$norm), remove.gene.id)
    
    #replace the reduced gene expression data set
    #since we only use the normalized, only replace
    #that matrix.
    subset.expr <- tissue.data[[tx]]$data$norm[,keep.gene.id]
    tissue.data[[tx]]$data$norm <- subset.expr

    #also remove pseudogenes from the eQTL tables
    keep.idx <- which(tissue.data[[tx]]$lod.peaks$additive$gene.id %in% keep.gene.id)
    tissue.data[[tx]]$lod.peaks$additive <- tissue.data[[tx]]$lod.peaks$additive[keep.idx,]
  }
}
```

Regress covariates out of the rank Z normalized gene 
expression data.

```{r adjust}
adj.expr.file <- file.path(data.results.dir, "Tissue_Expression_Adjusted.RDS")
if(!file.exists(adj.expr.file) || overwrite.results){
  #regress out covariates, and normalize residuals
  adj.expr <- lapply(tissue.data, function(x) adjust(x$data$norm, x$covar.matrix))
  saveRDS(adj.expr, adj.expr.file)
}else{
  adj.expr <- readRDS(adj.expr.file)
}
```

```{r match_expr}
#make transcript-matched and individual matched expression data sets
common.expr <- Reduce("intersect", lapply(adj.expr, colnames))
tx.matched.expr <- lapply(adj.expr, function(x) x[,common.expr])

common.ind <- Reduce("intersect", lapply(tissue.data, function(x) rownames(x$data$norm)))
ind.matched.expr <- lapply(tissue.data, function(x) x$data$norm[common.ind,])
```

```{r decomp, fig.width = 7, fig.height = 5}
#combine samples from all tissues into one matrix. Align along common transcripts

tissue.col <- unlist(lapply(1:length(tissue.names), 
  function(x) rep(tissue.cols[x], nrow(tx.matched.expr[[x]]))))
sample.sex <- unlist(lapply(tissue.data, function(x) x$covar.matrix[,1]))

#plot decomposition
decomp.file <- file.path(results.dir, "Tissue.Decomposition.RDS")
if(!file.exists(decomp.file) || overwrite.results){
  all.expr <- Reduce("rbind", tx.matched.expr)
  expr.decomp <- plot.decomp(all.expr, plot.results = FALSE)
  saveRDS(expr.decomp, decomp.file)
}else{
  expr.decomp <- readRDS(decomp.file)
}

par(mar = c(4,4,4,1))
layout(matrix(c(1,2), nrow = 1), widths = c(1, 0.3))
plot(expr.decomp$u,
  xlab = paste0("PC1 (", round(expr.decomp$var.exp[1]*100), "%)"),
  ylab = paste0("PC2 (", round(expr.decomp$var.exp[2]*100), "%)"), type = "n")
abline(v = seq(-0.02, 0.03, 0.01), col = rgb(0.9, 0.9, 0.9), lwd = 0.8)
abline(v = seq(-0.025, 0.035, 0.01), col = "lightgray", lwd = 0.5, lty = "dotted")
abline(h = seq(-0.03, 0.03, 0.01), col = rgb(0.9, 0.9, 0.9), lwd = 0.8)
abline(v = seq(-0.025, 0.035, 0.01), col = "lightgray", lwd = 0.5, lty = "dotted")
points(expr.decomp$u, col = tissue.col, pch = sample.sex+16)
par(mar = c(4,0,4,0))
plot.new()
plot.window(xlim = c(0,1), ylim = c(0,1))
legend(0, 1, legend = tissue.names, col = tissue.cols, pch = 16, adj = 0, title = "Tissue")
legend(0, 0.5, legend = c("Female", "Male"), col = "black", pch = unique(sample.sex)+16, 
  adj = 0, title = "Sex")
```

## Tissue-Specific Genes

The following barplot shows the number of transcripts
identified in each tissue.

```{r tissue_spec, warning = FALSE, error = FALSE, message = FALSE}
tx.list <- lapply(adj.expr, colnames)
n.tx <- sapply(tx.list, length)
barplot_with_num(n.tx, col = tissue.cols, main = "Number of Transcripts",
 text.gap = 0.8, text.shift = 0.2)
```

The following heatmap shows the Jaccard indices for the
transcripts found in each pair of tissues. Liver and 
islet are the most different from each other in terms 
of transcripts expressed.

```{r tissue_jacc}
jacc.mat <- jaccard.matrix(tx.list)
pheatmap(jacc.mat, display_numbers = TRUE)
```

## Pairwise Tissue Correlations  {.tabset .tabset-fade .tabset-pills}

For the transcripts that are shared across all tissues,
how well correlated are the levels of expression?

The following plot shows pairwise correlations for the
average expression value for each transcript common 
to all tissues. Again, heart and skeletal muscle are 
the most highly correlated. 

```{r gene_levels}
#calculate average expression for all transcripts
#across the tissues.
avg.expr <- sapply(tx.matched.expr, colMeans)
tissue.cor <- cor(avg.expr)
pheatmap(tissue.cor, display_numbers = TRUE)
```

The following plots show a more detailed picture of the
pairwise correlations between tissues. There is a subset
of genes that is very highly correlated across tissue
pairs (diagonal). However, for each tissue pair, there 
is also a substantial set of genes that are up in one
tissue of the pair and down in the other (off-diagonal).

```{r fine_cor, fig.width = 8, fig.height = 8}
par(mfrow = c(2,2))
for(i in 1:nrow(tissue.pairs)){
plot.hexbin.as.plot(avg.expr[,tissue.pairs[i,1]], avg.expr[,tissue.pairs[i,2]],
  xlab = tissue.names[tissue.pairs[i,1]], ylab = tissue.names[tissue.pairs[i,2]],
  n.bins = 10, round.legend = 10, 
  main = paste(tissue.names[tissue.pairs[i,1]], "vs.", tissue.names[tissue.pairs[i,2]]))
  abline(0,1)
}
```

I tried looking at which genes were significantly off the diagonal
by looking at genes that were significantly differentially expressed
between tissues, but they are basically all significantly differentially 
expressed.

```{r tx_enrich, results = "asis", eval = FALSE}
#pdf("~/Desktop/tx_enrich.pdf")
for(i in 1:nrow(combo.mat)){
  tx.yes <- which(combo.mat[i,] == 1)
  tx.no <- which(combo.mat[i,] == 0)
  in_text <- paste(tissue.names[tx.yes], collapse = ", ")
  out_text <- paste(tissue.names[tx.no], collapse = ", ")
  cat("### In", in_text, "\n")
  in_tx <- Reduce("intersect", tx.list[tx.yes])
  out_tx <- Reduce("intersect", tx.list[tx.no])
  in_tx_not_rest <- setdiff(in_tx, out_tx)
  enrich <- gost(in_tx_not_rest, organism = "mmusculus", sources = c("GO", "KEGG", "REACTOME"))
  plot.enrichment(enrich, plot.label = paste(length(in_tx_not_rest), "Genes in:", in_text, "\nNot in:", out_text), 
    num.terms = 20)
  cat("\n\n")
}
#dev.off()
```

## eQTL maps {.tabset .tabset-fade .tabset-pills}

The following plots show eQTL maps for each tissue. There are 
some prominent trans bands. 

```{r eqtl_maps, fig.width = 7, fig.height = 7, results = "asis", message = FALSE, warning = FALSE}
#pdf("~/Desktop/eQTL_maps.pdf")
for(tx in 1:length(tissue.data)){
  cat("###", tissue.names[tx], "\n")
  tissue.eqtl <- tissue.data[[tx]]$lod.peaks$additive
  tissue.eqtl <- tissue.eqtl[which(tissue.eqtl[,"lod"] >= min.lod),]
  split.eqtl <- strsplit(tissue.eqtl$marker.id, "_")
  eqtl.chr <- sapply(split.eqtl, function(x) x[1])
  eqtl.pos <- as.numeric(sapply(split.eqtl, function(x) x[2]))/1e6
  allowed.chr <- which(eqtl.chr %in% c(1:19, "X"))

  eqtl.table <- cbind("ID" = tissue.eqtl[allowed.chr,1], "chr" = eqtl.chr[allowed.chr], 
    "pos" = eqtl.pos[allowed.chr], "lod" = tissue.eqtl[allowed.chr,"lod"])
  transcript.pos.table <- as.matrix(tissue.data[[tx]]$annot.mrna[,c("gene.id", "chr", "start")])

  lod.col <- colors.from.values(log(log(as.numeric(eqtl.table[,"lod"]))), 
    col.scale = "blue", grad.dir = "high", light.dark = "d")

  plot.cistrans.table(as.matrix(eqtl.table), as.matrix(transcript.pos.table), 
    map, col = lod.col, cex = 0.5)
  cat("\n\n")
}
#dev.off()
```

## eQTL hotspots {.tabset .tabset-fade .tabset-pills}

The following plots show both local and distal eQTL counts 
along the genome. We used a sliding window of 4 Mb with an
overlap of 2 Mb. eQTL were considered local if they were
within 4 Mb of the gene body. 

Chromosome 17 has a large local eQTL hotspot in all tissues.
This is the chromosome housing the MHC locus, so this hotspot
is not surprising. 

```{r hotspot_thresh}
hotspot.min <- 20 #minimum number of gene targets for hotspot
```

There are also multiple distal eQTL hotspots in each tissue,
defined as an eQTL with at least `r hotspot.min` distal targets.

```{r hotspots, fig.width = 14, fig.height = 3, results = "asis"}
#pdf("~/Desktop/eqtl_hotspots.pdf", width = 10, height = 2.5)
local.target.list <- distal.target.list <- vector(mode = "list", length = length(tissue.names))
names(local.target.list) <- names(distal.target.list) <- tissue.names
for(tx in 1:length(tissue.data)){
  cat("###", tissue.names[tx], "\n")
  tissue.eqtl <- tissue.data[[tx]]$lod.peaks$additive
  tissue.eqtl <- tissue.eqtl[which(tissue.eqtl[,"lod"] >= min.lod),]
  split.eqtl <- strsplit(tissue.eqtl$marker.id, "_")
  eqtl.chr <- sapply(split.eqtl, function(x) x[1])
  eqtl.pos <- as.numeric(sapply(split.eqtl, function(x) x[2]))/1e6
  allowed.chr <- which(eqtl.chr %in% c(1:19, "X"))
  eqtl.table <- cbind("ID" = tissue.eqtl[allowed.chr,1], "chr" = eqtl.chr[allowed.chr], 
    "pos" = eqtl.pos[allowed.chr], "lod" = tissue.eqtl[allowed.chr,"lod"])
  transcript.pos.table <- as.matrix(tissue.data[[tx]]$annot.mrna[,c("gene.id", "chr", "start")])

  #count local and distal eQTL in a sliding window to look for hotspots
  #pdf("~/Desktop/test.pdf", width = 10, height = 5)
  gene.lists <- cistrans.hotspots(as.matrix(eqtl.table), as.matrix(transcript.pos.table), 
    map, mb.buffer = 4, window.mb = 4, window.gap.mb = 2)
  #dev.off()
  local.target.list[[tx]] <- gene.lists$local_targets
  distal.target.list[[tx]] <- gene.lists$distal_targets
  cat("\n\n")
}
#dev.off()
```


## eQTL composition

Here we count the number of local and distal eQTLs in 
each tissue. Local eQTLs are defined as being within
4 Mb of the gene position (start or end).

```{r eqtl_fun}
#this function separates local and distal eQTL and returns
#their coefficients. If there is more than one local eQTL
#for a transcript, this function returns the one with the 
#highest LOD score.
get_local_distal_coef <- function(transcript.id, eqtl.table, gene.table, bp.buffer = 4e6){
    
    tr.idx <- which(gene.table[,1] == transcript.id)

    if(length(tr.idx) == 0){
        cis.coef <- trans.coef <- rep(NA, ncol(eqtl.table))
        names(cis.coef) <- names(trans.coef) <- colnames(eqtl.table)
    }else{
        gene.chr <- gene.table[tr.idx,"chr"]
        gene.start <- as.numeric(gene.table[tr.idx,"start"])*1e6
        gene.end <- as.numeric(gene.table[tr.idx,"end"])*1e6

        eqtl.idx <- which(eqtl.table$gene.id == transcript.id)

        if(length(eqtl.idx) == 0){
          cis.coef <- trans.coef <- rep(NA, ncol(eqtl.table))
          names(cis.coef) <- names(trans.coef) <- colnames(eqtl.table)
        }else{

            eqtl.id <- strsplit(eqtl.table$marker.id[eqtl.idx], "_")
            eqtl.chr <- sapply(eqtl.id, function(x) x[1])
            eqtl.pos <- as.numeric(sapply(eqtl.id, function(x) x[2]))

            on.chr <- which(eqtl.chr %in% gene.chr)
            above.min <- eqtl.pos[on.chr] >= (gene.start-bp.buffer)
            below.max <- eqtl.pos[on.chr] <= (gene.end+bp.buffer)
            nearby <- union(which(above.min), which(below.max))

            if(length(nearby) > 0){
                if(length(nearby) == 1){
                  cis.coef <- eqtl.table[eqtl.idx[on.chr[nearby]],,drop=FALSE]
                }
                #if there is more than one local eQTL, take the one with the 
                #highest LOD score
                if(length(nearby) > 1){
                  max.lod.idx <- which.max(eqtl.table[eqtl.idx[on.chr[nearby]],]$lod)
                  cis.coef <- eqtl.table[eqtl.idx[on.chr[nearby[max.lod.idx]]],,drop=FALSE]
                }
            }else{
                cis.coef <- rep(NA, ncol(eqtl.table))
                names(cis.coef) <- colnames(eqtl.table)
            }

            trans.coef.idx <- setdiff(eqtl.idx, eqtl.idx[on.chr[nearby]])
            if(length(trans.coef.idx) == 0){
              trans.coef <- rep(NA, ncol(eqtl.table))
              names(trans.coef) <- colnames(eqtl.table)

            }else{
              trans.coef <- eqtl.table[trans.coef.idx,,drop=FALSE]
            }
        }
    }
    result <- list("local" = cis.coef, "distal" = trans.coef)
    return(result)
}
```


```{r get_coef}

#pull out only local eQTL coefficients for this test.
local.coef.file <- file.path(results.dir, "eQTL_coef_local.RDS")
distal.coef.file <- file.path(results.dir, "eQTL_coef_distal.RDS")

if(!file.exists(local.coef.file) || overwrite.results){
    local.coef <- distal.coef <- vector(mode = "list", length = length(tissue.names))
    names(local.coef) <- names(distal.coef) <- tissue.names
    for(tx in 1:length(tissue.names)){
        tissue.eqtl <- tissue.data[[tx]]$lod.peaks$additive
        tissue.eqtl <- tissue.eqtl[which(tissue.eqtl[,"lod"] >= min.lod),]
        gene.table <- tissue.data[[tx]]$annot.mrna
        eqtl.transcripts <- unique(tissue.eqtl$gene.id)

        split.coef <- lapply(eqtl.transcripts, 
          function(x) get_local_distal_coef(x, tissue.eqtl, gene.table))
        cis.coef <- Reduce("rbind", lapply(split.coef, function(x) x$local))
        trans.coef <- Reduce("rbind", lapply(split.coef, function(x) x$distal))

        cis.has.vals <- which(apply(cis.coef, 1, function(x) !all(is.na(x))))
        trans.has.vals <- which(apply(trans.coef, 1, function(x) !all(is.na(x))))

        cis.coef  <- cis.coef[cis.has.vals,]
        trans.coef <- trans.coef[trans.has.vals,]

        local.coef[[tx]] <- cis.coef
        distal.coef[[tx]] <- trans.coef
    }
    saveRDS(local.coef, local.coef.file)
    saveRDS(distal.coef, distal.coef.file)
}else{
    local.coef <- readRDS(local.coef.file)
    distal.coef <- readRDS(distal.coef.file)
}
```

The following barplot compares the number of total, distal, and 
local eQTL across tissues. The minimum LOD score for these
eQTL is `r min.lod`.

The bars are stacked, which makes comparing the total number
of eQTL across tissues easier. It is harder to compare the number
of distal eQTL because these start at different heights, but it
is clear from this plot that there are many more local eQTL than
distal eQTL.


```{r compare, fig.width = 7, fig.height = 5}
local.eQTL.count <- sapply(local.coef, nrow)
distal.eQTL.count <- sapply(distal.coef, nrow)
total.eQTL.count <- sapply(tissue.data, function(x) nrow(x$lod.peaks$additive))

eqtl.count.table <- rbind(total.eQTL.count, distal.eQTL.count, local.eQTL.count)
rownames(eqtl.count.table) <- c("Total", "Distal", "Local")

#par(mfrow = c(2,1))
#barplot(eqtl.table, beside = TRUE)
layout(matrix(c(1,2), nrow = 1), widths = c(1,0.2))
par(mar = c(4,4,4,1))
barplot(eqtl.count.table[c("Local", "Distal"),], col = c("#d95f02", "#1b9e77"))
par(mar = c(4,0,4,1))
plot.new()
plot.window(xlim = c(0,1), ylim = c(0,1))
legend(x = 0, y = 0.5, legend = c("Distal", "Local"), fill = c("#1b9e77", "#d95f02"))
```


## Tissue Specificity {.tabset .tabset-fade .tabset-pills}

The following plots explore the overlap of eQTLs across tissues.

Following Isabela's methods, local and distal eQTL for a given 
gene were considered shared across two tissues if their positions 
in their corresponding tissues (LOD > `r min.lod`) were within 4 Mb.

The first plot shows the counts of shared *local* eQTL across
tissues, and the second shows the counts of shared *distal* 
eQTL across tissues. The plots are complex, but the main 
take-away is that local eQTL are much more likely to be shared
across tissues than distal eQTL, which are much more likely 
to be tissue-specific: For the local eQTL, the largest group 
is the eQTL that are shared across all tissues. For the distal 
eQTL, the largest groups are those that are specific to a single 
tissue.


```{r kinship_mat}

#calculate the germline regardless of which kinship 
#type we have specified. If we have specified the 
#functional kinship, we will compare it to the 
#germline kinship.

kin.file <- file.path(data.results.dir, "overall.kinship.RDS")
if(!file.exists(kin.file) || overwrite.results){
    Kg = calc_kinship(genoprobs, "overall")
    saveRDS(Kg, kin.file)
}else{
    Kg <- readRDS(kin.file)
}
```

```{r eQTL_fun}
#functions for listing and comparing eQTL

#this function creates a list of eQTL for a 
get_eQTL <- function(transcript.id, coef.list){
  transcript.idx <- lapply(coef.list, function(x) which(x[,1] == transcript.id))
  transcript.eQTL <- lapply(1:length(coef.list), function(x) coef.list[[x]][transcript.idx[[x]],])
  return(transcript.eQTL)
}

#This function identifies shared eQTL within a given Mb range
assess_shared <- function(eqtl.list, mb.buffer = 4){

    #count the QTL for this transcript across all tissues
    n.qtl <- sapply(eqtl.list, nrow)
    #count the number of tissues with eQTL for this transcript
    has.eqtl <- which(n.qtl > 0)
    #if there are no eQTL, there is nothing to compare
    if(length(has.eqtl) < 1){return(NULL)} 

    #make a table for all eQTL for this transcript
    eqtl.mat <- Reduce("rbind", eqtl.list)
    #keep track of which tissue each came from
    eqtl.tissue <- unlist(lapply(1:length(eqtl.list), function(x) rep(tissue.names[x], nrow(eqtl.list[[x]]))))
    full.mat <- cbind(eqtl.tissue, eqtl.mat)
 
    #if there is only one eQTL, there is nothing to compare
    #return the lone eQTL
    if(nrow(full.mat) < 2){
      shared.eqtl.mat <- matrix(0, nrow = nrow(full.mat), ncol = length(tissue.names))
      colnames(shared.eqtl.mat) <- tissue.names
      rownames(shared.eqtl.mat) <- full.mat[,"marker.id"]
      shared.eqtl.mat[,tissue.names[has.eqtl]] <- 1
      shared.result = list("eqtl.tissues" = shared.eqtl.mat, "eqtl.coef" = list(full.mat))
      return(shared.result)
    }

    #test all pairs of eQTL for overlap
    #make a matrix holding all eQTL pairs, and indicate
    #whether each pair overlaps (1) or doesn't (0)
    test.mat <- pair.matrix(1:nrow(eqtl.mat))
    split.eqtl <- strsplit(eqtl.mat$marker.id, "_")
    eqtl.chr <- sapply(split.eqtl, function(x) x[1])
    eqtl.chr[which(eqtl.chr == "X")] <- 20
    eqtl.chr <- as.numeric(eqtl.chr)
    eqtl.pos <- as.numeric(sapply(split.eqtl, function(x) x[2]))
    full.mat <- cbind(full.mat, eqtl.chr, eqtl.pos)
    sorted.qtl <- sort.by.then.by(full.mat, sort.cols = c(13,14), col.type = c("n", "n")) 
    sorted.tissue <- sorted.qtl[,1]    

    eqtl.chr <- sorted.qtl[,13]
    eqtl.pos <- sorted.qtl[,14]    

    #cluster by position for each unique chromosome
    u_chr <-  unique(eqtl.chr)
    chr.pos <- lapply(u_chr, function(x) eqtl.pos[which(eqtl.chr == x)])
    names(chr.pos) <- u_chr
    #if any chromosomes have more than one eQTL logged, 
    #cluster them such that positions within 4Mb of each
    #other are clustered into a single QTL.        
    dist.tree <- lapply(chr.pos, function(x) if(length(x) > 1){hclust(dist(x/1e6))})
    #plot(dist.tree[[1]]);abline(h = 4)
    #split into multiple QTL based on our specified distance
    eqtl.mem <- lapply(dist.tree, 
      function(x) if(length(x) > 0){cutree(x, h = mb.buffer)}else{return(1)})

    num.eqtl <- sum(unlist(lapply(eqtl.mem, function(x) length(unique(x)))))
    eqtl.idx <- vector(mode = "list", length = length(eqtl.mem))
    start.idx <- 0
    for(eq in 1:length(eqtl.mem)){
      eqtl.idx[[eq]] <- eqtl.mem[[eq]] + start.idx
      start.idx <- max(eqtl.idx[[eq]])
    }
    eqtl.idx <- unlist(eqtl.idx)

    eqtl.names <- sapply(1:num.eqtl, 
        function(x) paste(eqtl.chr[which(eqtl.idx == x)[1]], 
        round(mean(eqtl.pos[which(eqtl.idx == x)])[1]), sep = "_"))

    #for the unique eQTL, tally which tissues each is in.
    shared.eqtl.mat <- matrix(0, nrow = num.eqtl, ncol = length(tissue.names))
    colnames(shared.eqtl.mat) <- tissue.names
    rownames(shared.eqtl.mat) <- eqtl.names

    shared.eqtl.coef <- vector(mode = "list", length = num.eqtl)
    names(shared.eqtl.coef) <- eqtl.names
    for(eq in 1:num.eqtl){
        eqtl.markers <- sorted.qtl[which(eqtl.idx == eq),"marker.id"]
        eqtl.tissues <- unique(sorted.tissue[which(sorted.qtl[,"marker.id"] %in% eqtl.markers)])
        shared.eqtl.mat[eq,eqtl.tissues] <- 1

        shared.eqtl.coef[[eq]] <- full.mat[which(full.mat[,"marker.id"] %in% eqtl.markers),1:12]
    }
    
    shared.result = list("eqtl.tissues" = shared.eqtl.mat, "eqtl.coef" = shared.eqtl.coef)

    return(shared.result)

}


```

### Local eQTL

```{r shared_local, fig.width = 8, fig.height = 5}
#get all unique transcripts across all tissues.

u_transcripts <- unique(unlist(lapply(tissue.data, function(x) x$annot.mrna[,1])))

#find all local eQTL associated with each transcript across all tissues
local.eQTL.pos <- lapply(u_transcripts, function(x) get_eQTL(x, local.coef))
names(local.eQTL.pos) <- u_transcripts

#identify unique local eQTLs that are shared across tissues
local.eqtl.binary.file <- file.path(results.dir, "Shared_eQTL_Local_Binary.RDS")
local.eqtl.coef.file <- file.path(results.dir, "Shared_eQTL_Local_Coef.RDS")
if(!file.exists(local.eqtl.binary.file) || overwrite.results){
  shared.local.list <- lapply(local.eQTL.pos, function(x) assess_shared(x, mb.buffer = 4))
  local.shared.binary <- Reduce("rbind", lapply(shared.local.list, function(x) if(length(x) > 0){x$eqtl.tissues}))
  local.shared.coef <- lapply(shared.local.list, function(x) if(length(x) > 0){x$eqtl.coef})
  saveRDS(local.shared.binary, local.eqtl.binary.file)
  saveRDS(local.shared.coef, local.eqtl.coef.file)
}else{
  local.shared.binary <- readRDS(local.eqtl.binary.file)
  local.shared.coef <- readRDS(local.eqtl.coef.file)
}

#eqtl.list = local.eQTL.pos[[94]]
#shared.local.list[[94]]
#assess_shared(eqtl.list)

combo.list <- apply(combo.mat, 1, 
  function(x) which(apply(local.shared.binary, 1, function(y) all(x == y))))
combo.counts <- sapply(combo.list, length)

bar.col <- rep("gray", nrow(combo.mat))
single.tissue <- which(rowSums(combo.mat) == 1)
bar.col[single.tissue] <- rev(tissue.cols)
layout(matrix(c(1,2), ncol = 1), heights = c(1, 0.7))
par(mar = c(0,6,4,4))
a <- barplot(combo.counts, ylab = "eQTL Count", col = bar.col, ylim = c(0, 5000))
text(x = a[single.tissue], y = combo.counts[single.tissue]+100, 
  labels = rev(tissue.names), srt = 90, adj = 0)
par(mar = c(1,6,0,4))
imageWithText(t(combo.mat), col.scale = "blue", row.text.shift = 0.02)
mtext("Shared Local eQTL", side = 3, outer = TRUE, line = -1.5)
```

### Distal eQTL

```{r shared_distal, fig.width = 8, fig.height = 5}

#find all distal eQTL associated with each transcript across all tissues
distal.eQTL.pos <- lapply(u_transcripts, function(x) get_eQTL(x, distal.coef))
names(distal.eQTL.pos) <- u_transcripts

#identify unique local eQTLs that are shared across tissues
distal.eqtl.binary.file <- file.path(results.dir, "Shared_eQTL_Distal_Binary.RDS")
distal.eqtl.coef.file <- file.path(results.dir, "Shared_eQTL_Distal_Coef.RDS")
if(!file.exists(distal.eqtl.binary.file) || overwrite.results){
  shared.distal.list <- lapply(distal.eQTL.pos, function(x) assess_shared(x, mb.buffer = 4))
  distal.shared.binary <- Reduce("rbind", lapply(shared.distal.list, function(x) if(length(x) > 0){x$eqtl.tissues}))
  distal.shared.coef <- lapply(shared.distal.list, function(x) if(length(x) > 0){x$eqtl.coef})
  saveRDS(distal.shared.binary, distal.eqtl.binary.file)
  saveRDS(distal.shared.coef, distal.eqtl.coef.file)
}else{
  distal.shared.binary <- readRDS(distal.eqtl.binary.file)
  distal.shared.coef <- readRDS(distal.eqtl.coef.file)
}

combo.list <- apply(combo.mat, 1, 
  function(x) which(apply(distal.shared.binary, 1, function(y) all(x == y))))
combo.counts <- sapply(combo.list, length)

bar.col <- rep("gray", nrow(combo.mat))
single.tissue <- which(rowSums(combo.mat) == 1)
bar.col[single.tissue] <- rev(tissue.cols)
layout(matrix(c(1,2), ncol = 1), heights = c(1, 0.7))
par(mar = c(0,6,4,4))
a <- barplot(combo.counts, ylab = "eQTL Count", col = bar.col, ylim = c(0, 5000))
text(x = a[single.tissue], y = combo.counts[single.tissue]+100, 
  labels = rev(tissue.names), srt = 90, adj = 0)
par(mar = c(1,6,0,4))
imageWithText(t(combo.mat), col.scale = "blue", row.text.shift = 0.02)
mtext("Shared Distal eQTL", side = 3, outer = TRUE, line = -1.5)
```

## Allele Effects and Expression Correlation

How do allele effects and gene expression correspond across
tissues? Do two local eQTL shared across tissues also have 
correlated allele effects? And if so, is gene expression 
correlated across the tissues?

### Allele Effect Correlations {.tabset .tabset-fade .tabset-pills}

The following plots show the distribution of correlations
between allele effects for transcripts with shared
eQTL in all pairs of tissues. The boxes are ordered by 
increasing mean correlation. Heart and skeletal muscle 
have the most highly correlated allele effects for shared
local eQTL, as we might expect. 

Allele effects for shared distal eQTL are weirdly highly
correlated. Much more highly correlated than allele effects
for local eQTL. 

#### Shared Local eQTL

```{r local_allele_effects, fig.width = 8, fig.height = 6}

allele_effect_cor <- function(shared.coef, filename){
  
  if(file.exists(filename) && !overwrite.results){
    coef.cor <- readRDS(filename)
    return(coef.cor)
  }
  
  total.eqtl <- sum(sapply(shared.coef, length))
  coef.cor <- matrix(NA, ncol = nrow(tissue.pairs), nrow = total.eqtl)
  colnames(coef.cor) <- pair.names
  eqtl_tID <- rep(NA, total.eqtl)
  
  qtl.idx <- 1
  for(exp in 1:length(shared.coef)){
    #count the number of eQTL for this transcript
    n.qtl <- length(shared.coef[[exp]])
    if(n.qtl > 0){
      #for each qtl that has more than one tissue involved...
      for(eq in 1:n.qtl){
          one.eqtl.coef <- shared.coef[[exp]][[eq]]
          num.tissues <- length(unique(one.eqtl.coef[,"eqtl.tissue"]))
          if(num.tissues < 2){next()}
          eqtl.tissues <- sort(unique(one.eqtl.coef[,1])) #only compare correlations across tissues
          u_mat <- one.eqtl.coef[match(eqtl.tissues, one.eqtl.coef[,1]),]
          eqtl.tissue.pairs <- apply(pair.matrix(eqtl.tissues), 1, function(x) paste(x[1], x[2], sep = "-"))
          pair.cor <- cor(t(u_mat[,LETTERS[1:8]]))
          #test.mat <- u_mat[,LETTERS[1:8]]
          #rownames(test.mat) <- u_mat[,1]
          #barplot(as.matrix(test.mat), beside = TRUE);abline(h = 0)
          #pheatmap(test.mat, cluster_rows = FALSE, cluster_cols = FALSE)
          #pheatmap(cor(t(test.mat)), display_numbers = TRUE)
          tissue.pair.cor <- as.numeric(pair.cor[upper.tri(pair.cor, diag = FALSE)])
          #par(mar = c(12,4,4,4));barplot(tissue.pair.cor, names = eqtl.tissue.pairs, las = 2, main = names(shared.coef)[exp])
          name.idx <- match(eqtl.tissue.pairs, pair.names)
          coef.cor[qtl.idx,name.idx] <- tissue.pair.cor 
          eqtl_tID[qtl.idx] <- paste0(names(shared.coef)[exp], "_qtl", eq)
          qtl.idx <- qtl.idx + 1
      }
    }  
  }
  rownames(coef.cor) <- eqtl_tID
  saveRDS(coef.cor, filename)
  return(coef.cor)
}

#for(i in 1:length(tissue.names)){
#  write.table(tissue.data[[i]]$lod.peaks$additive, 
#  paste0("~/Desktop/", paste0(tissue.names[i], "_eQTL.csv")), sep = ",", quote = FALSE,
#  row.names = FALSE)
#}

local.eqtl.cor.file <- file.path(results.dir, "Correlation_Allele_Effects_Local.RDS")
local.eqtl.cor <- allele_effect_cor(local.shared.coef, local.eqtl.cor.file)

mean.order <- order(apply(local.eqtl.cor, 2, function(x) mean(x, na.rm = TRUE)))
par(mar = c(12,4,4,4))
boxplot(local.eqtl.cor[,mean.order], las = 2, 
  main = "Correlation of Allele Effects\nFor Shared Local eQTL")
abline(h = c(0.5, 0, -0.5), lty = c(2,1,2))
```


#### Shared Distal eQTL

```{r distal_eqtl_cor, fig.width = 8, fig.height = 6}
distal.eqtl.cor.file <- file.path(results.dir, "Correlation_Allele_Effects_Distal.RDS")
distal.eqtl.cor <- allele_effect_cor(distal.shared.coef, distal.eqtl.cor.file)

mean.order <- order(apply(distal.eqtl.cor, 2, function(x) mean(x, na.rm = TRUE)))
par(mar = c(12,4,4,4))
boxplot(distal.eqtl.cor[,mean.order], las = 2,
  main = "Correlation of Allele Effects\nFor Genes with Shared Distal eQTL")
abline(h = c(0.5, 0, -0.5), lty = c(2,1,2))
```

#### Local vs Distal Comparison

This plot shows the correlations of the local allele effects
(upper triangle) and the distal allele effects (lower triangle)
across pairs of tissues.

```{r compare_mean_cor}
mean.cor.mat <- matrix(NA, nrow = length(tissue.names), ncol = length(tissue.names))
rownames(mean.cor.mat) <- colnames(mean.cor.mat) <- tissue.names
for(i in 1:nrow(tissue.pairs)){
  mean.cor.mat[tissue.pairs[i,1], tissue.pairs[i,2]] <- median(local.eqtl.cor[,i], na.rm = TRUE)
  mean.cor.mat[tissue.pairs[i,2], tissue.pairs[i,1]] <- median(distal.eqtl.cor[,i],  na.rm = TRUE)
}
pheatmap(mean.cor.mat, cluster_rows = FALSE, cluster_cols = FALSE, display_numbers = TRUE)
```

### Gene Expression Correlation {.tabset .tabset-fade .tabset-pills}

The following plots show expression correlation across 
tissues for transcripts with shared eQTL, both local
and distal.

Boxes are ordered by mean correlation. For the transcripts
with local eQTL, heart and skeletal muscle have the highest 
correlation between gene expression. The gene expression 
correlation is lower than the allele effect correlations.

The correlations between transcripts with distal eQTL
is again weirdly highly correlated. I have spot-checked
some of these, and they do seem right...

#### Local eQTL

```{r expr_cor, fig.width = 8, fig.height = 6}

expr_cor <- function(shared.coef, expr.mat, filename){

  if(!file.exists(filename) || overwrite.results){

    total.eqtl <- sum(sapply(shared.coef, length))
    expr.cor <- matrix(NA, ncol = nrow(tissue.pairs), nrow = total.eqtl)
    colnames(expr.cor) <- pair.names
    qtl.idx <- 1
    tx.id <- rep(NA, total.eqtl)

      for(i in 1:length(shared.coef)){
        if(length(shared.coef[[i]]) == 0){next()}
          qtl.info <- shared.coef[[i]]
          n.qtl <- length(qtl.info)
          for(qtl in 1:n.qtl){
            transcript.id <- qtl.info[[qtl]][1,2]
            tx.id[qtl.idx] <- paste0(transcript.id, "_qtl", qtl)
            tissue.id <- sort(unique(qtl.info[[qtl]][,1])) #only compare cross-tissue expression
            eqtl.tissue.pairs <- apply(pair.matrix(tissue.id), 1, function(x) paste(x[1], x[2], sep = "-"))
            
            tissue.idx <- match(tissue.id, tissue.names)
            tx.expr <- sapply(expr.mat[tissue.idx], function(x) x[,transcript.id])
            pair.cor <- cor(tx.expr)

            tissue.pair.cor <- as.numeric(pair.cor[upper.tri(pair.cor, diag = FALSE)])
            name.idx <- match(eqtl.tissue.pairs, pair.names)
            expr.cor[qtl.idx,name.idx] <- tissue.pair.cor 
            qtl.idx <- qtl.idx + 1

          } #end looping through qtl
      } #end looping through shared.coef
    rownames(expr.cor) <- tx.id

    saveRDS(expr.cor, filename)
  }else{
    expr.cor <- readRDS(filename)
  }
  return(expr.cor)

}

local.expr.cor.file <- file.path(results.dir, "Correlation_Expression_Local.RDS")
local.expr.cor <- expr_cor(shared.coef = local.shared.coef, 
  expr.mat = ind.matched.expr, filename = local.expr.cor.file)

mean.order <- order(apply(local.expr.cor, 2, function(x) mean(x, na.rm = TRUE)))
par(mar = c(12,4,2,4))
boxplot(local.expr.cor[,mean.order], las = 2, 
  main = "Correlation of Gene Expression\nFor Genes with Shared Local eQTL")
abline(h = c(0.5, 0, -0.5), lty = c(2,1,2))

```

#### Distal eQTL

```{r expr_cor_distal, fig.width = 8, fig.height = 6}
distal.expr.cor.file <- file.path(results.dir, "Correlation_Expression_Distal.RDS")
distal.expr.cor <- expr_cor(shared.coef = distal.shared.coef, 
  expr.mat = ind.matched.expr, filename = distal.expr.cor.file)

mean.order <- order(apply(distal.expr.cor, 2, function(x) mean(x, na.rm = TRUE)))
par(mar = c(12,4,4,4))
boxplot(distal.expr.cor[,mean.order], las = 2,
  main = "Correlation of Gene Expression\nFor Genes with Shared Distal eQTL")
abline(h = c(0.5, 0, -0.5), lty = c(2,1,2))
```

#### Local vs Distal Comparison

This plot shows the correlations of the local gene expression
(upper triangle) and the distal gene expression (lower triangle)
across pairs of tissues.


```{r compare_mean_expr_cor}
mean.cor.mat <- matrix(NA, nrow = length(tissue.names), ncol = length(tissue.names))
rownames(mean.cor.mat) <- colnames(mean.cor.mat) <- tissue.names
for(i in 1:nrow(tissue.pairs)){
  mean.cor.mat[tissue.pairs[i,1], tissue.pairs[i,2]] <- median(local.expr.cor[,i], na.rm = TRUE)
  mean.cor.mat[tissue.pairs[i,2], tissue.pairs[i,1]] <- median(distal.expr.cor[,i],  na.rm = TRUE)
}
pheatmap(mean.cor.mat, cluster_rows = FALSE, cluster_cols = FALSE, display_numbers = TRUE)
```

#### eQTL Groups

Genes can be broken into four groups depending on whether they have
or don't have local or distal eQTL. What is the correlation of expression
of these groups of genes?

From the plots below, it looks as if transcripts with distal eQTL,
regardless of whether they are shared across tissues, have higher
correlations across tissues than genes with local eQTL or genes with
no eQTL.

```{r all_four, fig.width = 8, fig.height = 8}

num.local <- t(sapply(local.eQTL.pos, function(x) sapply(x, nrow)))
num.distal <- t(sapply(distal.eQTL.pos, function(x) sapply(x, nrow)))
colnames(num.local) <- colnames(num.distal) <- tissue.names

#pdf("~/Desktop/test.pdf", width = 16, height = 8)
#for each tissue pair
par(mfrow = c(2,2))
for(txp in 1:nrow(tissue.pairs)){
  local.qtl.combo <- cbind(num.local[,tissue.pairs[txp,1],drop=FALSE], num.local[,tissue.pairs[txp,2],drop=FALSE])
  distal.qtl.combo <- cbind(num.distal[,tissue.pairs[txp,1],drop=FALSE], num.distal[,tissue.pairs[txp,2],drop=FALSE])  
  #identify genes that do and don't have local and distal eQTL across both tissues
  local.no <- Reduce("intersect", apply(local.qtl.combo, 2, function(x) which(x == 0)))
  local.yes <- Reduce("intersect", apply(local.qtl.combo, 2, function(x) which(x > 0)))
  distal.no <- Reduce("intersect", apply(distal.qtl.combo, 2, function(x) which(x == 0)))
  distal.yes <- Reduce("intersect", apply(distal.qtl.combo, 2, function(x) which(x > 0)))

  expr.mat1 <- ind.matched.expr[[tissue.pairs[txp,1]]]
  expr.mat2 <- ind.matched.expr[[tissue.pairs[txp,2]]]
  common.tx <- intersect(colnames(expr.mat1), colnames(expr.mat2))

  local.yes.distal.yes <- intersect(rownames(num.local)[intersect(local.yes, distal.yes)], common.tx)
  local.yes.distal.no <- intersect(rownames(num.local)[intersect(local.yes, distal.no)], common.tx)
  local.no.distal.yes <- intersect(rownames(num.local)[intersect(local.no, distal.yes)], common.tx)
  local.no.distal.no <- intersect(rownames(num.local)[intersect(local.no, distal.no)], common.tx)

  #barplot(c(length(local.yes.distal.yes), length(local.yes.distal.no), 
  #  length(local.no.distal.yes), length(local.no.distal.no)))

  local.yes.distal.yes.cor <- sapply(local.yes.distal.yes, 
    function(x) cor(expr.mat1[,x], expr.mat2[,x]))

  local.yes.distal.no.cor <- sapply(local.yes.distal.no, 
    function(x) cor(expr.mat1[,x], expr.mat2[,x]))

  local.no.distal.yes.cor <- sapply(local.no.distal.yes, 
    function(x) cor(expr.mat1[,x], expr.mat2[,x]))

  local.no.distal.no.cor <- sapply(local.no.distal.no, 
    function(x) cor(expr.mat1[,x], expr.mat2[,x]))

  all.cor <- sapply(common.tx, function(x) cor(expr.mat1[,x], expr.mat2[,x]))

  boxplot(list("local_distal" = local.yes.distal.no.cor, 
    "no.local_distal" = local.no.distal.yes.cor, 
    "local_no.distal" = local.yes.distal.no.cor, 
    "no.local_no.distal" = local.no.distal.no.cor,
    "all" = all.cor), ylab = "Correlation",
    main = paste(tissue.names[tissue.pairs[txp,]], collapse = " vs. "),
    ylim = c(-1,1), las = 2)
    abline(h = 0)

}
#dev.off()
```

## Correlation between gene expression and coefficient correlation {.tabset .tabset-fade .tabset-pills}

Are transcripts with highly correlated eQTL coefficients more
highly correlated in expression than genes with lower correlations
between eQTL coefficients?

```{r expression_coef_cor}

plot_expr_cor_v_eqtl_cor <- function(expr.cor, eqtl.cor, 
  xlab = "eQTL Coefficient Correlation", ylim = c(-1, 1)){
  common.tx <- intersect(rownames(expr.cor), rownames(eqtl.cor))
  expr.idx <- match(common.tx, rownames(expr.cor))
  coef.idx <- match(common.tx, rownames(eqtl.cor))
  tx.pair.stats <- matrix(NA, nrow = nrow(tissue.pairs), ncol = 2)
  colnames(tx.pair.stats)  <- c("R2", "p")
  rownames(tx.pair.stats) <- pair.names
  layout(get.layout.mat(nrow(tissue.pairs)))
  for(txp in 1:nrow(tissue.pairs)){
    tx.pair.stats[txp,] <- plot.with.model(eqtl.cor[coef.idx,txp], 
      expr.cor[expr.idx,txp],ylim = ylim,
      main = paste(tissue.names[tissue.pairs[txp,]], collapse = " vs. "), 
    xlab = xlab, ylab = "Expression Correlation", col = "gray")
    abline(h = 0, v = 0, lty = "dashed", lwd = 2)
  }

  invisible(tx.pair.stats)
}

```

### Local eQTL

The following plot shows the relationship between eQTL coefficient correlations
and expression correlation across tissue pairs for local eQTL.

```{r local_cor, fig.width = 12, fig.height = 9}
meta.cor.local <- plot_expr_cor_v_eqtl_cor(local.expr.cor, local.eqtl.cor, 
  "local eQTL Coefficient Correlation")
```

### Distal eQTL

The following plot shows the relationship between eQTL coefficient
correlations and expression correlation across tissue pairs for 
distal eQTL.

```{r distal_cor, fig.width = 12, fig.height = 9}
meta.cor.distal <- plot_expr_cor_v_eqtl_cor(distal.expr.cor, distal.eqtl.cor, 
  "distal eQTL Coefficient Correlation")
```

It turns out that the distal eQTL are probably not worth
digging into. Most of the targets are pseudogenes, so there
are issues with alignments. We will ignore distal eQTL
for the remainder of this analysis. 

## Tissue-specific eQTL

We are interested in whether there are tissue-specific
local allele effects. To get at this, we will look for 
interaction effects between genotype and tissue in a 
linear model.

The model looks for an interaction between genotype and
tissue influencing expression.

```{r get_effects_fun}
get_transcript_effects <- function(transcript.id, return.data = FALSE){

    tx.idx <- which(all.tx == transcript.id)
    tissue.idx <- lapply(tissue.tx, function(x) which(x == all.tx[tx.idx]))
    has.expr <- which(sapply(tissue.idx, length) > 0)
    
    if(length(has.expr) < 2){return(NULL)}

    tx.expr <- sapply(1:length(has.expr), function(x) ind.matched.expr[[has.expr[[x]]]][,tissue.idx[has.expr][[x]]])
    colnames(tx.expr) <- names(ind.matched.expr)[has.expr]
    
    tx.info  <- tissue.data[[has.expr[1]]]$annot.mrna[which(tissue.data[[has.expr[1]]]$annot.mrna$gene.id == all.tx[tx.idx]),]
    tx.chr <- tx.info$chr
    chr.idx <- which(names(K) == tx.chr)
    gene.name <- tx.info$symbol

    if(!(tx.chr %in% c(1:19, "X"))){return(NULL)}

    nearest.marker <- tx.info$nearest.marker.id
    nearest.geno <- pull_genoprobpos(genoprobs, marker = nearest.marker)

    tx.expr <- lapply(1:length(has.expr), function(x) ind.matched.expr[[has.expr[[x]]]][,tissue.idx[has.expr][[x]]])
    names(tx.expr) <- tissue.names[has.expr]
    long.expr <- unlist(tx.expr)
    tissue.factor <- as.factor(unlist(lapply(1:length(tx.expr), 
      function(x) rep(tissue.names[x], length(tx.expr[[x]])))))
    rep.geno <- Reduce("rbind", lapply(1:length(has.expr), function(x) nearest.geno[rownames(ind.matched.expr[[1]]),]))      
    rep.K <- K[[chr.idx]][rownames(rep.geno), rownames(rep.geno)]
    rownames(rep.K) <- colnames(rep.K) <- names(long.expr)

    if(use.kinship){
      #use cape code to adjust variables
      adj_data <- kin_adjust(kin_obj = rep.K, geno = rep.geno, 
        phenoV = as.matrix(long.expr))
      test <- lm(adj_data$corrected_pheno~tissue.factor*adj_data$corrected_geno)
    }else{
      test <- lm(long.expr~tissue.factor*rep.geno)
      #anova(test)
    }
    test.summary <- anova(test)
    component.p <- test.summary[,"Pr(>F)"][1:3]
    names(component.p) <- rownames(anova(test))[1:3]
    
    tissue.effect <- test.summary[1,"Mean Sq"]
    geno.effect <- test.summary[2,"Mean Sq"]
    int.effect <- test.summary[3,"Mean Sq"]
    resid.effect <- test.summary[4,"Mean Sq"]

    relative.effects <- c("Tissue" = tissue.effect, 
      "Genotype" = geno.effect, 
      "Interaction" = int.effect,
      "Residual" = resid.effect)

    if(!return.data){
      result = list("tissues.with.expr" = has.expr,
        "p.values" = component.p, "rel.effects" = relative.effects)
    }else{
      result = list("tissues.with.expr" = has.expr,
        "p.values" = component.p, "rel.effects" = relative.effects,
        "genotypes" = nearest.geno, "expression" = tx.expr,
        "chr" = chr.idx, "gene.name" = gene.name)
    }


    return(result)  
}
```

```{r tissue_specific}

save.every = 100
use.kinship = FALSE

tissue.tx <- lapply(ind.matched.expr, function(x) colnames(x))
all.tx <- Reduce("union", tissue.tx)

#calculate all local LOD scores for future comparisons
#and thresholding
local.lod.file <- here("Results", "Transcriptomes", "Local_LOD.RDS")
if(!file.exists(local.lod.file) || overwrite.results){
  local.lod <- sapply(local.coef, function(x) x$lod[match(all.tx, x$gene.id)])
  rownames(local.lod) <- all.tx
  saveRDS(local.lod, local.lod.file)
}else{
  local.lod <- readRDS(local.lod.file)
}

max.lod <- apply(local.lod, 1, function(x) if(!all(is.na(x))){max(x, na.rm = TRUE)}else{NA})

if(use.kinship){
  int.p.file <- file.path(results.dir, "Tissue-Specific_eQTL_p_values_with_kin.RDS")
}else{
  int.p.file <- file.path(results.dir, "Tissue-Specific_eQTL_p_values_no_kin.RDS") 
}
relative.effect.file <- file.path(results.dir, "Transcripts_Relative_Effects.RDS")
tissue.grid.file <- file.path(results.dir, "Transcripts_by_Tissue.RDS")

if(file.exists(int.p.file) && !overwrite.results){
  p.mat <- readRDS(int.p.file)
  tissue.grid <- readRDS(tissue.grid.file)
  relative.effect <- readRDS(relative.effect.file)
  start.at <- max(which(!is.na(p.mat[,1]))) + 1 
}else{
  p.mat <- matrix(NA, nrow = length(all.tx), ncol = 3)
  rownames(p.mat) <- all.tx
  colnames(p.mat) <- c("Tissue", "Genotype", "Tissue_by_Genotype")

  relative.effect <- matrix(NA, nrow = length(all.tx), ncol = 4)
  rownames(relative.effect) <- all.tx
  colnames(relative.effect) <- c("Tissue", "Genotype", "Interaction", "Residual")

  tissue.grid <- matrix(0, nrow = length(all.tx), ncol = length(tissue.names)) #keep track of which tissues each transcript is expressed in
  rownames(tissue.grid) <- all.tx
  colnames(tissue.grid) <- tissue.names
  start.at <- 1
}

if(start.at < length(all.tx)){
  sink(file.path(results.dir, "progress.txt"))
  for(tx.idx in start.at:length(all.tx)){
  #for(tx.idx in 1:200){
    #tx.idx <- which(all.tx == transcript.id)
    print(tx.idx)
    tx.effects <- get_transcript_effects(transcript.id = all.tx[tx.idx])
    if(is.null(tx.effects)){next()}

    tissue.grid[tx.idx,tx.effects$tissues.with.expr] <- 1  
    p.mat[tx.idx,] <- tx.effects$p.values
    relative.effect[tx.idx,] <- tx.effects$rel.effects

      if(tx.idx %% save.every == 0){
        saveRDS(p.mat, int.p.file)
        saveRDS(tissue.grid, tissue.grid.file)
        saveRDS(relative.effect, relative.effect.file)
      }

    }
  saveRDS(p.mat, int.p.file)
  saveRDS(tissue.grid, tissue.grid.file)
  saveRDS(relative.effect, relative.effect.file)
  sink()
}

```

Most transcripts are expressed in all five tissues.

```{r tissue_grid}
num.tissues <- rowSums(tissue.grid)
hist(num.tissues, main = "Number of Expressing Tissues")
```

The boxplot below shows the relative effect of tissue,
local genotype, and the interaction between them on 
gene expression across all transcripts.
Tissue had the largest effect, as we would expect.

```{r effect_ratio}

total.effect <- rowSums(relative.effect)
prop.effect <- relative.effect/total.effect

boxplot(prop.effect, ylab = "Relative Effect")
```

The following plot shows the number of transcripts
with significant effects from each term. 

```{r sig_counts}
fdr = 0.01
alpha <- 0.05/nrow(p.mat)
adj.mat <- apply(p.mat, 2, function(x) p.adjust(x, "holm"))
sig.count <- apply(p.mat, 2, function(x) length(which(x <= alpha)))
total.genes <- length(which(!is.na(p.mat[,1])))
all.counts <- c(total.genes, sig.count)
names(all.counts)[1] <- "Total"
barplot_with_num(all.counts, 
  ylab = "count", main = "Number of Significant Genes")
```


Tissue is the highest contributor to variation in 
transcript level for the vast majority of genes.
There are a handful of genes for which genotype 
has the maximum influence, and even a few for 
which the interaction has the largest effect.

```{r contribution_count}
max.cont <- apply(prop.effect, 1, function(x) if(!all(is.na(x))){which.max(x)}else{NA})
highest.count <- sapply(1:ncol(prop.effect), function(x) length(which(max.cont == x)))
barplot_with_num(highest.count, main = "Which contributes most?",
  names = colnames(prop.effect))

#only LOD > 6
lod.thresh = 6
sig.lod.idx <- which(max.lod > lod.thresh)
highest.count <- sapply(1:ncol(prop.effect), function(x) length(intersect(which(max.cont == x), sig.lod.idx)))
barplot_with_num(highest.count, main = paste("Which contributes most?\nLOD >", lod.thresh),
  names = colnames(prop.effect))
```

The following dot plot compares the contributions
from local genotype and the interaction. The color
of each point corresponds to the maximum LOD score
for the transcript across all tissues. 

For most transcripts, the majority of the variance is 
explained by local genotype. For those with more variance 
explained by the interaction between tissue and genotype, 
the maximum LOD score across all tissues tends to be low.

The counts for above and below the line y=x are
also shown.

```{r rel_cont}

max.lod.col <- colors.from.values(max.lod, use.pheatmap.colors = TRUE)

layout(matrix(c(1,2,1,3), ncol = 2, byrow = TRUE), widths = c(1, 0.3), heights = c(0.3, 1))
par(mar = c(4,4,4,4))
plot(prop.effect[,2], prop.effect[,3], col = max.lod.col, pch = 16,
  xlab = "Genotype", ylab = "Interaction")
abline(0,1)
par(mar = c(0,0,0,0))
plot.text("LOD", y = 0.2, font = 2)
par(mar = c(4,4,0,4))
imageWithTextColorbar(matrix(max.lod, ncol = 1), use.pheatmap.colors = TRUE,
  cex = 1, bar.lwd = 2.5)

par(mfrow = c(1,1))
rel.cont <- apply(prop.effect[,2:3], 1, function(x) if(!all(is.na(x))){which.max(x)}else{NA})
higher.count <- sapply(1:2, function(x) length(which(rel.cont == x)))
barplot_with_num(higher.count, main = "Which contributes more?",
  names = colnames(prop.effect)[2:3])
```

```{r plot_tissue_int}

plot_tissue_interaction <- function(transcript.id, use.kinship = FALSE,
  text.gap = 0.1, text.shift = 0.05){

    tx.effects <- get_transcript_effects(transcript.id, return.data = TRUE)

    if(is.null(tx.effects)){
      return("Not expressed in multiple tissues")
    }

    has.expr <- tx.effects$tissues.with.expr
    nearest.geno <- tx.effects$genotypes
    chr.idx <- tx.effects$chr
    expr.mat <- Reduce("cbind", tx.effects$expression)
    colnames(expr.mat) <- names(has.expr)
    relative.effects <- tx.effects$rel.effects
    total.effect <- sum(relative.effects)
    prop.effect <- relative.effects/total.effect

    test <- apply(expr.mat, 2, function(x) fit1(nearest.geno, x, kinship = K[[chr.idx]]))
    tissue.lod <- sapply(test, function(x) x$lod)

    layout(matrix(c(1,2,3,4,4,4), nrow = 2, byrow = TRUE))
    par(mar = c(4,4,4,4))
    boxplot(expr.mat, main = "Expression by tissue", col = tissue.cols[has.expr], 
      las = 2, ylab = "TPM")
    bar.y <- get_plot_bounds(0, round(max(tissue.lod)), scale.factor = 0.2)
    barplot_with_num(round(tissue.lod), main = "LOD score by tissue", 
      col = tissue.cols[has.expr], las = 2, text.gap = text.gap, 
      text.shift = text.shift)
    barplot_with_num(signif(prop.effect, 2), main = "Relative Effects",
      ylab = "Mean Effect")
    
    all.coef <- sapply(test, function(x) x$coef)
    #pheatmap(all.coef[1:8,], scale = "none")
    par(mar = c(4,4,0,4))
    a <- barplot(t(all.coef[1:8,]), beside = TRUE, col = tissue.cols[has.expr], 
      ylab = "Allele Effects", names = rep("", 8))
    bar.range <- max(all.coef[1:8,]) - min(all.coef[1:8,])
    par(xpd = NA)
    text(x = colMeans(a), y = rep(min(all.coef[1:8,])-bar.range*0.05, 8), 
      labels = names(CCcolors), col = CCcolors, font = 2, cex = 2)
    par(xpd = FALSE)

    #I think this barplot is less useful
    #barplot(all.coef[1:8,], beside = TRUE, col = CCcolors, ylab = "Allele Effects")
    
    gene.name <- tx.effects$gene.name
    int.p <- signif(tx.effects$p.values[3], 2)
    mtext(paste0(gene.name, " (p = ", int.p, ")"), 
      side = 3, outer = TRUE, line = -1.5, font = 4, cex = 1.5)
 
}



#This function plots expression for all alleles
#in a given tissue.
expr_across_alleles <- function(transcript.id, tissue.name){
  
  tx.effects <- get_transcript_effects(transcript.id, return.data = TRUE)

  if(is.null(tx.effects)){
    plot.text("No comparable expression.")
    return(NULL)
  }

  nearest.geno <- tx.effects$genotypes
  chr.idx <- tx.effects$chr
  expr.list <- tx.effects$expression
  tissue.idx <- which(names(expr.list) == tissue.name)
  relative.effects <- tx.effects$rel.effects
  gene.name <- tx.effects$gene.name

  par(mfrow = c(2,4))
  for(a in 1:ncol(nearest.geno)){
    rounded.geno <- bin.vector(nearest.geno[names(expr.list[[tissue.idx]]),a], c(0, 0.5, 1))
    boxplot(expr.list[[tissue.idx]]~rounded.geno, 
      xlab = "Genotype", ylab = paste(gene.name, "expression"),
      main = names(CCcolors[a]), col = CCcolors[a])
    #plot.with.model(nearest.geno[names(expr.list[[tissue.idx]]),a], 
    #  expr.list[[tissue.idx]], 
    #  xlab = "Genotype", ylab = paste(gene.name, "expression"),
    #  main = names(CCcolors[a]), col = CCcolors[a], report = "cor.test")
  }  
  mtext(tissue.name, side = 3, outer = TRUE, line = -1.5, font = 2)
}


#This function plots expression across tissues for a 
#given allele.
expr_across_tissues <- function(transcript.id, allele.name){

  allele.idx <- c(which(names(CCcolors) == allele.name), which(LETTERS[1:8] == allele.name))
  allele.label <- names(CCcolors)[allele.idx]

  tx.effects <- get_transcript_effects(transcript.id, return.data = TRUE)
  
  if(is.null(tx.effects)){
    plot.text("No comparable expression.")
    return(NULL)
  }

  has.expr <- tx.effects$tissues.with.expr
  nearest.geno <- tx.effects$genotypes
  chr.idx <- tx.effects$chr
  expr.list <- tx.effects$expression
  relative.effects <- tx.effects$rel.effects  
  gene.name <- tx.effects$gene.name

  layout(get.layout.mat(length(expr.list)))
  for(tx in 1:length(expr.list)){
    rounded.geno <- bin.vector(nearest.geno[names(expr.list[[tx]]),allele.idx], c(0, 0.5, 1))
    boxplot(expr.list[[tx]]~rounded.geno, col = tissue.cols[has.expr[tx]],
      xlab = "Genotype", ylab = paste(gene.name, "expression"),
      main = names(expr.list)[tx])
      #plot.with.model(nearest.geno[names(expr.list[[tx]]),allele.idx], expr.list[[tx]], 
      #xlab = "Genotype", ylab = paste(gene.name, "expression"),
      #main = names(expr.list)[tx], col = tissue.cols[has.expr[tx]],
      #report = "cor.test")
  }  
  mtext(allele.label, side = 3, outer = TRUE, line = -1.5, font = 2)

}
```

## Examples

### No tissue-specific effect

The following plot shows an example of a transcript
with no tissue-specific effects. The allele 
effects are very similar across all tissues.

```{r no_int, fig.width = 8, fig.height = 7}
int.diff <- relative.effect[,"Interaction"] - relative.effect[,"Genotype"]
sorted.int.diff <- sort(int.diff)
trans.id <- names(sorted.int.diff)[2] #small interaction
plot_tissue_interaction(trans.id, FALSE, text.gap = 3, text.shift = 2)
```

### Strong tissue-specific effect

The following plot shows an example of a transcript 
with strongly tissue-specific eQTL. The CAST allele
has a large positive effect on expression, but only
in islet.


```{r strong_int, fig.width = 8, fig.height = 7}
#pick a transcript with a large LOD score and a 
#strong interaction effect
high.int <- int.diff[which(max.lod > 50)]
#boxplot(list(int.diff, high.int))
sorted.high.int <- sort(high.int, decreasing = TRUE)
trans.id <- names(sorted.high.int)[2] #large interaction
plot_tissue_interaction(trans.id, FALSE)
```

This can be clearly seen in the average 
expression for each allele shown below.

```{r expr_plot, fig.width = 8, fig.height = 8}
expr_across_tissues(trans.id, "CAST")
```

```{r int_stronger_examples, eval = FALSE}
trans.id <- names(sorted.high.int)[4]
plot_tissue_interaction(trans.id, FALSE)
expr_across_tissues(trans.id, "PWK")

pdf("~/Desktop/test.pdf", width = 12, height = 7)
for(tx in 1:length(tissue.names)){
  expr_across_alleles(trans.id, tissue.names[tx])
}
dev.off()
```

```{r, eval = FALSE}
clin.pheno <- readRDS(file.path(data.results.dir, "Clinical_Phenotypes_Adjusted.RDS"))
clin.decomp <- plot.decomp(clin.pheno, plot.results = FALSE)
clin.pc <- clin.decomp$u[,1,drop=FALSE]
rownames(clin.pc) <- rownames(clin.pheno)[clin.decomp$rows.used]

gene.name = "Wfdc21";tissue.name = "Adipose"
gene.name <- "Myo19";tissue.name = "Adipose"
gene.name = "Lep";tissue.name = "Adipose"
gene.name = "Alox12"; tissue.name = "Islet"
gene.name = "Mogat2";tissue.name = "Adipose"

tissue.idx <- which(names(tissue.data) == tissue.name)
trans.id  <- tissue.data[[tissue.idx]]$annot.mrna$gene.id[which(tissue.data[[tissue.idx]]$annot.mrna$symbol == gene.name)]

plot_tissue_interaction(trans.id, FALSE)

#expr_across_tissues(transcript.id = trans.id, "CAST")
#expr_across_tissues(transcript.id = trans.id, "NZO")
#expr_across_alleles(transcript.id = trans.id, "Islet")

pdf("~/Desktop/test.pdf", width = 12, height = 7)
for(tx in 1:length(tissue.names)){
  expr_across_alleles(trans.id, tissue.names[tx])
}
dev.off()


gene.expr.data <- get_transcript_effects(trans.id, return.data = TRUE)
gene.geno <- gene.expr.data$genotypes

#does this locus have an effect on clinical PC1?
up.allele <- which(names(CCcolors) == "PWK")
down.allele <- which(names(CCcolors) == "WSB")

common.ind <- intersect(rownames(clin.pc), rownames(gene.geno))

dev.off()
par(mfrow = c(1,2))
up.test <- aov(clin.pc[common.ind,1]~as.factor(bin.vector(gene.geno[common.ind,up.allele])))
up.p <- anova(up.test)$"Pr(>F)"[1]
boxplot(clin.pc[common.ind,1]~as.factor(bin.vector(gene.geno[common.ind, up.allele])),
  xlab = paste(gene.name, "genotype"), ylab = "Clin PC1",
  main = paste("Up Allele\np = ", signif(up.p, 2)))

down.test <- aov(clin.pc[common.ind,1]~as.factor(bin.vector(gene.geno[common.ind,down.allele])))
down.p <- anova(down.test)$"Pr(>F)"[1]
boxplot(clin.pc[common.ind,1]~as.factor(bin.vector(gene.geno[common.ind, down.allele])),
  xlab = paste(gene.name, "genotype"), ylab = "Clin PC1",
  main = paste("Down Allele\np = ", signif(down.p, 2)))

dev.off()
#check for associations with individual phenotypes
common.ind <- intersect(rownames(clin.pc), rownames(gene.geno))
pdf("~/Desktop/gene_test.pdf", width = 12, height = 12)
par(mfrow = c(4,4))
for(i in 1:ncol(clin.pheno)){
  boxplot(rankZ(clin.pheno[common.ind,i])~as.factor(bin.vector(gene.geno[common.ind,up.allele])),
    main = paste(colnames(clin.pheno)[i], "Up Allele"), 
    xlab = gene.name, ylab = colnames(clin.pheno)[i])
  boxplot(rankZ(clin.pheno[common.ind,i])~as.factor(bin.vector(gene.geno[common.ind,down.allele])),
    main = paste(colnames(clin.pheno)[i], "Down Allele"), 
    xlab = gene.name, ylab = colnames(clin.pheno)[i])

}
dev.off()

#Does the locus have an effect on the clinical PC
#that could be mediated by transcription?
genotype_effect <- lm(clin.pc[common.ind,1]~gene.geno[common.ind,])
summary(genotype_effect)


plot_module_expr_v_clin_pc <- function(gene.list){

  module.info <- lapply(gene.list, function(x) get_transcript_effects(x, return.data = TRUE))
  module.info <- module.info[which(sapply(module.info, length) > 0)]
  module.expr <- lapply(module.info, function(x) x$expression)
  clin.cols <- colors.from.values(clin.pc, use.pheatmap.colors = TRUE)

  par(mfrow = c(2,2))
  for(tx in 1:length(tissue.names)){
    has.tx <- which(sapply(module.expr, function(x) length(which(names(x) == tissue.names[tx])) > 0))
    tx.expr <- sapply(has.tx, function(x) module.expr[[x]][[which(names(module.expr[[x]]) == tissue.names[tx])]])
    common.ind <- intersect(rownames(clin.pc), rownames(tx.expr))
    tx.expr[,which(trait.cor < 0)] <- tx.expr[,which(trait.cor < 0)] * -1
    tx.decomp <- plot.decomp(tx.expr, plot.results = FALSE)
    tx.pc <- tx.decomp$u
    rownames(tx.pc) <- rownames(tx.expr)[tx.decomp$rows.used]
    plot.with.model(tx.pc[common.ind,1], clin.pc[common.ind,1], 
      main = tissue.names[tx], xlab = "Expr PC1", ylab = "Expr PC2", 
      report = "cor.test")
  }
}

alox.net <- read.csv("~/Downloads/network_weights.csv")
u_gene <- unique(c(alox.net[,1], alox.net[,2]))
gene.idx <- match(u_gene, tissue.data[[2]]$annot.mrna$symbol)
gene.idx <- gene.idx[which(!is.na(gene.idx))]
gene.id <- tissue.data[[2]]$annot.mrna$gene.id[gene.idx]
plot_module_expr_v_clin_pc(gene.id)


eic.genes <- read.delim("~/Downloads/AA_metabolism.txt")
eic.idx <- lapply(tissue.data, function(x) match(eic.genes[,"Symbol"], x$annot.mrna$symbol))
eic.id <- vector(mode = "list", length = length(eic.idx))
names(eic.id) <- tissue.names
for(i in 1:length(eic.idx)){
  eic.idx[[i]] <- eic.idx[[i]][which(!is.na(eic.idx[[i]]))]
  eic.id[[i]] <- tissue.data[[i]]$annot.mrna$gene.id[eic.idx[[i]]]
}
plot_module_expr_v_clin_pc(eic.id[[4]])
```

## Gene Correlation Degree

In prepration for making a genetic architecture statement later
on, we want to calculate degree of each transcript in each 
transcript correlation network. High-loading transcripts might be
high-, low-, or mid-degree transcripts in the network.

Here we calculate degree without saving the full transcript by
transcript correlation network. They are very large.

```{r network_degree}
gene.deg.file <- here("Results", "Transcription_Networks", "Transcript_Degree.RDS")

if(!file.exists(gene.deg.file)){
  gene.deg <- vector(mode = "list", length = length(tissue.names))
  names(gene.deg) <- tissue.names

  for(tx in 2:length(tissue.names)){
      
    num.tx <- ncol(adj.expr[[tx]])
    pop.cor.result <- bigcor(adj.expr[[tx]])
    
    #pull out the matrix
    expr.cor <- pop.cor.result[1:num.tx, 1:num.tx]
    gene.deg[[tx]] <- colMeans(abs(expr.cor))
    names(gene.deg[[tx]]) <- colnames(adj.expr[[tx]])
        
    }
    
  saveRDS(gene.deg, gene.deg.file)
}
```