---
title: "local imputations"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
bibliography: hdm.bib 
---

```{r param}
rm(list = ls()) #clear out previous R environment
remove.pseudogenes = TRUE #if TRUE, removes pseudogenes and predicted genes 
tissue.names <- c("Adipose", "Islet", "Liver", "SkeletalMuscle")
tissue.cols <- c("orange", "#8dd3c7", "tan", "brown") #colors that are not related to the DO/CC colors
is.interactive = FALSE
overwrite.results = FALSE
```

## Introduction
In this workflow, we impute gene expression in the DO
using the effects of the local marker. We fit the measured
transcript abundance to the nearest marker with a kinship 
correction. We use the fitted values from that model as 
the imputed transcript values.

We also estimate measures of heritability, both overall and
distal. We calculate overall heritability using est_herit
on measured gene expression. We may not use this, but we 
want to have this. We calculate distal heritability by 
using est_herit on the residuals after fitting the linear 
model with the nearest marker. The product of the variance
of the residuals times their estimated heritability is the 
distal heritability component for the transcript.

We then compare these estimates of heritability to a measure 
of trait relevance, which is the maximum correlation of each
transcript with the traits.

We see that local heritability is anti-correlated with 
trait relevance. That is, transcripts whose abundance
is highly related to genotype at the local marker tend 
to have low correlations with traits. On the other hand,
transcripts with high trait relevance, tend to have low
local heritability.

We see a different relationship between trait relevance
and the distal component of heritability. Transcripts
with high trait relevance tend to have low-mid to mid-level 
distal heritability.

This workflow depends on results from 
1a.Tissue_Expression.Rmd, 1b.Trait_Selection.Rmd, and 
2a.Kinship_Expression_Traits.Rmd.

```{r here, message = FALSE, warning = FALSE}
library(here)
imputed.results.dir <- here("Results", "Imputed_Transcriptomes")
if(!file.exists(imputed.results.dir)){dir.create(imputed.results.dir, recursive = TRUE)}
```

Load code and libraries.

```{r load_code, message = FALSE, warning = FALSE}
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}

all.packages <- c("qtl2")
load_libraries(all.packages, personal.library = TRUE)
```

Read in data. We use expression and traits that have
already been adjusted for covariates and filtered for
pseudogenes in previous workflows. See code for details.


```{r read_transcriptomes}
#generated by 1a.Tissue_Expression.Rmd
#these transcripts are normalized and adjusted
#for covariates, but are not mean centered or 
#scaled
adj.expr <- readRDS(here("Data", "DO", "Tissue_Expression_Adjusted.RDS")) 
tissue.names <- names(adj.expr)
tissue.cols <- c("orange", "#8dd3c7", "tan", "brown") #colors that are not related to the DO/CC colors

#mean center and standardize gene expression so all variances are 1
scaled.expr <- lapply(adj.expr, function(x) apply(x, 2, scale))
for(tx in 1:length(scaled.expr)){
    dimnames(scaled.expr[[tx]]) <- dimnames(adj.expr[[tx]])
}

load(here("Data", "DO", "QTLViewer_Geno_V10.Rdata")) #ensembl.version, genoprobs, K, map, markers
covar <- readRDS(here("Data", "DO", "Clinical_Phenotype_Covariates.RDS")) #generated by 1b.Trait_Selection.Rmd
gene.tables <- readRDS(here("Data", "DO", "Gene_Tables.RDS")) #generated by 1a.Tissue_Expression.Rmd

local.coef <- readRDS(here("Results", "Transcriptomes", "eQTL_coef_local.RDS")) #generated in 1a.Tissue_Expression.Rmd
```


## Impute gene expression

Impute gene expression. We do this by fitting the transcript
values with the local marker genotype matrix. We use the predicted
values from fit1 as the locally imputed transcript. Because the
transcripts are all scaled to have a variance of 1, the variance
of the fitted values is the variance explained by the local genotype.

When we do this, we can also estimate the variance explained by 
the distal genome. To do this, we estimate the heritability of 
the residuals of each transcript abundance after controlling 
for the local eQTL.


```{r kin}
kin.file <- here("Data", "DO", "overall.kinship.RDS")
if(!file.exists(kin.file)){
    Kg = calc_kinship(genoprobs, "overall")
    saveRDS(Kg, kin.file)
}else{
    Kg <- readRDS(kin.file)
}
```

```{r overall_herit}
transcript.herit.file <- here("Results", "Transcriptomes", "Transcript_Heritability.RDS")
if(!file.exists(transcript.herit.file)){
    all.tx.herit <- vector(mode = "list", length = length(tissue.names))
    names(all.tx.herit) <- tissue.names
    for(tx in 1:length(tissue.names)){
        tx.herit <- est_herit(scaled.expr[[tx]], kinship = kin)
        all.tx.herit[[tx]] <- tx.herit
    }
    saveRDS(all.tx.herit, transcript.herit.file)
}
```

## Imputation

Here we use the nearest marker to each gene to impute
gene expression. We use fit1 with a kinship correction
and take the fitted values of the model as the imputed
expression.

We also calculate variance explained by the distal component 
of gene expression. To get this we estimates the heritability 
of the residuals of each fitted model above.


```{r impute_transcriptomes}
save.every = 100

for(tx in 1:length(tissue.names)){
    start.idx <- 1
    distal.herit.file <- file.path(imputed.results.dir, 
      paste0("Distal_Heritability_", tissue.names[tx], ".RDS"))
    transcript.id <- colnames(adj.expr[[tx]])

    local.imp.file <- file.path(imputed.results.dir, paste0("Local_Imputation_Expr_", tissue.names[tx], ".RDS"))
    
    #check to see if we've already saved some results that we can use
    #we only need to use one of the files to figure this out
    if(file.exists(distal.herit.file) && !overwrite.results){
      herit.resid <- readRDS(distal.herit.file)
      start.idx <- max(which(!is.na(herit.resid[,1]))) + 1
      local.imp <- readRDS(local.imp.file)
    }
      
  #start where we left off, or at the beginning if there are no data
  #or we are overwriting previous data.
  if(start.idx < length(transcript.id)){
    sink(file.path(imputed.results.dir, paste0("progress_", tissue.names[tx], ".txt")))

    #if we are starting fresh...
    if(!file.exists(distal.herit.file) || overwrite.results){
      herit.resid <- matrix(NA, ncol = 3, nrow = length(transcript.id))
      rownames(herit.resid) <- transcript.id
      colnames(herit.resid) <- c("local_var", "residual_variance", "residual_heritability")

      local.imp <- matrix(NA, ncol = length(transcript.id), nrow = nrow(adj.expr[[tx]]))
      dimnames(local.imp) <- dimnames(adj.expr[[tx]])
    }
  
    #start where we left off, or at the beginning, if that is warranted
    for(tr in start.idx:length(transcript.id)){
      print(tr)
      
      gene.info.idx <- which(gene.tables[[tx]]$gene.id == transcript.id[tr])
      nearest.marker <- gene.tables[[tx]]$nearest.marker.id[gene.info.idx]

       #we won't be able to find nearest markers for genes on Y or MT chromosomes
      if(nearest.marker == ""){next()}  

      marker.geno <- pull_genoprobpos(genoprobs, marker = nearest.marker)

      trans.expr <- scaled.expr[[tx]][,transcript.id[tr]]
  
      #with kinship
      #I am keeping the variance of the fitted values to use as the local variance explained
      #reserve as much of the signal from kinship as we can by performing a kinship correction
      model.kin <- fit1(marker.geno, trans.expr, kinship = Kg)
      imp.local <- model.kin$fitted
      local.imp[,tr] <- imp.local
      distal.resid <- model.kin$resid
      result <- c(var(imp.local), var(distal.resid), est_herit(distal.resid, Kg))
      herit.resid[tr,] <- result

      if(tr %% save.every == 0){
        saveRDS(herit.resid, distal.herit.file)
        saveRDS(local.imp, local.imp.file)
      }
    }
    saveRDS(herit.resid, distal.herit.file)
    saveRDS(local.imp, local.imp.file)
    sink()
  }
}
```

```{r read_imp}
local.imp.files <- list.files(imputed.results.dir, pattern = "Local", full.names = TRUE)
tx.local.imp <- lapply(local.imp.files, readRDS)
```


## Correlation Check {.tabset .tabset-fade .tabset-pills}

### Imputed and measured 

Check the correlations of the imputed transcripts with
transcript measurements. Transcripts with higher local 
LOD scores should have higher correlations to the 
measured transcript.

```{r check_correlations, results = "asis", fig.height = 4, fig.width = 8}
for(tx in 1:length(tissue.names)){
    cat("###", tissue.names[tx], "\n")
    measured <- scaled.expr[[tx]]
    imputed <- tx.local.imp[[tx]]

    common.tx <- intersect(colnames(measured), colnames(imputed))
    common.ind <- intersect(rownames(measured), rownames(imputed))
    all.cor <- sapply(1:length(common.tx), 
        function(x) cor(measured[common.ind,common.tx[x]], 
        imputed[common.ind,common.tx[x]]))
    local.lod <- local.coef[[tx]]$lod[match(common.tx, local.coef[[tx]]$gene.id)]
    par(mfrow = c(1,2))
    hist(all.cor, main = "Histogram of Correlations", 
        xlab = "Correlation", xlim = c(0,1))
    plot(local.lod, all.cor, xlab = "local LOD score", 
        main = "Correlation v. LOD",
        ylab = "Correlation", ylim = c(0,1))
    mtext(tissue.names[tx], side = 3, outer = TRUE, line = -1.5, font = 2)
    cat("\n\n")

}
```

### Distal and local heritabilities

Check that local and distal heritabilities are anti-correlated
and never sum to more than 1.

```{r local_distal_check}
distal.herit <- lapply(tissue.names, function(x) readRDS(file.path(imputed.results.dir, 
      paste0("Distal_Heritability_", tissue.names[tx], ".RDS"))))


for(tx in 1:length(tissue.names)){
    local.var <- apply(tx.local.imp[[tx]], 2, var)
    distal.var <- distal.herit[[tx]][,"residual_variance"]
    common.tx <- intersect(names(local.var), names(distal.var))
    plot(local.var[common.tx], distal.var[common.tx], xlab = "Local Variance Explained",
    ylab = "Distal Variance Explained")
}

```

## Heritability and Trait Relevance

We want to examine the relationship between heritability and 
trait relevance. We define trait relevance here as the correlation
between the transcript and the trait. Heritability we break into
two components: local heritability and distal heritability.

Local heritability can be defined either with the local LOD
score, or the variance explained by the local marker. We will
use the variance explained by the local marker to be more analogous
both with the human literature and with the distal heritability
measure. Distal heritability is the variance eplained by the 
kinship matrix for the transcript levels after controlling for the
local marker.


### Local Heritability and Trait Relavance {.tabset .tabset-fade .tabset-pills}

The following plots show the correlation between the 
variance explained by the local marker and the trait
relevance (maximum correlation between the transcript
and all traits.)

#### Variance Explained - Loess

```{r local_v_traits, fig.height = 8, fig.width = 8}
trait.cor.files <- list.files(here("Results", "Transcriptomes"), 
    pattern = "Maximum_Trait_Correlation", full.names = TRUE) #generated in 2a.Kinship_Expression_Traits.Rmd
trait.cor <- lapply(trait.cor.files, readRDS)
names(trait.cor) <- tissue.names

par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){
    imp.expr <- tx.local.imp[[tx]]
    var.exp <- apply(imp.expr, 2, var)
    common.tx <- intersect(names(trait.cor[[tx]]), names(var.exp))

    binned_ts <- bin.vector2(var.exp[common.tx], seq(0, 1, 0.05))
    bin_means <- sapply(binned_ts, function(x) mean(abs(trait.cor[[tx]][names(x)]), na.rm = TRUE))
    x_means <- sapply(binned_ts, mean)

    plot(var.exp[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Local Heritability", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], pch = 16)
    points(x_means, bin_means, col = "lightblue", type = "l", lwd = 3)
}
```

#### Variance Explained - Linear

The following plots are the same as above, but show the linear fit.

```{r local_lin_fit, fig.height = 8, fig.width = 8}
par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){
    imp.expr <- tx.local.imp[[tx]]
    var.exp <- apply(imp.expr, 2, var)
    common.tx <- intersect(names(trait.cor[[tx]]), names(var.exp))

    plot.with.model(var.exp[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Local Heritability", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], report = "cor.test")
}
```

#### LOD - Linear

The following plots show the relationship between trait
relevance and local LOD score.

```{r herit_lod, fig.width = 8, fig.height = 8}
par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    local.lod <- unlist(local.coef[[tx]][,"lod"])
    names(local.lod) <- unlist(local.coef[[tx]][,"gene.id"])
    common.tx <- intersect(names(local.lod), names(trait.cor[[tx]]))

    plot.with.model(local.lod[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Local Heritability", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], report = "cor.test")
}

```

### Distal Heritability and Trait Relavance {.tabset .tabset-fade .tabset-pills}

#### Variance Explained - Loess

```{r, fig.width = 8, fig.height = 8}

herit.files <- list.files(imputed.results.dir, pattern = "Distal", full.names = TRUE)
tx.herit <- lapply(herit.files, readRDS)

par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){
    distal.herit <- tx.herit[[tx]][,"residual_variance"]*tx.herit[[tx]][,"residual_heritability"]
    common.tx <- intersect(names(trait.cor[[tx]]), names(distal.herit))

    binned_ts <- bin.vector2(distal.herit[common.tx], seq(0, 1, 0.05))
    bin_means <- sapply(binned_ts, function(x) mean(abs(trait.cor[[tx]][names(x)]), na.rm = TRUE))
    x_means <- sapply(binned_ts, mean)

    plot(distal.herit[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Heritability", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], pch = 16)
    points(x_means, bin_means, col = "lightblue", type = "l", lwd = 3)
}
```

#### Variance Explained - Linear

The following figures show the same data, but with linear fit models.

```{r lin_fig, fig.width = 8, fig.height = 8}
par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){
    distal.herit <- tx.herit[[tx]][,"residual_variance"]*tx.herit[[tx]][,"residual_heritability"]
    common.tx <- intersect(names(trait.cor[[tx]]), names(distal.herit))

    plot.with.model(distal.herit[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Heritability", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], report = "cor.test")
}
```
