---
title: "local imputations"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
bibliography: hdm.bib 
---

```{r param}
remove.pseudogenes = TRUE #if TRUE, removes pseudogenes and predicted genes 
tissue.names <- c("Adipose", "Islet", "Liver", "SkeletalMuscle")
tissue.cols <- c("orange", "#8dd3c7", "tan", "brown") #colors that are not related to the DO/CC colors
```

## Introduction
In this workflow, we use local eQTL effects estimated previously
by the Churchill lab to impute the local component of gene 
expression in the DO. This workflow depends on results from 
1a.Tissue_Expression.Rmd, 1b.Trait_Selection.Rmd, and 
2a.Kinship_Expression_Traits.Rmd.

Set up results directory.

```{r here, message = FALSE, warning = FALSE}
library(here)
results.dir <- here("Results", "Imputed_Transcriptomes")
if(!file.exists(results.dir)){dir.create(results.dir, recursive = TRUE)}
```

Load code and libraries.

```{r load_code, message = FALSE, warning = FALSE}
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}

all.packages <- c("qtl2")
load_libraries(all.packages, personal.library = TRUE)
```

Read in data. We use experssion and traits that have
already been adjusted for covariates and filtered for
pseudogenes in previous workflows. See code for details.


```{r read_transcriptomes}
adj.expr <- readRDS(here("Data", "DO", "Tissue_Expression_Adjusted.RDS")) #generated by 1a.Tissue_Expression.Rmd
tissue.names <- names(adj.expr)
tissue.cols <- c("orange", "#8dd3c7", "tan", "brown") #colors that are not related to the DO/CC colors

load(here("Data", "DO", "QTLViewer_Geno_V10.Rdata")) #ensembl.version, genoprobs, K, map, markers
covar <- readRDS(here("Data", "DO", "Clinical_Phenotype_Covariates.RDS")) #generated by 1b.Trait_Selection.Rmd
gene.tables <- readRDS(here("Data", "DO", "Gene_Tables.RDS")) #generated by 1a.Tissue_Expression.Rmd

#tables with local eQTL coefficients and LOD scores.
#generated in 1a.Tissue_Expression.Rmd
local.coef <- readRDS(here("Results", "Transcriptomes", "eQTL_coef_local.RDS"))
```

Impute gene expression from local allele effects.

```{r impute}
impute_transcript <- function(marker.id, marker.coef){
    marker.geno <- pull_genoprobpos(genoprobs, marker = marker.id)
    imp.expr <- marker.geno%*%t(marker.coef)
    names(imp.expr) <- rownames(marker.geno)
    return(imp.expr)
}

imputed.expr <- vector(mode = "list", length = length(tissue.names))
names(imputed.expr) <- tissue.names

imputed.expr.files <- file.path(results.dir, paste0("Locally_Imputed_Expr_", tissue.names, ".RDS"))

for(tx in 1:length(local.coef)){
    
    if(!file.exists(imputed.expr.files[tx])){
        tissue.coef <- local.coef[[tx]]        
        tissue.imp <- sapply(1:nrow(tissue.coef), 
            function(x) impute_transcript(marker.id = unlist(tissue.coef[x,"marker.id"]),
            marker.coef = matrix(unlist(tissue.coef[x,LETTERS[1:8]]), nrow = 1)))
        colnames(tissue.imp) <- unlist(tissue.coef[,1])
        imputed.expr[[tx]] <- tissue.imp
        saveRDS(tissue.imp, imputed.expr.files[tx])
    }else{
        imputed.expr[[tx]] <- readRDS(imputed.expr.files[tx])
    }
}

```

## Correlation Check {.tabset .tabset-fade .tabset-pills}

Check the correlations of the imputed transcripts with
transcript measurements. Transcripts with higher local 
LOD scores should have higher correlations to the 
measured transcript.

```{r check_correlations, results = "asis", fig.height = 4, fig.width = 8}
for(tx in 1:length(tissue.names)){
    cat("###", tissue.names[tx], "\n")
    measured <- adj.expr[[tx]]
    imputed <- imputed.expr[[tx]]

    common.tx <- intersect(colnames(measured), colnames(imputed))
    common.ind <- intersect(rownames(measured), rownames(imputed))
    all.cor <- sapply(1:length(common.tx), 
        function(x) cor(measured[common.ind,common.tx[x]], 
        imputed[common.ind,common.tx[x]]))
    local.lod <- unlist(local.coef[[tx]][,"lod"])
    par(mfrow = c(1,2))
    hist(all.cor, main = "Histogram of Correlations", 
        xlab = "Correlation", xlim = c(0,1))
    plot(local.lod, all.cor, xlab = "local LOD score", 
        main = "Correlation v. LOD",
        ylab = "Correlation", ylim = c(0,1))
    mtext(tissue.names[tx], side = 3, outer = TRUE, line = -1.5, font = 2)
    cat("\n\n")
}
```

## Heritability

What is the heritability of each transcript
after adjusting for the local eQTL?


```{r kin}
kin.file <- here("Data", "DO", "overall.kinship.RDS")
if(!file.exists(kin.file)){
    Kg = calc_kinship(genoprobs, "overall")
    saveRDS(Kg, kin.file)
}else{
    Kg <- readRDS(kin.file)
}
```


```{r herit}
transcript_herit <- function(gene.expr, marker.id){
    marker.geno <- pull_genoprobpos(genoprobs, marker = marker.id)
    model.fit <- fit1(marker.geno, gene.expr, kinship = Kg)
    adj_expr <- model.fit$resid
    #adj_expr <- adjust(gene.expr, marker.geno)
    herit <- est_herit(adj_expr, Kg)
    return(as.numeric(herit))
}

tx.herit <- vector(mode = "list", length = length(tissue.names))
names(tx.herit) <- tissue.names
for(tx in 1:length(tissue.names)){
    herit.file <- file.path(results.dir, paste0("Transcript_Heritability_", tissue.names[tx], ".RDS"))
    if(!file.exists(herit.file)){
        expr <- adj.expr[[tx]]
        tissue.coef <- local.coef[[tx]]
        gene.id <- tissue.coef[,"gene.id"]
        eqtl.marker <- tissue.coef[,"marker.id"]
        gene.herit <- sapply(1:length(gene.id), 
            function(x) transcript_herit(gene.expr = tissue.expr[,gene.id[[x]],drop=FALSE], 
                marker.id = eqtl.marker[[x]]))
        names(gene.herit) <- gene.id
        saveRDS(gene.herit, herit.file)
    }else{
        gene.herit <- readRDS(herit.file)
    }
    tx.herit[[tx]] <- gene.herit
}
```

How does heritability relate to local LOD score?
How does heritability relate to trait correlations?

The following plots show the relationship between
heritability and correlation with traits.

```{r, fig.width = 8, fig.height = 8}
trait.cor.files <- list.files(here("Results", "Transcriptomes"), 
    pattern = "Maximum_Trait_Correlation", full.names = TRUE) #generated in 2a.Kinship_Expression_Traits.Rmd
trait.cor <- lapply(trait.cor.files, readRDS)
names(trait.cor) <- tissue.names
par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){
    common.tx <- intersect(names(trait.cor[[tx]]), names(tx.herit[[tx]]))
    plot.with.model(tx.herit[[tx]][common.tx], trait.cor[[tx]][common.tx],
        xlab = "Heritability", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], report = "cor.test")
}
```

The following plots show the relationship between heritability
and local LOD score.

```{r herit_lod, fig.width = 8, fig.height = 8}
par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    local.lod <- unlist(local.coef[[tx]][,"lod"])
    names(local.lod) <- unlist(local.coef[[tx]][,"gene.id"])
    common.tx <- intersect(names(local.lod), names(tx.herit[[tx]]))
    plot.with.model(tx.herit[[tx]][common.tx], local.lod[common.tx],
        report = "cor.test", xlab = "Heritability", ylab = "Local LOD")
}

```

## Imputation of Distal Component

Can we impute the distal component of the gene expression?
We could fit a linear model to PCs of the kinship matrix, and
then predict expression from that. But how many PCs do we use?
Too few, and we can't predict gene expression. Too many, and
we overfit the model.

Do we just say that the distal component is everything that
is not the local component? No, because what if a gene has
no local eQTL, but low heritability?

```{r distal_imputation, eval = FALSE}
n.pc <- 450
kin.decomp <- plot.decomp(Kg, pc = n.pc, plot.results = FALSE)
kin.pc <- kin.decomp$u[,1:n.pc]
null.kin <- sapply(1:ncol(kin.pc), function(x) rnorm(nrow(kin.pc)))
rownames(kin.pc) <- rownames(null.kin) <- rownames(Kg)
cumulative.var <- cumsum(kin.decomp$var.exp)

for(tx in 1:length(tissue.names)){
    distal.file <- file.path(results.dir, 
        paste0("Distally_Imputed_Expr_", tissue.names[tx], ".RDS"))
    if(!file.exists(distal.file)){

        tissue.coef <- local.coef[[tx]]
        tissue.expr <- adj.expr[[tx]]

        var.exp <- matrix(NA, nrow = nrow(tissue.coef), ncol = 2)
        colnames(var.exp) <- c("local", "distal")

        local.imp.mat <- distal.imp.mat <- matrix(NA, nrow = nrow(tissue.expr), 
            ncol = nrow(tissue.coef))
        rownames(local.imp.mat) <- rownames(distal.imp.mat) <- rownames(tissue.expr)
        colnames(local.imp.mat) <- colnames(distal.imp.mat) <- unlist(tissue.coef[,1])

        test.n <- 200
        for(idx in 1:test.n){
            #unlist(local.coef[[tx]][idx,])
            print(idx)
            gene.id <- unlist(tissue.coef[idx,1])
            gene.expr <- tissue.expr[,gene.id,drop=FALSE]
            marker.id <- unlist(tissue.coef[idx,2])
            marker.chr <- strsplit(marker.id, "_")[[1]][1]
            marker.geno <- pull_genoprobpos(genoprobs, marker = marker.id)
            gene.herit <- tx.herit[[tx]][gene.id]
            
            local.model <- fit1(marker.geno, gene.expr, kinship = Kg)
            local.imp <- local.model$fitted
            resid.expr <- local.model$resid
            #as.numeric(est_herit(resid.expr, Kg))
            local.imp.mat[,idx] <- local.imp
            
            distal.model <- fit1(genoprobs = NULL, pheno = gene.expr, kinship = Kg)
            #distal.model <- fit1(genoprobs = marker.geno, pheno = resid.expr, kinship = Kg)
            distal.imp <- distal.model$fitted
            #test <- distal.model$resid
            #plot(resid.expr, test)


            distal.imp.mat[,idx] <- distal.imp

            #plot.with.model(local.imp, gene.expr)
            #plot.with.model(distal.imp, gene.expr)
            #plot.with.model(distal.imp, adj_expr)

            local.var.exp <- plot.with.model(x = local.imp, y = gene.expr, plot.results = FALSE) # nolint
            distal.var.exp <- plot.with.model(distal.imp, gene.expr, plot.results = FALSE)
            #plot.with.model(distal.imp, adj_expr) 
            var.exp[idx,1] <- local.var.exp[1]*100
            var.exp[idx,2] <- distal.var.exp[1]*100
        }

        test.cor <- sapply(1:test.n, function(x) cor(local.imp.mat[,x], distal.imp.mat[,x]))
        hist(test.cor)
        
        plot.with.model(var.exp[,1], var.exp[,2], report = "cor.test",
            xlab = "local", ylab = "distal")
    }
    
}


```