---
title: "Cell Type Validation"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction

The purpose of this workflow is to look at markers of 
specific cell types in the adipose tissue to see if we
can see elevated levels of specific cell types in different
subsets of mice. 

Ehrlund et al. 2017 published a set of markers specific
to different cell types in human adipose tissue. We can
use these markers to look for cell composition variation
in the DO mice.

```{r load_code}
rm(list = ls())

#is.interactive = TRUE
library("here")

exp.name = "tissue_together-_-complete_mediation-germline_kinship"
data.results.dir <- here("Results", "High_Dim_Med", exp.name)

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "gprofiler2", "RColorBrewer", "GEOquery", "limma", "vioplot")
load_libraries(all.packages, personal.library = TRUE)
```


```{r read_data}

model_scores <- readRDS(file.path(data.results.dir, "Model_Scores_0.RDS"))

transcript_loadings <- readRDS(file.path(data.results.dir, "Loadings_Transcripts_0.RDS"))
tissue.names <- names(transcript_loadings)

adj.expr <- readRDS(here("Results", "Data", "Tissue_Expression_Adjusted.RDS")) #generated by 1a.Tissue_Expression.Rmd

mus.hum <- read.delim(here("Data", "general", "human.mouse.orthologs.txt"))

#cell_type_data
cell_type_row <- read.delim(here("Data", "Human", "Ehrlund_et_al", "Selected_enriched_genes_all_fractions.txt"),
    skip = 2, nrows = 1, strip.white = TRUE, header = FALSE)
cell_types <- cell_type_row[which(!is.na(cell_type_row))]
cell_type_col <- lapply(cell_types, function(x) c(which(cell_type_row == x), which(cell_type_row == x)+1))
names(cell_type_col) <- cell_types

cell_type_table <- read.delim(here("Data", "Human", "Ehrlund_et_al", "Selected_enriched_genes_all_fractions.txt"),
    skip = 3, strip.white = TRUE)
cell_tables <- lapply(cell_type_col, function(x) cell_type_table[,x])

matched_cell_tables <- vector(mode = "list", length = length(cell_tables))
names(matched_cell_tables) <- cell_types
#find orthologs of cell type data
for(ct in 1:length(cell_tables)){
    hum.gene.name <- cell_tables[[ct]][,1]
    common.name <- intersect(hum.gene.name, mus.hum[,"Human.Gene.Name"])
    matched.table <- mus.hum[match(common.name, mus.hum[,"Human.Gene.Name"]),]
    matched_cell_tables[[ct]] <- matched.table
}
```

The plot below shows the relative distribution of loadings across
cell types. Negative loadings mean that higher proportions of these
cells are associated with less obesity and insulin resistance.

Positive loadings mean that higher proportions of these
cells are associated with greater obesity and insulin 
resistance.

So for example, if the tissue is mostly adipose cells, as opposed
to some adipose cells and lots of macrophages, that indicates
a healthier animal. Tissue that has a higher abundance of macrophages
is likely to be from an obese animal with higher insulin resistance.

Progenitor cells don't seem to be different in obese and lean animals.
This is not really what they saw in Ehrlund et al. They saw that obese
people had slightly fewer progenitor cells, and more macrophages.

```{r loadings}

adipose.loadings <- transcript_loadings$Adipose

matched.genes <- lapply(matched_cell_tables, function(x) intersect(x[,"Mouse.Ortholog.Ensembl"], rownames(adipose.loadings)))
cell_type_loadings <- lapply(matched.genes, function(x) adipose.loadings[x,])

all.load <- unlist(cell_type_loadings)
fact <- as.factor(unlist(lapply(names(cell_type_loadings), function(x) rep(x, length(cell_type_loadings[[x]])))))
vioplot(all.load~fact, xlab = "Cell Type", ylab = "Loading", col = "lightgray")
abline(h = 0)
#stripchart(cell_type_loadings, vertical = TRUE, add = TRUE, pch = 16, col = "darkgray", method = "jitter")

```

The following figures show that the cell-type specific 
signatures are highly correlated with phenotypic outcome
in the DO mice. I'm not sure why the progenitor cell 
profile does so well at predicting lean and obese animals
when it didn't seem that the loadings were different
between them.

```{r cell_types, fig.width = 8, fig.height = 8}
adipose.expr <- adj.expr$Adipose
scaled.expr <- apply(adipose.expr, 2, scale)
dimnames(scaled.expr) <- dimnames(adipose.expr)

adipose.model <- model_scores$Adipose
cell.type.pred.mat <- matrix(NA, ncol = length(cell_types), nrow = nrow(scaled.expr))
colnames(cell.type.pred.mat) <- cell_types
rownames(cell.type.pred.mat) <- rownames(scaled.expr)

par(mfrow = c(2,2))
for(ct in 1:length(cell_types)){
    ct_genes <- intersect(matched_cell_tables[[ct]][,"Mouse.Ortholog.Ensembl"], colnames(scaled.expr))
    ct_expr <- scaled.expr[,ct_genes]
    loaded_expr <- apply(ct_expr, 1, function(x) x*adipose.loadings[ct_genes,])
    #pheatmap(loaded_expr)
    pred_expr <- colMeans(loaded_expr)
    cell.type.pred.mat[,ct] <- pred_expr
    common.ind <- intersect(names(pred_expr), rownames(adipose.model))
    plot.with.model(pred_expr[common.ind], adipose.model[common.ind,"Outcome"], 
        main = paste(cell_types[ct], "\n", length(ct_genes), "Genes"), 
        xlab = "Cell Type Prediction", ylab = "Phenotypic Outcome")
}
```

The following shows the loaded transcripts across all individuals.
The matrix is scaled by column, to show relative strenghts of each
transcript for the individual. This shows us that individuals with 
higher phenotype scores have higher proportions of macrophages and
and lower proportions of other leukocytes are more obese (?)

```{r cell_type_heat, fig.width = 9, fig.height = 4}
common.ind <- intersect(rownames(scaled.expr), rownames(adipose.model))
df <- data.frame("Phenotype Score" = adipose.model[common.ind,"Outcome"])

pheno.order <- order(df[,1])
pheatmap(t(cell.type.pred.mat[rownames(df)[pheno.order],]), cluster_col = FALSE, 
    annotation_col = df, show_colnames = FALSE, scale = "column")



```

```{r ind_expr}

thresh = 4
par(mfrow = c(2,2))
for(ct in 1:length(cell_types)){
    ct_genes <- intersect(matched_cell_tables[[ct]][,"Mouse.Ortholog.Ensembl"], colnames(scaled.expr))
    gene.names <- matched_cell_tables[[ct]][match(ct_genes, matched_cell_tables[[ct]][,"Mouse.Ortholog.Ensembl"]),"Mouse.Ortholog.Name"]
    ct_expr <- scaled.expr[,ct_genes]
    colnames(ct_expr) <- gene.names
    loaded_expr <- apply(ct_expr, 1, function(x) x*adipose.loadings[ct_genes,])
    common.ind <- intersect(rownames(ct_expr), rownames(adipose.model))
    ordered.ind <- rownames(adipose.model)[order(adipose.model[common.ind,"Outcome"])]
    df <- data.frame("Phenotype" = adipose.model[common.ind,"Outcome"])
    mat <- ct_expr[ordered.ind,]
    mat[which(mat > thresh)] <- thresh
    mat[which(mat < thresh*-1)] <- thresh*-1
    pheatmap(mat, cluster_rows = FALSE, show_rownames = FALSE,
        annotation_row = df, main = cell_types[ct])

    pheatmap(loaded_expr[,ordered.ind], cluster_cols = FALSE, annotation_col = df, 
        show_colnames = FALSE, main = paste(cell_types[ct], "Loaded"))

    plot(as.vector(mat), as.vector(loaded_expr[,ordered.ind]))

}

