---
title: "Literature Validation"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow is to look into the literature
to see how much is known about genes with high MGE vs. 
genes with low MGE.

```{r setup}
library(here)
mge.results <- "tissue_together-full_imp-_-complete_mediation"
mge.results.dir <- here("Results", "High_Dim_Med", mge.results)

if(length(grep("full_imp", mge.results)) > 0){
    #were local effects included in MGE calculations?
    #This can be TRUE or FALSE, if the full model imputation was used.
    kinship.effect.only <- TRUE
}else{
    #if the local imputation was used, we can't remove kinship effects.
    kinship.effect.only <- FALSE
    if(kinship.effect.only){stop("Can only remove local effects if full genetic model was used.")}
}

kinship.options <- c("local_effects_included", "local_effects_removed")
kinship.text <- kinship.options[as.numeric(kinship.effect.only)+1]

```

This workflow used `r mge.results`.
And `r kinship.text`.

```{r load_code, message = FALSE, warning = FALSE}
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}

all.packages <- c("pheatmap", "easyPubMed", "fgsea")
load_libraries(all.packages, personal.library = TRUE)
```

```{r read_previous_results}
mge <- readRDS(here("Results", "High_Dim_Med", mge.results, paste0("Population_MGE_Loadings_", kinship.text, ".RDS")))
transcript.loadings <- readRDS(here("Results", "High_Dim_Med", mge.results, "Loadings_Transcripts.RDS"))
gene.tables <- readRDS(here("Data", "DO", "Gene_Tables.RDS"))
```

```{r, eval = FALSE}
#Correlation between MGE and transcript loading
#Genes with mid-range loadings have highest MGE
#because of high LOD scores
#transcripts with high loadings have a large effect
#on phenotype, but don't vary much in expression
#across animals. Transcripts with high MGE
#have some effect on phenotype, and have heritable
#variation across individuals.
par(mfrow = c(2,2))
for(tx in 1:length(mge)){
    if(length(mge[[tx]]) > 0){
        common.genes <- intersect(names(mge[[tx]]), rownames(transcript.loadings[[tx]]))
        plot.with.model(abs(mge[[tx]][common.genes]), 
            abs(transcript.loadings[[tx]][common.genes,]), xlab = "MGE",
            ylab = "Loading", main = names(mge)[tx])
    }
}
```

## T2D gene lists

There are several sources of gene lists associating 
individual genes with type 2 diabetes. One is these
sources is the T2D knowledge portal, which is
accessible through the Human Genetics Amplifier (HuGeAMP)
BioIndex. 

The portal provides "bottom-line integrative analysis"
scores described here: (https://t2d.hugeamp.org/help.html?page=911)
This method uses a meta-analysis that combines similar
studies to increase power to detect trait-associated variants
that are not necessarily significant in individual studies.
This analysis also accounts for the possibility of samples
overlapping across multiple studies. These analyses combine
results from studies including multiple ancestries.

The code for downloading T2D-associated genes from the
knowledge portal is below. This data set is re-downloaded
each time the workflow is run.

```{r huge, warning = FALSE}
con <- curl::curl("https://bioindex.hugeamp.org/api/bio/query/huge-phenotype?q=t2d&fmt=row")
huge_data = readLines(con, -1)
huge_scores = jsonlite::fromJSON(huge_data)

# Add codes for evidence levels
# Compelling: HuGE Score >= 350 | Extreme: >=100 | Very Strong: >=30 | Strong: >=10 | Moderate: >=3 | Anecdotal: >1 | No Evidence: <=1

t2d.gene.table <- huge_scores$data
score.codes <- c("compelling" = 350, "extreme" = 100, 
    "very strong" = 30, "strong" = 10, "moderate" = 3, 
    "anecdotal" = 1, "no evidence" = 0)

code_fun <- function(scores, code.vector, min.score, label){
    need.code.idx <- which(is.na(code.vector))
    meet.criteria <- which(scores >= min.score)
    overlap.idx <- intersect(need.code.idx, meet.criteria)
    code.vector[overlap.idx] <- label
    return(code.vector)
}

evidence.code <- rep(NA, nrow(t2d.gene.table))
for(i in 1:length(score.codes)){
    evidence.code <- code_fun(scores = t2d.gene.table[,"huge"], 
        code.vector = evidence.code, min.score = score.codes[i], 
        label = names(score.codes)[i])
}

```


Using each evidence category is a little too granular, 
so we grouped the genes into categories based on their
T2D evidence. These groups are shown below:

```{r cat_groups}
cat.groups <- list(
    "T2D" = c("compelling", "extreme", "very strong", "strong", "moderate"), 
    "Weak Association" = c("anecdotal"), 
    "No Association" = c("no evidence"))

for(i in 1:length(cat.groups)){
    cat(paste0(names(cat.groups)[i], ": "))
    cat(paste(cat.groups[[i]], collapse = ", "), "\n")
}
```


We converted the human gene names to mouse orthologs.
There were no genes in the table with "no association" so for 
this category we use the remaining genes in our ortholog table
that were not included in the T2D list. 

```{r subset}

gene.sets <- lapply(cat.groups, 
    function(x) t2d.gene.table[which(evidence.code %in% x),1])
names(gene.sets) <- names(cat.groups)
no.assoc <- setdiff(hum.mus.ortho[,"Human.Gene.Name"], unlist(gene.sets))
gene.sets[[length(cat.groups)]] <- no.assoc
```

## Genes by category

The following table shows how many genes are in each of the
category. There are a lot of genes in each group. Maybe too
many to get reasonable enrichment values from GSEA?

```{r counts}
score.counts <- sapply(gene.sets, length)
barplot_with_num(score.counts, ylab = "Number of Genes")
```


```{r ortho}
hum.mus.ortho <- read.delim(here("Data", "general", "human.mouse.orthologs.txt")) 

ortho.tables <- lapply(gene.sets, 
    function(x) hum.mus.ortho[match(x, hum.mus.ortho[,"Human.Gene.Name"]),])

gsea.ensembl.list <- lapply(ortho.tables, 
    function(x) x[which(!is.na(x[,"Mouse.Ortholog.Ensembl"])),"Mouse.Ortholog.Ensembl"])

```

## Loadings by category

The boxplot below shows the distributions of transcript loadings
for orthologs of genes with varying degrees of association
with T2D. The loadings seem to be evenly distributed across
all classes.

```{r loadings}
t2d.gene.loadings <- vector(mode = "list", length = length(gene.sets))
names(t2d.gene.loadings) <- names(gene.sets)

for(i in 1:length(gsea.names.list)){
    t2d.gene.loadings[[i]] <- lapply(transcript.loadings, 
        function(x) abs(x[which(rownames(x) %in% gsea.ensembl.list[[i]])]))
}
plot.grouped.boxes(t2d.gene.loadings, 
    main = "|Loadings| of orthologs to T2D-associated genes in humans",
    print.vals = NA)
```

## MGE by category

The boxplot below shows the distributions of MGE for orthologs 
of genes with varying degrees of association with T2D. The MGE
scores seem to be evenly distributed across all classes.


```{r mge}
#MGE
t2d.gene.mge <- vector(mode = "list", length = length(gene.sets))
names(t2d.gene.mge) <- names(gene.sets)

for(i in 1:length(gsea.ensembl.list)){
    t2d.gene.mge[[i]] <- lapply(mge, 
        function(x) abs(x[which(names(x) %in% gsea.ensembl.list[[i]])]))
}
plot.grouped.boxes(t2d.gene.mge, 
    main = "|MGE| of orthologs to T2D-associated genes in humans",
    print.vals = NA)
```

## GSEA sorted by loading

We then used GSEA to see if any of the groups of 
genes were enriched among our genes with high
loadings or high MGE. 

The following plot shows the -log10(p values) for
enrichments of each gene set among our highest
loading transcripts. The genes with large loadings
do seem to be slightly enriched in the Strong Association
gene set, especially for adipose tissue.

```{r fgsea, warning = FALSE}
loading.fgsea <- vector(mode = "list", length = length(transcript.loadings))
for(tx in 1:length(transcript.loadings)){
    #cat(names(transcript.loadings)[tx], "\n")
    stats <- sort(abs(transcript.loadings[[tx]][,1]), decreasing = TRUE)
    #plot(stats)
    loading.fgsea[[tx]] <- fgsea(gsea.ensembl.list, stats, scoreType = "pos")    
    #plotEnrichment(gsea.ensembl.list[["Strong Association"]],stats) + ggplot2::labs(title="Strong Association")
    #plotEnrichment(gsea.ensembl.list[["Weak Association"]], stats) + ggplot2::labs(title="Weak Association")
    #plotEnrichment(gsea.ensembl.list[["No Association"]], stats) + ggplot2::labs(title="No Association")
    #all.p <- loading.fgsea[[tx]]$"padj"
    #barplot_with_num(signif(-log10(loading.p), 2), names = names(gsea.ensembl.list))
}

loading.p <- sapply(loading.fgsea, function(x) x$"padj")
colnames(loading.p) <- names(transcript.loadings)
rownames(loading.p) <- names(gsea.ensembl.list)
cat.cols <- c("#66c2a5", "#fc8d62", "#8da0cb", "#e78ac3", "#a6d854", "#ffd92f", "#e5c494", "#b3b3b3")[1:length(gene.sets)]
barplot(-log10(loading.p), beside = TRUE, col = cat.cols,
    main = "-log10(Enrichment Loading p)")
legend("topright", fill = cat.cols, legend = names(gsea.ensembl.list))
```

## GSEA sorted by MGE

The following plot shows the enrichment -log10(p values)
for each gene set with the genes sorted by |MGE|. Genes 
with large MGE are not enriched for any group in any tissue.

```{r mge_enrich, warning = FALSE}
mge.fgsea <- vector(mode = "list", length = length(transcript.loadings))
for(tx in 1:length(mge)){
    if(length(mge[[tx]]) > 0){
        #cat(names(mge)[tx], "\n")
        stats <- sort(abs(mge[[tx]]), decreasing = TRUE)
        #plot(stats)
        mge.fgsea[[tx]] <- fgsea(gsea.ensembl.list, stats, scoreType = "pos")
        #plotEnrichment(gsea.ensembl.list[["Strong Association"]],stats) + ggplot2::labs(title="Strong Association")
        #plotEnrichment(gsea.ensembl.list[["Weak Association"]], stats) + ggplot2::labs(title="Weak Association")
        #plotEnrichment(gsea.ensembl.list[["No Association"]], stats) + ggplot2::labs(title="No Association")
        #mge.p <- mge.fgsea[[tx]]$"padj"
        #barplot_with_num(signif(-log10(mge.p), 2), names = names(gsea.names.list))
    }
}

mge.p <- sapply(mge.fgsea, function(x) x$"padj")
colnames(mge.p) <- names(transcript.loadings)
rownames(mge.p) <- names(gsea.names.list)
barplot(-log10(mge.p), beside = TRUE, col = cat.cols,
    main = "-log10(Enrichment MGE p)")
legend("topright", fill = cat.cols, legend = names(gsea.names.list))

```