---
title: "CC-RIX genotypes"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction

The purpose of this workflow is to generate genotypes for the 
CC-RIX animals using CC genotype data for the CC lines.
Individual level genotypes are used to impute gene expression.


```{r load_code}
rm(list = ls()) #clear workspace
is.interactive = FALSE
overwrite.files = FALSE
#is.interactive = TRUE

library("here")

data.dir <- here("Data", "CC-RIX") #where original data are kept
results.dir <- here("Results", "CC-RIX", "Genotypes") #directory for results generated by this workflow
if(!file.exists(results.dir)){dir.create(results.dir)}

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}

genotypes <- c("MRCA", "sequenced")
genotypes <- genotypes[1] #use MRCA genotypes
```

```{r set_filenames}
annot.file <- file.path(data.dir, "final_sample_annotations.csv") 

#files that we will generate
genoprobs.file <- file.path(results.dir, "Genoprobs.RDS")
kinship.file <- file.path(results.dir, "Kinship.RDS")
map.file <- file.path(results.dir, "Map.RDS")
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "qtl2", "abind")

all.paths <- .libPaths()
personal.path <- grep("atyler", all.paths)
load_libraries(all.packages, lib.loc = all.paths[personal.path])

data(CCcolors) 
#barplot(rep(1, length(CCcolors)), col = CCcolors, names = names(CCcolors))
```

```{r read_data}
annot <- read.csv(annot.file)
manifest <- read.csv(file.path(data.dir, "Original_Samples_Manifest.csv"))
```


## CC genotypes

In this workflow, we used the `r genotypes` genotypes.

I downloaded the CC genotypes from the UNC website:
http://csbio.unc.edu/CCstatus/CCGenomes/#genotypes

I downloaded the Most Recent Common Ancestor file. 
This has marker names from the mega muga. I downloaded the
mega muga marker map from the Jackson Laboratory 
[resource for CC/DO mice](https://www.jax.org/research-and-faculty/genetic-diversity-initiative/tools-data/diversity-outbred-reference-data)

```{r MRCA_genotypes, warning = FALSE}

if(genotypes == "MRCA"){
    cc.geno <- readRDS(here("Data", "CC-RIX", "MRCA10_modelprobs_trim.rds"))
    marker.info.file <- here("Data", "CC-RIX", "mm_uwisc_v1.csv")
    marker.info <- read.csv(marker.info.file)

    #convert the map and the genoprobs into a qtl2 object
    cross_obj <- array2qtl2(cc.geno, marker.info, chr = c(1:19, "X"), chr.col = "chr", 
        pos.col = "bp_mm10", crosstype = "risib8")

    cc.genoprobs <- cross_obj$genoprobs

    map <- cross_obj$map
    saveRDS(map, map.file)

    K <- calc_kinship(cc.genoprobs, "loco")
    saveRDS(K, kinship.file)
}
```

Or read genotype data from \url{https://github.com/rqtl/qtl2data/tree/main/CC}
if specified.

```{r sequenced_genotypes}

if(genotypes == "sequenced"){
    if(!file.exists(map.file)){
        file <- paste0("https://raw.githubusercontent.com/rqtl/", "qtl2data/master/CC/cc.zip")
        cc <- read_cross2(file)
        map <- map
        saveRDS(map, map.file)
        
        cc.official <- rownames(cc$geno[[1]])
        cc.alias <- sapply(strsplit(cc.official, "/"), function(x) x[1])
        write.table(cbind(cc.official, cc.alias), here("Data", "CC_names.txt"), sep = "\t", 
            quote = FALSE, row.names = FALSE, col.names = FALSE)
        
        cc.genoprobs <- calc_genoprob(cc, cores = cores)    
        new.genoprobs <- cc.genoprobs

        #add strains 83 and 84 to the genoprobs 
        new.strains <- c("CC083", "CC084")
        new.strain.idx <- match(new.strains, rownames(cc.geno))
        for(i in 1:length(cc.genoprobs)){
            chr.markers <- dimnames(cc.genoprobs[[i]])[[3]]
            common.markers <- intersect(chr.markers, dimnames(cc.geno)[[3]])
            orig.idx <- match(common.markers, dimnames(cc.genoprobs[[i]])[[3]])
            new.idx <- match(common.markers, dimnames(cc.geno)[[3]])
            new.chr <- abind(cc.genoprobs[[i]][,,orig.idx], cc.geno[new.strain.idx,,new.idx], along = 1)
            new.genoprobs[[i]] <- new.chr
        }

        #overwrite the original objecct
        cc.genoprobs <- new.genoprobs

        K <- calc_kinship(new.genoprobs, "loco")
        saveRDS(K, kinship.file)
    }else{
        map <- readRDS(map.file)
        K <- readRDS(kinkship.file)
    }
}
```


```{r read_genotype_data, eval = FALSE}
#original code from this workflow. switching to MRCA genotypes
all.var <- ls()
cc.data.loaded <- as.logical(length(which(all.var == "cc")))

if(!cc.data.loaded){
    file <- paste0("https://raw.githubusercontent.com/rqtl/",
        "qtl2data/master/CC/cc.zip")
    cc <- read_cross2(file)
}

#calculate genoprobs using qtl2
map <- cc$pmap
saveRDS(map, map.file)
cc.genoprobs <- calc_genoprob(cc, cores = 4)

```

```{r, eval = FALSE}
#test known marker 
#the marker at chr 4, 107159244 bp
#should have E genotypes for strains
#CC051, CC059, CC012, and CC002

marker.chr = 4
marker.pos = 107159244
test.strains <- c("CC051", "CC059", "CC012", "CC002")
nearest.geno <- pull_genoprobpos(cc.genoprobs, map = map, chr = marker.chr, pos = marker.pos)
strain.idx <- sapply(test.strains, function(x) grep(x, rownames(nearest.geno)))
round(nearest.geno[strain.idx,])

```

```{r get_geno_fun}
get_rix_geno <- function(rix.strain, cc.genoprobs, sex){

    split.strain <- strsplit(rix.strain, "_")
    strain1 <- split.strain[[1]][1]
    strain2 <- split.strain[[1]][2]
    strain1.locale <- grep(strain1, rownames(cc.genoprobs[[1]]))
    strain2.locale <- grep(strain2, rownames(cc.genoprobs[[1]]))
    #take the average of the genoprobs for the RIX
    #test <- cc.genoprobs[[1]][c(strain1.locale, strain2.locale),,]
    #i = 5000
    #pheatmap(test[,,i], cluster_rows = FALSE, cluster_cols = FALSE)
    rix.genoprobs <- lapply(cc.genoprobs, function(x) colMeans(x[c(strain1.locale, strain2.locale),,]))
    #barplot(rix.genoprobs[[1]][,i])
    #pheatmap(rix.genoprobs[[4]], cluster_rows = FALSE, cluster_cols = FALSE, show_colnames = FALSE)

    #recalulate the X chromosome if the individual is male
    #in this case the X will be exclusively from the mother.
    #assume strain1 is the mother
    if(sex == "Male"){
        x.chr <- cc.genoprobs[[20]][strain1.locale,,]
        rix.genoprobs[[20]] <- x.chr
    }
    return(rix.genoprobs)
}

if(is.interactive){
    lapply_fun <- match.fun("lapply_pb")
}else{
    lapply_fun <- match.fun("lapply")
}
```

There are many repeated genotypes in the CC-RIX.
We only need to calculate genoprobs for unique combinations of genomes.
We also need to calculate males and females separately.
We will label the individual rows with the strain and sex label so we can
grab the correct names later for scanning.

```{r ind_genoprobs}
unique_combos <- unique(annot[,c("Strain", "Sex")])
all.rix.genoprobs <- lapply_fun(1:nrow(unique_combos), function(x) 
    get_rix_geno(unique_combos[x,"Strain"], cc.genoprobs, unique_combos[x,"Sex"]))
```


```{r concatenate_rix_genoprobs}
geno.id <- apply(unique_combos, 1, function(x) paste(x, collapse = "_"))
cc.rix.genoprobs <- cc.genoprobs
for(ch in 1:length(cc.genoprobs)){
    if(is.interactive){
        report.progress(ch, length(cc.genoprobs))
    }
    all.chr <- abind(lapply(all.rix.genoprobs, function(x) x[[ch]]), along = 3)
    chr.array <- aperm(all.chr, c(3,1,2))
    rownames(chr.array) <- geno.id
    cc.rix.genoprobs[[ch]] <- chr.array
}
attr(cc.rix.genoprobs, "alleles") <- colnames(cc.rix.genoprobs[[1]])

saveRDS(cc.rix.genoprobs, genoprobs.file)

```

## Allele Frequencies {.tabset .tabset-fade .tabset-pills}

Look at the allele frequencies by chromosome for this validation cohort.

```{r allele_freq, results = "asis", fig.width = 10, fig.height = 4.5}

for(ch in 1:length(cc.rix.genoprobs)){
    cat("### Chr", names(cc.rix.genoprobs)[ch], "\n")
    chr.geno <- cc.rix.genoprobs[[ch]]
    freq.mat <- apply(chr.geno, 2, function(x) colMeans(x))
    #quartz(width = 10, height = 4.5)
    layout(matrix(c(1,2), ncol = 2), widths = c(1, 0.2))
    ymax <- max(c(0.5, max(freq.mat)))
    
    common.markers <- intersect(rownames(freq.mat), names(map[[ch]]))

    par(mar = c(4,4,4,0))
    plot.new()
    plot.window(xlim = c(1,max(map[[ch]])), ylim = c(0, ymax))
    for(a in 1:ncol(chr.geno)){
        points(x = map[[ch]][common.markers], y = freq.mat[common.markers,a], col = CCcolors[a], type = "l", lwd = 3)
    }
    axis(1);axis(2)
    mtext("Position (Mb)", side = 1, line = 2.5)
    mtext("Allele Frequency", side = 2, line = 2.5)
    mtext(paste0("Chr", names(cc.rix.genoprobs)[ch]), side = 3)
    
    par(mar = c(0,0,0,0))
    plot.new()
    plot.window(xlim = c(0,1), ylim = c(0,1))
    legend(x = 0, y = 0.8, legend = names(CCcolors), lty = 1, col = CCcolors, lwd = 3)
    cat("\n\n")
}

```

## Kinship

Calculate kinship.

```{r kinship}
if(!file.exists(kinship.file) || overwrite.files){
    K <- calc_kinship(cc.rix.genoprobs, type = "loco", cores = 4)
    saveRDS(K, kinship.file)
}
```

The kinship matrix has the following structure.

```{r kin_heat, fig.width = 7, fig.height = 7}
pheatmap(K[[1]])
```
