---
title: "Impute CC-RIX Gene Expression"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction
The purpose of this workflow is use local eQTL identified
in the DO to impute gene expression in the CC-RIX. We
compare imputed values to measured values and check to make
sure that transcripts with strong eQTL in the DO are imputed
accurately in the CC-RIX.


```{r load_code}
rm(list = ls()) #clear out workspace
is.interactive = FALSE
#is.interactive = TRUE
library("here")

data.dir <- here("Data", "CC-RIX") #directory holding original data
data.results.dir <- here("Results", "CC-RIX", "Data") #directory holding parsed data for downstream workflows
results.dir <- here("Results", "CC-RIX", "Imputed") #directory holding results from this workflow
if(!file.exists(results.dir)){dir.create(results.dir)}

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}

test_plot = FALSE
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "stringr", "qtl2", "vioplot")
load_libraries(all.packages)
```

Read in the following data:

1. local eQTL coefficients from the DO generated by 1.DO/3a.Imputation.Rmd
2. genotypes of CC-RIX individuals generated by 2.CC-RIX/1b.CC-RIX_Genotypes.Rmd
3. map for CC-RIX markers generated by 2.CC-RIX/1b.CC-RIX_Genotypes.Rmd
4. measured gene expression in CC-RIX generated by 2.CC-RIX/1a.CC-RIX_Phenotypes_Expression.Rmd
  for comparison to imputed values
5. gene information and nearest marker information generated in 1.DO/1a.Tissue_Expression.Rmd

```{r read_data}
#generated by 1.DO/3a.Imputation.Rmd
local.eqtl.files <- list.files(here("Results", "DO", "Imputed_Transcriptomes"), 
  pattern = "eQTL", full.names = TRUE)
tissue.names <- sapply(strsplit(gsub(".RDS", "", basename(local.eqtl.files)), "_"), function(x) tail(x, 1))
local.eqtl <- lapply(local.eqtl.files, readRDS)
names(local.eqtl) <- tissue.names

#generated in 2.CC-RIX/1b.CC-RIX_Genotypes.Rmd
#F1 genotypes for all individuals used in this 
#experiment
cc.geno <- readRDS(here("Results", "CC-RIX", "Genotypes", "Genoprobs.RDS")) 

#get just the strain labels, ignoring male/female
geno.strain <- sapply(strsplit(rownames(cc.geno[[1]]), "_"), function(x) paste(x[1], x[2], sep = "_"))
u_strain <- unique(geno.strain)
u_strain_idx <- match(u_strain, geno.strain)

#generated in 2.CC-RIX/1b.CC-RIX_Genotypes.Rmd
cc.map <- readRDS(here("Results", "CC-RIX", "Genotypes", "Map.RDS")) 

#there are a few markers different in the map and genotype object. Line them up
for(i in 1:length(cc.geno)){
  common.markers <- intersect(dimnames(cc.geno[[i]])[[3]], names(cc.map[[i]]))
  cc.geno[[i]] <- cc.geno[[i]][,,common.markers]
  cc.map[[i]] <- cc.map[[i]][common.markers]
}

#generated by 2.CC-RIX/1a.CC-RIX_Phenotypes_Expression.Rmd
#adjusted for sex, diet, treatment, and age. The only
#effect left is strain. Should we adjust for strain too,
#so we can look only at the local allele effects, rather
#than kinship effects?
cc.expr <- readRDS(here("Results", "CC-RIX", "Data", "Expression.by.Tissue.Adjusted.RDS"))
cc.tissues <- names(cc.expr)

tissue.cols <- as.matrix(read.delim(here("Data", "general", "tissue_colors.txt"), header = FALSE, row.names = 1))[,1]
tissue.idx <- sapply(cc.tissues, function(x) grep(x, names(tissue.cols), ignore.case = TRUE))
tissue.cols <- tissue.cols[tissue.idx]

#gene marker information 
marker.info <- readRDS(here("Results", "DO", "Data", "Gene_Tables.RDS"))

#sample annotations
manifest <- read.csv(here("Data", "CC-RIX", "Original_Samples_Manifest.csv"))

```

We have multiple animals of each genotype with measured gene expression.
To compare to the imputed, we use the average expression for each genotype.

We also want to compare the within-strain to between-strain variation
for the transcripts. How is this related to eQTL LOD score in the DO?


```{r measured_avg}

cc.avg.expr.file <- file.path(results.dir, "Strain_Avg_Expression.RDS")

cc.avg.expr <- vector(mode = "list", length = length(cc.tissues))
names(cc.avg.expr) <- cc.tissues

for(tx in 1:length(cc.expr)){
  tx.expr <- cc.expr[[tx]]
  ind.expr <- rownames(tx.expr)

  ind_info_table <- t(sapply(ind.expr, function(x) get_mouse_info(x, manifest)))
  ind.strain <-ind_info_table[,"User.Defined.Strain"]
  expr.strain <- gsub(" x ", "_", gsub(" X ", "_", ind.strain))

  #find individual indices for each unique strain
  strain.idx <- lapply(u_strain, function(x) which(expr.strain == x))
  #par(mar = c(8, 4, 4, 4)); boxplot(strain.idx, names = u_strain, las = 2)
  
  #rnd.tx <- sample(1:ncol(tx.expr), 1)
  #rnd.tx <- which(colnames(tx.expr) == "ENSMUSG00000051748") #Wfdc21
  #rnd.expr <- lapply(strain.idx, function(x) tx.expr[x,rnd.tx])
  #strain.order <- order(sapply(rnd.expr, mean))
  #par(mar = c(8, 4, 4, 4)); boxplot(rnd.expr[strain.order], names = geno.strain[strain.order], las = 2)

  #instrain.var <- sapply(rnd.expr, var)
  #allstrain.var <- var(sapply(rnd.expr, mean))
  #xlim <- c(min(c(instrain.var, allstrain.var))*0.95, max(c(instrain.var, allstrain.var))*1.05)
  #hist(instrain.var, xlim = xlim)
  #abline(v = allstrain.var, col = "red")

  avg.expr <- t(sapply(strain.idx, function(x) colMeans(tx.expr[x,])))
  rownames(avg.expr) <- u_strain
  cc.avg.expr[[tx]] <- avg.expr
}

saveRDS(cc.avg.expr, cc.avg.expr.file)

```

The imputation process is as follows:

For each transcript in each tissue, 

  1. identify the nearest marker
  2. find the genotype at that marker for each CC F1
  3. multiply the local eQTL coefficients by the F1 genotypes 
    to get the imputed expression for each unique genotype


```{r impute_expr}
save.every <- 100

for(tx in 1:length(cc.tissues)){ #only impute for tissues with measured expression
  imputed.expr.file <- file.path(results.dir, paste0("Imputed_Expr_", cc.tissues[tx], ".RDS"))
  
  if(file.exists(imputed.expr.file)){
    imputed.expr <- readRDS(imputed.expr.file)
    start.idx <- max(which(apply(imputed.expr, 2, function(x) !all(is.na(x))))) + 1
  }else{
    imputed.expr <- matrix(NA, nrow = length(u_strain), ncol = ncol(cc.expr[[tx]]))
    rownames(imputed.expr) <- u_strain
    colnames(imputed.expr) <- colnames(cc.expr[[tx]])
    start.idx <- 1
  }

  if(start.idx < ncol(cc.expr[[tx]])){
      sink(file.path(results.dir, paste0("progress_", cc.tissues[tx], ".txt")))

    for(tr in start.idx:ncol(cc.expr[[tx]])){
      print(tr)
      #tr <- which(colnames(cc.expr[[tx]]) == "ENSMUSG00000051748") #Wfdc21
      #tr <- which(colnames(cc.expr[[tx]]) == "ENSMUSG00000028619") #very high lod from before
      tr.id <- colnames(cc.expr[[tx]])[tr]
      
      info.idx <- grep(cc.tissues[tx], names(marker.info), ignore.case = TRUE)
      tr.info.idx <- which(marker.info[[info.idx]]$gene.id == tr.id)
    
      #find the genotypes at the nearest marker
      if(length(tr.info.idx) == 0){next()}
      nearest.marker <- marker.info[[info.idx]]$nearest.marker.id[tr.info.idx]

      if(nearest.marker == ""){next()}
      split.marker <- strsplit(nearest.marker, "_")[[1]]
      marker.chr <- split.marker[1]
      marker.pos <- as.numeric(split.marker[2])
      nearest.geno <- pull_genoprobpos(cc.geno, map = cc.map, chr = marker.chr, pos = marker.pos)[u_strain_idx,]
      rownames(nearest.geno) <- u_strain

      #find the coefficients from the DO
      do.tx.idx <- grep(cc.tissues[tx], names(local.eqtl), ignore.case = TRUE)
      do.tr.idx <- which(rownames(local.eqtl[[do.tx.idx]]) == tr.id) 
      tr.coef <- local.eqtl[[do.tx.idx]][do.tr.idx,]
      #barplot(tr.coef[1:8])

      imp.expr <- colMeans(apply(nearest.geno, 1, function(x) x*tr.coef[LETTERS[1:8]]))
      imputed.expr[,tr] <- imp.expr
    
      if(test_plot){
        png("~/Desktop/test_imputation.png", width = 9, height = 10, units = "in", res = 200)
        meas.expr <- lapply(strain.idx, function(x) cc.expr[[tx]][x,tr.id])
        strain.order <- order(sapply(meas.expr, median))
        layout(matrix(c(1,5, 2, 3, 4, 0), ncol = 2, byrow = TRUE))
        par(mar = c(8, 4, 4, 0))
        boxplot(meas.expr[strain.order], names = u_strain[strain.order], las = 2, 
          main = "Measured Expression")
        par(mar = c(0, 4, 8, 0))
        imageWithText(t(nearest.geno[strain.order,]), show.text = FALSE, 
          use.pheatmap.colors = TRUE, row.text.shift = 0.05, col.text.shift = -1.8)
        par(mar = c(0,0,8,4))
        ordered.tr.coef  <- tr.coef[8:1]
        barplot(ordered.tr.coef, horiz = TRUE, las = 2, xlab = "Haplotype Coefficient", 
          names = names(ordered.tr.coef))
        par(mar = c(8, 4, 0, 0))
        barplot(imp.expr[strain.order], las = 2)
        par(mar = c(4, 4, 4, 4))
        plot.with.model(sapply(meas.expr[strain.order], mean), imp.expr[strain.order], 
          xlab = "Measured Expr", ylab = "Imputed Expr", report = "cor.test")
        dev.off()
      }

      #imp.df <- data.frame("Imputed_Expression" = imp.expr, "Measured_Expression" = cc.avg.expr[[tx]][,tr.id])
      #imp.order <- order(imp.df[,1])
      #pheatmap(nearest.geno[imp.order,], annotation_row = imp.df, cluster_rows = FALSE, cluster_cols = FALSE)
      #plot.with.model(imp.expr, cc.avg.expr[[tx]][,tr.id], report = "cor.test")

      if(tr %% save.every == 0){
        saveRDS(imputed.expr, imputed.expr.file)
      }
    }
    saveRDS(imputed.expr, imputed.expr.file)
    sink()
  
  }
}

tx.imp.expr <- lapply(cc.tissues, function(x) readRDS(file.path(results.dir, paste0("Imputed_Expr_", x, ".RDS"))))
names(tx.imp.expr) <- cc.tissues
```

## Imputation Check {.tabset .tabset-fade .tabset-pills}

Compare the measured gene expression to the imputed gene expression.
The imputed transcript values are positively correlated with the
measured values overall. Also, the higher the LOD score of the 
transcript eQTL in the DO, the better we are at imputing it in the
CC-RIX.

There are a substantial number of transcripts for which the measured
and imputed expression are negatively correlated. This was not the
case in the DO (3a.Imputation.Rmd).

```{r compare, fig.width = 8, fig.height = 4, results = "asis", warning = FALSE}

max.lod <- max(sapply(local.eqtl, function(x) max(x[,"lod"], na.rm = TRUE)))
tx.measured.imp.cor <- tx.measured.imp.var <- vector(mode = "list", length = length(cc.tissues))
names(tx.measured.imp.cor) <- names(tx.measured.imp.var) <- cc.tissues

#pdf("~/Desktop/test_var_exp.pdf", height = 8, width = 8)
par(mfrow = c(1,2))
for(tx in 1:length(cc.tissues)){
  cat("###", cc.tissues[tx], "\n")
  eqtl.idx <- grep(cc.tissues[tx], names(local.eqtl), ignore.case = TRUE)
  imp.idx <- which(names(tx.imp.expr) == cc.tissues[tx])

  common.tx <- Reduce("intersect", list(colnames(cc.avg.expr[[tx]]), 
    colnames(tx.imp.expr[[imp.idx]]), rownames(local.eqtl[[eqtl.idx]])))
  
  has.var.expr <- which(apply(cc.avg.expr[[tx]][,common.tx], 2, var) > 0)
  has.var.imp <- which(apply(tx.imp.expr[[tx]][,common.tx], 2, var) > 0)
  common.with.var <- intersect(names(has.var.expr), names(has.var.imp))
  measured.imp.cor <- sapply(common.with.var, function(x) cor(cc.avg.expr[[tx]][,x], tx.imp.expr[[tx]][,x]))
  measured.imp.var <- sapply(common.with.var, function(x) summary(lm(tx.imp.expr[[tx]][,x]~cc.avg.expr[[tx]][,x]))$adj.r.squared)
  tx.measured.imp.cor[[tx]] <- measured.imp.cor
  tx.measured.imp.var[[tx]] <- measured.imp.var

  a <- hist(measured.imp.cor, xlab = "Imputed-Measured Correlation", col = tissue.cols[tx], main = "")
  #abline(v = mean(measured.imp.cor, na.rm = TRUE))  
  mean.imp.cor <- mean(measured.imp.cor, na.rm = TRUE)
  arrows(x0 = mean.imp.cor, y0 = max(a$counts)*0.1, y1 = 0, col = "black", lwd = 3, length = 0.1)

  plot.new()
  plot.window(xlim = c(6, max.lod), ylim = c(-0.9, 1))
  axis(1);axis(2)
  mtext("eQTL LOD", side = 1, line = 2.5)
  mtext("Measured-Imputed Correlation", side = 2, line = 2.5)
  #if(tx == 1){
    points(local.eqtl[[eqtl.idx]][common.with.var,"lod"], measured.imp.cor,
      xlim = c(6, max.lod), col = "gray10") 
  #}
  points(local.eqtl[[eqtl.idx]][common.with.var,"lod"], measured.imp.cor, 
    pch = 16, col = tissue.cols[tx])
  abline(h = 0)

  a <- hist(measured.imp.var, col = tissue.cols[tx], xlab = "Variance Explained", main = "")
  mean.imp.var <- mean(measured.imp.var, na.rm = TRUE)
  arrows(x0 = mean.imp.var, y0 = max(a$counts)*0.1, y1 = 0, col = "black", lwd = 3, length = 0.1)


  plot.new()
  plot.window(xlim = c(6, max.lod), ylim = c(-0.1, 1))
  axis(1);axis(2)
  mtext("eQTL LOD", side = 1, line = 2.5)
  mtext("Variance Explained", side = 2, line = 2.5)

  #if(tx == 1){
    points(local.eqtl[[eqtl.idx]][common.with.var,"lod"], measured.imp.var,
      xlim = c(6, max.lod), col = "gray10") 
  #}
  points(local.eqtl[[eqtl.idx]][common.with.var,"lod"], measured.imp.var, 
    pch = 16, col = tissue.cols[tx])

  cat("\n\n")
}
#dev.off()
```

The boxplots below compare the imputed-measured correlation across
tissues. 

```{r comp_cor, fig.width = 5, fig.height = 5}
boxplot(tx.measured.imp.cor, ylab = "Correlation", col = tissue.cols)
abline(h = 0)
```

## Comparison of imputation across tissues

The boxplots below show the comparison of the variance explained
of the measured gene expression by the local genotype as compared
to what we saw in the DO. 

```{r do_cc_comp, fig.width = 7, fig.height = 5}
local.distal.comp <- readRDS(here("Results", "DO", "Imputed_Transcriptomes", 
  "Local_Distal_Comparison.RDS"))

just.local <- local.distal.comp$Local
common.tx <- intersect(rownames(just.local), Reduce("intersect", lapply(tx.measured.imp.var, names)))
cc.var <- sapply(tx.measured.imp.var, function(x) x[common.tx])

comp.list <- list(just.local[,c("Adipose", "Liver", "SkeletalMuscle")], cc.var)
names(comp.list) <- c("DO", "CC-RIX")
plot.grouped.boxes(group.list = comp.list, 
  type = "matrix", group.cols = c("#af8dc3", "#7fbf7b"),
  ylab = "Variance Explained", plot.type = "vioplot",
  within.group.sep = 1, between.group.sep = 1.5)
plot.dim <- par("usr")
segments(x0 = plot.dim[1], x1 = plot.dim[2], y0 = 0)
```

## Imputation and Trait Relevance {.tabset .tabset-fade .tabset-pills}

We expect that transcripts that are easy to impute accurately, 
those with strong eQTL are less trait-relevant than transcripts
that are more difficult to impute. Here we look at ability to 
impute accurately (correlation between measured and imputed),
and trait-relevance (loading or max trait correlation).

```{r trait_rel, fig.width = 5, fig.height = 5, results = "asis"}

transcript.loadings <- readRDS(here("Results", "DO", "High_Dim_Med_no_MGE", 
  "tissue_together-_-complete_mediation-germline_kinship", "Loadings_Transcripts_0.RDS"))

max.trait.cor.files <- list.files(here("Results", "DO", "Transcriptomes"), 
  pattern = "Maximum_Trait_Correlation", full.names = TRUE)
max.trait.cor <- lapply(max.trait.cor.files, readRDS)

for(tx in 1:length(cc.tissues)){
  cat("###", cc.tissues[tx], "\n")
  tx.idx <- grep(cc.tissues[tx], names(transcript.loadings), ignore.case = TRUE)
  common.tx <- intersect(names(tx.measured.imp.cor[[tx]]), rownames(transcript.loadings[[tx.idx]]))

    plot_with_marginal_hist(tx.measured.imp.cor[[tx]][common.tx], abs(transcript.loadings[[tx.idx]][common.tx,]),
      xlab = "Measured-Imputed Correlation", ylab = "Transcript Loading", report = "cor.test",
      main = "Transcript Loading")
  
  common.tx <- intersect(names(tx.measured.imp.cor[[tx]]), names(max.trait.cor[[tx.idx]]))
  plot_with_marginal_hist(tx.measured.imp.cor[[tx]][common.tx], abs(max.trait.cor[[tx.idx]][common.tx]),
    xlab = "Measured-Imputed Correlation", ylab = "Maximum Trait Correlation", report = "cor.test",
    main = "Maximum Trait Correlation")
  
  cat("\n\n")
}

```

## Imputation and Correlation Structure {.tabset .tabset-fade .tabset-pills}

How well does imputation recapitulate the correlation structure
in the transcriptome? Transcripts tend to be highly correlated
in real life, and we know that this correlation structure is 
important. However, alleles in the genome are inherited randomly
and independently. Thus, we expect that the imputed transcriptome
completely mis-specifies the transcript correlation structure. 

For now we are just sampling the correlation matrix instead of
calculating it exhaustively. If this is a useful analysis, we'll
do it properly.

Much to my surprise, it looks as if the correlations are actually
related to each other. Transcripts that are negatively/positively 
correlated in the measured transcriptomes tend to be 
negatively/positively correlated in the imputed transcriptomes. 

The blue lines below show the line of best fit. The red lines show
y = x.

```{r transcript_correlations, fig.width = 5, fig.height = 5, warning = FALSE, results = "asis"}
sample.size = 100
n.samples = 100

measured.cor <- imputed.cor <- vector(mode = "list", length = length(cc.tissues))
names(measured.cor) <- names(imputed.cor) <- cc.tissues
for(tx in 1:length(cc.tissues)){
  common.tx <- intersect(colnames(cc.avg.expr[[tx]]), colnames(tx.imp.expr[[tx]]))
  measured.cor.v <- imp.cor.v <- NULL
  for(sp in 1:n.samples){
    tx.sample <- sample(common.tx, sample.size)
    sampled.measured.cor <- cor(cc.avg.expr[[tx]][,tx.sample], use = "pairwise.complete.obs")
    sampled.imp.cor <- cor(tx.imp.expr[[tx]][,tx.sample], use = "pairwise.complete.obs")
    measured.cor.v <- c(measured.cor.v, sampled.measured.cor[upper.tri(sampled.measured.cor, diag = FALSE)])
    imp.cor.v <- c(imp.cor.v, sampled.imp.cor[upper.tri(sampled.imp.cor, diag = FALSE)])
  }
  measured.cor[[tx]] <- measured.cor.v
  imputed.cor[[tx]] <- imp.cor.v
}

for(tx in 1:length(cc.tissues)){
  cat("###", cc.tissues[tx], "\n")
  plot_with_marginal_hist(measured.cor[[tx]], imputed.cor[[tx]], main = cc.tissues[tx],
    xlab = "Measured Correlation", ylab = "Imputed Correlation", report = "cor.test")
    cat("\n\n")
}
```

### DO-CC Kinship Comparison

The high correlation between measured and imputed in the CC-RIX 
may be related to LD blocks and relatively high kinship in the 
CC-RIX. The imputed transcripts are mostly uncorrelated in the 
DO, which are much less related to each other.

The following histograms show the distribution of kinship coefficients
among the DO and the CC-RIX.


```{r kin_comparison, fig.width = 8, fig.height = 8}
cc.kin <- calc_kinship(cc.geno)
do.kin <- readRDS(here("Results", "DO", "Data", "overall.kinship.RDS"))
cc.kin.v <- cc.kin[upper.tri(cc.kin,diag = FALSE)]
do.kin.v <- do.kin[upper.tri(do.kin, diag = FALSE)]
xmax <- max(c(do.kin.v, cc.kin.v))

par(mfrow = c(2,2))
hist(cc.kin.v, xlab = "Kinship Coefficient", xlim = c(0, xmax), main = "CC Kinship")
hist(do.kin.v, xlab = "Kinship Coefficient", main = "DO Kinship", xlim = c(0, xmax), breaks = 100)
stripchart(list("CC_kin" = cc.kin.v, "DO_kin" = do.kin.v), main = "Kinship Comparison",
  vertical = TRUE, pch = 16, method = "jitter")
```