---
title: "Kinship, Expression, and Traits"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
bibliography: hdm.bib 
---

The purpose of this workflow is to examine the relationship
between kinshp, gene expression, and traits. It uses output
from 1a.Tissue_Expression.Rmd and 1b.Trait_Selection.Rmd.

```{r here, message = FALSE, warning = FALSE}
rm(list = ls())
library(here)
results.dir <- here("Results", "Transcriptomes")
if(!file.exists(results.dir)){dir.create(results.dir)}
parsed.data.dir <- here("Results", "DO", "Data") #a results folder that holds parsed data that we don't want to mix in with downloaded data
is.interactive = FALSE #set to true to show verbose reporting of progress on some longer processes
```

```{r load_code, message = FALSE, warning = FALSE}
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}

all.packages <- c("pheatmap")
load_libraries(all.packages, personal.library = TRUE)
```

```{r read_data}

expr <- readRDS(file.path(parsed.data.dir, "Tissue_Expression_Adjusted.RDS")) #generated in 1a.Tissue_Expression.Rmd
tissue.names <- names(expr)

common.ind <- Reduce("intersect", lapply(expr, rownames))
ind.matched.expr <- lapply(expr, function(x) x[common.ind,])
local.lod <- readRDS(here("Results", "DO", "Transcriptomes", "Local_LOD.RDS")) #generated in 1a.Tissue_Expression.Rmd
gene.tables <- readRDS(file.path(parsed.data.dir, "Gene_Tables.RDS")) #generated in 1a.Tissue_Expression.Rmd

pheno <- readRDS(file.path(parsed.data.dir, "Clinical_Phenotypes_Adjusted.RDS")) #generated in 1b.Trait_Selection.Rmd
pheno.names <- colnames(pheno)
pheno.class <- c("1.phys", "1.phys", "4.glucose", "4.glucose", "5.insulin", "5.insulin",
    "6.lipid", "6.lipid", "6.lipid", "6.lipid", "2.islet", "2.islet", "2.islet", "3.HOMA", 
    "3.HOMA", "6.lipid", "6.lipid", "6.lipid", "6.lipid", "7.chem", "7.chem", "1.phys")

load(here("Data", "DO", "QTLViewer_Geno_V10.Rdata")) #ensembl.version, genoprobs, K, map, markers

```

## Trait Matrix Decomposition

The trait matrix clusters into several clusters of related traits.

```{r trait_clust, fig.width = 6, fig.height = 6}
trait.cor <- cor(pheno, use = "pairwise.complete.obs")
pheatmap(trait.cor)
```

Decomposition of the trait correlation matrix shows
how traits cluster in PC space.

```{r trait_cor, fig.width = 6, fig.height = 6}
plot.decomp(trait.cor, label.points = TRUE, xlim = c(-0.4, 0.55))
```

```{r trait_decomp, warning = FALSE}
trait.decomp <- plot.decomp(pheno, plot.results = FALSE, scale = TRUE, pc = ncol(pheno))
```

The loadings on the traits are shown below.

```{r trait_loadings, fig.width = 7, fig.height = 7}
#png("~/Desktop/test.png", width = 7, height = 7, units = "in", res = 300)
var.exp <- round(trait.decomp$var.exp*100)

layout(matrix(c(1,0,2,3), ncol = 2, byrow = TRUE), heights = c(0.5, 1), widths = c(1, 0.1 ))
par(mar = c(0,5.7,2,3.7))
barplot_with_num(var.exp, ylab = "% Var. Exp.")

trait.loadings <- trait.decomp$v
rownames(trait.loadings) <- colnames(pheno)
colnames(trait.loadings) <- paste0("PC", 1:ncol(trait.loadings))

row.order <- sort.by.then.by(cbind(pheno.class, trait.loadings[,1]), col.type = c("c", "n"))

par(mar = c(4,6,0,4))
imageWithText(trait.loadings[rownames(row.order),], use.pheatmap.colors = TRUE, show.text = FALSE,
    row.text.shift = 0.05, col.text.shift = 0.05)
par(mar = c(4,0,4,2))
imageWithTextColorbar(trait.loadings, use.pheatmap.colors = TRUE, cex = 1, bar.lwd = 3)
#dev.off()
```

## Transcripts, Traits, and eQTL

The figures below show correlations between eQTL LOD
scores and trait correlations. Trait correlations are
the maximum absolute value correlation for a transcript
across all traits.

```{r max_lod}
#calculate the maximum LOD score for each transcript
tissue.data <- lapply(tissue.names, 
    function(x) readRDS(here("Data", "DO", paste0(x, ".RDS")))) #actual data downloaded from Gary

get_max_lod <- function(eqtl.data){
    all.tx <- unique(eqtl.data$gene.id)
    max.lod <- sapply(all.tx, function(x) max(eqtl.data[which(eqtl.data$gene.id == x),"lod"]))
    return(max.lod)
}

max.lod.file <- file.path(results.dir, "Maximum_LOD.RDS")
if(!file.exists(max.lod.file)){
    tx.max.lod <- lapply(tissue.data, function(x) get_max_lod(x$lod.peaks$additive))
    saveRDS(tx.max.lod, max.lod.file)
}else{
    tx.max.lod <- readRDS(max.lod.file)
}
```

```{r trait_trans_cor, fig.width = 8, fig.height = 8}

#calculate correlations for each transcript
#in each tissue with all traits in the trait matrix.
#keep the maximum correlation.
par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    tx.expr <- expr[[tx]]
    common.ind <- intersect(rownames(tx.expr), rownames(pheno))

    max.trait.cor.file <- file.path(results.dir, paste0("Maximum_Trait_Correlation_", tissue.names[tx], ".RDS"))
    if(!file.exists(max.trait.cor.file)){
        max.trait.cor <- rep(NA, ncol(tx.expr))
        names(max.trait.cor) <- colnames(tx.expr)
        for(trans in 1:ncol(tx.expr)){
            cat("\n")
            if(is.interactive){report.progress(trans, ncol(tx.expr))}
            max.trait.cor[trans] <- max(abs(apply(pheno[common.ind,], 2, 
                function(x) cor(x, tx.expr[,trans], use = "pairwise.complete.obs", 
                method = "spearman"))), na.rm = TRUE)
        }
    saveRDS(max.trait.cor, max.trait.cor.file)
    }else{
    max.trait.cor <- readRDS(max.trait.cor.file)
    }

    common.tx <- intersect(colnames(tx.expr), names(tx.max.lod[[tx]]))
    plot.with.model(tx.max.lod[[tx]][common.tx], abs(max.trait.cor[common.tx]),
        main = tissue.names[tx], report = "cor.test", xlab = "Local LOD",
        ylab = "Maximum Trait Correlation")
}
```

