---
title: "local imputations"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
bibliography: hdm.bib 
---

```{r param}
rm(list = ls()) #clear out previous R environment
remove.pseudogenes = TRUE #if TRUE, removes pseudogenes and predicted genes 
tissue.names <- c("Adipose", "Islet", "Liver", "SkeletalMuscle")
tissue.cols <- c("orange", "#8dd3c7", "tan", "brown") #colors that are not related to the DO/CC colors
is.interactive = FALSE
overwrite.results = FALSE #set to TRUE to overwrite previous results
```

## Introduction
In this workflow, we impute gene expression in the DO
using the effects of the local marker. We fit the measured
transcript abundance to the nearest marker with a kinship 
correction. We use the fitted values from that model as 
the imputed transcript values.

We also estimate measures of heritability, both overall and
distal. We calculate overall heritability using est_herit
on measured gene expression. We calculate distal heritability 
by using est_herit on the residuals after fitting the linear 
model with the nearest marker. The product of the variance
of the residuals and their estimated heritability is the 
distal heritability component for the transcript.

We then compare these estimates of heritability to a measure 
of trait relevance, which is the maximum correlation of each
transcript with the traits.

We see that local heritability is anti-correlated with 
trait relevance. That is, transcripts whose abundance
is highly related to genotype at the local marker tend 
to have low correlations with traits. On the other hand,
transcripts with high trait relevance, tend to have low
local heritability.

We see a different relationship between trait relevance
and the distal component of heritability. Transcripts
with high trait relevance tend to have low-mid to mid-level 
distal heritability.

After performing the imputation, we can do a TWAS-type study
within the DO. In TWAS we calculate the correlation between
the locally-inherited component of transcription and the
trait. This should prioritize genes/transcripts for which
there is high local heritability and high trait relevance.
We want to be able to compare TWAS results with HDM. 

This workflow depends on results from 
1a.Tissue_Expression.Rmd, 1b.Trait_Selection.Rmd, and 
2a.Kinship_Expression_Traits.Rmd.

```{r here, message = FALSE, warning = FALSE}
library(here)
imputed.results.dir <- here("Results", "DO", "Imputed_Transcriptomes") #result generated by this workflow
if(!file.exists(imputed.results.dir)){dir.create(imputed.results.dir, recursive = TRUE)}
data.results.dir <- here("Results", "DO", "Data") #a results folder that holds parsed data that we don't want to mix in with downloaded data
transcriptome.data.dir <- here("Results", "DO", "Transcriptomes") #directory holding results from 1a.Tissue_Expression.Rmd
```

Load code and libraries.

```{r load_code, message = FALSE, warning = FALSE}
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}

all.packages <- c("qtl2", "gprofiler2", "pheatmap", "knitr", "DT", "hexbin")
load_libraries(all.packages, personal.library = TRUE)
```

Read in data. We use expression and traits that have
already been adjusted for covariates and filtered for
pseudogenes in previous workflows. See code for details.


```{r read_transcriptomes}
#generated by 1a.Tissue_Expression.Rmd
#these transcripts are normalized and adjusted
#for covariates, but are not mean centered or 
#scaled
adj.expr <- readRDS(file.path(data.results.dir, "Tissue_Expression_Adjusted.RDS")) 
tissue.names <- names(adj.expr)
tissue.cols <- c("orange", "#8dd3c7", "tan", "brown") #colors that are not related to the DO/CC colors

#mean center and standardize gene expression so all variances are 1
scaled.expr <- lapply(adj.expr, function(x) apply(x, 2, scale))
for(tx in 1:length(scaled.expr)){
    dimnames(scaled.expr[[tx]]) <- dimnames(adj.expr[[tx]])
}

load(here("Data", "DO", "QTLViewer_Geno_V10.Rdata")) #ensembl.version, genoprobs, K, map, markers
covar <- readRDS(file.path(data.results.dir, "Clinical_Phenotype_Covariates.RDS")) #generated by 1b.Trait_Selection.Rmd
gene.tables <- readRDS(file.path(data.results.dir, "Gene_Tables.RDS")) #generated by 1a.Tissue_Expression.Rmd

local.coef <- readRDS(file.path(transcriptome.data.dir, "eQTL_coef_local.RDS")) #generated in 1a.Tissue_Expression.Rmd
```


## Impute gene expression

Impute gene expression. We do this by fitting the transcript
values with the local marker genotype matrix. We use the predicted
values from fit1 as the locally imputed transcript. Because the
transcripts are all scaled to have a variance of 1, the variance
of the fitted values is the variance explained by the local genotype.

When we do this, we can also estimate the variance explained by 
the distal genome. To do this, we estimate the heritability of 
the residuals of each transcript abundance after controlling 
for the local eQTL.


```{r kin}
kin.file <- file.path(data.results.dir, "overall.kinship.RDS")
if(!file.exists(kin.file)){
    Kg = calc_kinship(genoprobs, "overall")
    saveRDS(Kg, kin.file)
}else{
    Kg <- readRDS(kin.file)
}
```

```{r overall_herit}
transcript.herit.file <- file.path(transcriptome.data.dir, "Transcript_Heritability.RDS")
if(!file.exists(transcript.herit.file)){
    all.tx.herit <- vector(mode = "list", length = length(tissue.names))
    names(all.tx.herit) <- tissue.names
    for(tx in 1:length(tissue.names)){
        tx.herit <- est_herit(scaled.expr[[tx]], kinship = Kg)
        all.tx.herit[[tx]] <- tx.herit
    }
    saveRDS(all.tx.herit, transcript.herit.file)
}
```

## Imputation

Here we use the nearest marker to each gene to impute
gene expression. We use fit1 with a kinship correction
and take the fitted values of the model as the imputed
expression.

We also calculate variance explained by the distal component 
of gene expression. To get this we estimates the heritability 
of the residuals of each fitted model above.


```{r impute_transcriptomes}
save.every = 100

for(tx in 1:length(tissue.names)){
    start.idx <- 1
    distal.herit.file <- file.path(imputed.results.dir, 
      paste0("Distal_Heritability_", tissue.names[tx], ".RDS"))
    transcript.id <- colnames(adj.expr[[tx]])

    local.imp.file <- file.path(imputed.results.dir, paste0("Local_Imputation_Expr_", tissue.names[tx], ".RDS"))
    
    eQTL.coef.file <- file.path(imputed.results.dir, paste0("eQTL_Coef_", tissue.names[tx], ".RDS"))

    #check to see if we've already saved some results that we can use
    #we only need to use one of the files to figure this out
    if(file.exists(distal.herit.file) && !overwrite.results){
      herit.resid <- readRDS(distal.herit.file)
      start.idx <- max(which(!is.na(herit.resid[,1]))) + 1
      local.imp <- readRDS(local.imp.file)
      eQTL.coef <- readRDS(eQTL.coef.file)
    }
      
  #start where we left off, or at the beginning if there are no data
  #or we are overwriting previous data.
  if(start.idx < length(transcript.id)){
    sink(file.path(imputed.results.dir, paste0("progress_", tissue.names[tx], ".txt")))

    #if we are starting fresh...
    if(!file.exists(distal.herit.file) || overwrite.results){
      herit.resid <- matrix(NA, ncol = 3, nrow = length(transcript.id))
      rownames(herit.resid) <- transcript.id
      colnames(herit.resid) <- c("local_var", "residual_variance", "residual_heritability")

      local.imp <- matrix(NA, ncol = length(transcript.id), nrow = nrow(adj.expr[[tx]]))
      dimnames(local.imp) <- dimnames(adj.expr[[tx]])

      eQTL.coef <- matrix(NA, nrow = length(transcript.id), ncol = 9)
      rownames(eQTL.coef) <- colnames(adj.expr[[tx]])
      colnames(eQTL.coef) <- c(LETTERS[1:8], "lod")
    }
  
    #start where we left off, or at the beginning, if that is warranted
    for(tr in start.idx:length(transcript.id)){
      print(tr)
      
      gene.info.idx <- which(gene.tables[[tx]]$gene.id == transcript.id[tr])
      nearest.marker <- gene.tables[[tx]]$nearest.marker.id[gene.info.idx]

       #we won't be able to find nearest markers for genes on Y or MT chromosomes
      if(nearest.marker == ""){next()}  

      marker.geno <- pull_genoprobpos(genoprobs, marker = nearest.marker)

      trans.expr <- scaled.expr[[tx]][,transcript.id[tr]]
  
      #with kinship
      #I am keeping the variance of the fitted values to use as the local variance explained
      #reserve as much of the signal from kinship as we can by performing a kinship correction
      model.kin <- fit1(marker.geno, trans.expr, kinship = Kg)
      eQTL.coef[tr,] <- c(model.kin$coef[LETTERS[1:8]], model.kin$lod)
      imp.local <- model.kin$fitted
      local.imp[,tr] <- imp.local
      distal.resid <- model.kin$resid
      result <- c(var(imp.local), var(distal.resid), est_herit(distal.resid, Kg))
      herit.resid[tr,] <- result

      if(tr %% save.every == 0){
        saveRDS(herit.resid, distal.herit.file)
        saveRDS(local.imp, local.imp.file)
        saveRDS(eQTL.coef, eQTL.coef.file)
      }
    }
    saveRDS(herit.resid, distal.herit.file)
    saveRDS(local.imp, local.imp.file)
    saveRDS(eQTL.coef, eQTL.coef.file)
    sink()
  }
}
```

```{r read_imp}
local.imp.files <- list.files(imputed.results.dir, pattern = "Local", full.names = TRUE)
tx.local.imp <- lapply(local.imp.files, readRDS)

distal.herit.files <- list.files(imputed.results.dir, pattern = "Distal", full.names = TRUE)
tx.distal.herit <- lapply(distal.herit.files, readRDS)

local.var.exp <- lapply(tx.distal.herit, function(x) x[,"local_var"])
distal.var.exp <- lapply(tx.distal.herit, function(x) x[,"residual_variance"]*x[,"residual_heritability"])
```


## Correlation Check

### Imputed and measured {.tabset .tabset-fade .tabset-pills}
 
Check the correlations of the imputed transcripts with
transcript measurements. Transcripts with higher local 
LOD scores should have higher correlations to the 
measured transcript.

```{r check_correlations, results = "asis", fig.height = 4, fig.width = 8}
for(tx in 1:length(tissue.names)){
    cat("####", tissue.names[tx], "\n")
    measured <- scaled.expr[[tx]]
    imputed <- tx.local.imp[[tx]]

    common.tx <- intersect(colnames(measured), colnames(imputed))
    common.ind <- intersect(rownames(measured), rownames(imputed))
    all.cor <- sapply(1:length(common.tx), 
        function(x) cor(measured[common.ind,common.tx[x]], 
        imputed[common.ind,common.tx[x]]))
    local.lod <- local.coef[[tx]]$lod[match(common.tx, local.coef[[tx]]$gene.id)]
    par(mfrow = c(1,2))
    hist(all.cor, main = "Histogram of Correlations", 
        xlab = "Correlation", xlim = c(0,1))
    plot(local.lod, all.cor, xlab = "local LOD score", 
        main = "Correlation v. LOD",
        ylab = "Correlation", ylim = c(0,1))
    mtext(tissue.names[tx], side = 3, outer = TRUE, line = -1.5, font = 2)
    cat("\n\n")

}
```

### Distal and local heritabilities

Check that local and distal heritabilities are anti-correlated
and never sum to more than 1.

```{r local_distal_check, fig.width = 8, fig.height = 8}

par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    #check.local.var <- apply(tx.local.imp[[tx]], 2, var)
    local.var <- tx.distal.herit[[tx]][,"local_var"]
    #plot(check.local.var, local.var)
    distal.var <- tx.distal.herit[[tx]][,"residual_variance"]
    plot(local.var, distal.var, xlab = "Local Variance Explained",
    ylab = "Distal Variance Explained", main = tissue.names[tx])
}
```

### Local and distal contribution to transcripts

What are the relative contributions of local and distal 
regulation to trait variation?

The following scatter plots show the local and distal variance
explained for each transcript in each tissue. The two measures
are weakly negatively correlated as we expect. Most transcripts
have low heritability on both measures. Some transcripts have
high local heritability and others have high distal heritability.

```{r local_v_distal_by_tx, fig.width = 8, fig.height = 8}
par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    plot.hexbin.as.plot(local.var.exp[[tx]], distal.var.exp[[tx]], xlab = "Local Variance Explained",
        ylab = "Distal Variance Explained", n.bins = 5)
}
```


The following plot compares the overall variance explained by 
local and distal components across the tissues. Both components 
contribute about the same to transcript levels.

We calculated these values in the following way:

1. The variance of the measured transcript is the sum of the 
    variance contributions from covariates, local genotype, 
    distal genotype, and error. We already removed the variance
    from the covariates, so we are left with all the other 
    variance components. 

*transcript = cov + local_genotype + kinship + error*

2. We scaled the remaining variance each transcript (measured minus 
    covariates) to be equal to 1. Thus, the variance contributions
    from the local, distal, and error components sum to 1.

*adj_transcript = local_genotype + kinship + error*

3. We then fit this variance with a model using the local genotype
    and kinship matrix. The residual variance from this model is
    the variance explained by distal genetics and by error. We can
    estimate the proportion explained by distal genetics by estimating
    the heritability of this residual. The *distal variance explained* 
    is the variance of the residual of the above model times the 
    heritability of the residual. 
 

```{r local_var, fig.width = 8, fig.height = 5}
common.tx <- Reduce("intersect", lapply(distal.var.exp, names))
local.distal.comp <- list("Local" = Reduce("cbind", lapply(local.var.exp, function(x) x[common.tx])),
    "Distal" = Reduce("cbind", lapply(distal.var.exp, function(x) x[common.tx])))
for(i in 1:length(local.distal.comp)){
    colnames(local.distal.comp[[i]]) <- tissue.names
}
plot.grouped.boxes(local.distal.comp, type = "matrix", ylab = "Variance Explained",
    main = "Variance Explained by Local and Distal Components")
```

```{r fig, eval = FALSE}
par(mar = c(3,4,1,2))
plot.grouped.boxes(local.distal.comp, type = "matrix", ylab = "Variance Explained",
    main = "")
```

## Heritability and Trait Relevance

If a transcript mediates the effect of local genotype on 
phenotype, we would expect to see a correlation between the
locally inferred transcript and trait.

```{r imp_trait}
clin.pheno <- readRDS(here("Results", "DO", "Data", "Clinical_Phenotypes_Adjusted.RDS")) #generated by 1b.Trait_Selection.Rmd
trait.tx.cor <- vector(mode = "list", length = length(tissue.names))
names(trait.tx.cor) <- tissue.names

for(tx in 1:length(tissue.names)){
    common.ind <- intersect(rownames(tx.local.imp[[tx]]), rownames(clin.pheno))
    trait.tx.cor[[tx]] <- apply(clin.pheno[common.ind,], 2, function(x) apply(tx.local.imp[[tx]][common.ind,], 2, function(y) cor(x,y, use = "pairwise.complete.obs")))
}

plot.grouped.boxes(trait.tx.cor, type = "matrix", print.vals = NA, 
    group.cols = tissue.cols, cex = 0.5, label.srt = 90, within.group.sep = 0.7, 
    between.group.sep = 1.5, legend.x = 69, legend.y = 0.1)
plot.dim <- par("usr")
segments(x0 = 0, x1 = 68, y0 = seq(-0.2, 0.3, 0.1), lty = 2, col = "darkgray")
```

We want to examine the relationship between heritability and 
trait relevance. We define trait relevance here as the correlation
between the transcript and the trait. Heritability we break into
two components: local heritability and distal heritability.

Local heritability can be defined either with the local LOD
score, or the variance explained by the local marker. We will
use the variance explained by the local marker to be more analogous
both with the human literature and with the distal heritability
measure. Distal heritability is the variance eplained by the 
kinship matrix for the transcript levels after controlling for the
local marker.


### Local Variance Explained and Trait Relavance {.tabset .tabset-fade .tabset-pills}

The following plots show the correlation between the 
variance explained by the local marker and the trait
relevance (maximum correlation between the transcript
and all traits.)

#### Variance Explained - Loess

```{r local_v_traits, fig.height = 8, fig.width = 8}
trait.cor.files <- list.files(file.path(transcriptome.data.dir), 
    pattern = "Maximum_Trait_Correlation", full.names = TRUE) #generated in 2a.Kinship_Expression_Traits.Rmd
trait.cor <- lapply(trait.cor.files, readRDS)
names(trait.cor) <- tissue.names

par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){

    common.tx <- intersect(names(trait.cor[[tx]]), names(local.var.exp[[tx]])) #transcripts should be in the same order, but just make sure

    binned_ts <- bin.vector2(local.var.exp[[tx]][common.tx], seq(0, 1, 0.05))
    bin_means <- sapply(binned_ts, function(x) mean(abs(trait.cor[[tx]][names(x)]), na.rm = TRUE))
    x_means <- sapply(binned_ts, mean)

    plot(local.var.exp[[tx]][common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Local Variance Explained", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], pch = 16)
    points(x_means, bin_means, col = "lightblue", type = "l", lwd = 3)
}
```

#### Variance Explained - Linear

The following plots are the same as above, but show the linear fit.

```{r local_lin_fit, fig.height = 8, fig.width = 8}
par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){
    common.tx <- intersect(names(trait.cor[[tx]]), names(local.var.exp[[tx]]))

    plot.with.model(local.var.exp[[tx]][common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Local Variance Explained", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], report = "cor.test")
}
```

#### LOD - Linear

The following plots show the relationship between trait
relevance and local LOD score.

```{r herit_lod, fig.width = 8, fig.height = 8}
par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    local.lod <- unlist(local.coef[[tx]][,"lod"])
    names(local.lod) <- unlist(local.coef[[tx]][,"gene.id"])
    common.tx <- intersect(names(local.lod), names(trait.cor[[tx]]))

    plot.with.model(local.lod[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Local Variance Explained", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], report = "cor.test")
}

```

### Distal Variance Explained and Trait Relavance {.tabset .tabset-fade .tabset-pills}

#### Variance Explained - Loess

```{r, fig.width = 8, fig.height = 8}

par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){
    distal.herit <- tx.distal.herit[[tx]][,"residual_variance"]*tx.distal.herit[[tx]][,"residual_heritability"]
    common.tx <- intersect(names(trait.cor[[tx]]), names(distal.herit))

    binned_ts <- bin.vector2(distal.herit[common.tx], seq(0, 1, 0.05))
    bin_means <- sapply(binned_ts, function(x) mean(abs(trait.cor[[tx]][names(x)]), na.rm = TRUE))
    x_means <- sapply(binned_ts, mean)

    plot(distal.herit[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Variance Explained", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], pch = 16)
    points(x_means, bin_means, col = "lightblue", type = "l", lwd = 3)
}

```

```{r hexbin, eval = FALSE}

par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){
    distal.herit <- tx.distal.herit[[tx]][,"residual_variance"]*tx.distal.herit[[tx]][,"residual_heritability"]
    common.tx <- intersect(names(trait.cor[[tx]]), names(distal.herit))

    binned_ts <- bin.vector2(distal.herit[common.tx], seq(0, 1, 0.05))
    bin_means <- sapply(binned_ts, function(x) mean(abs(trait.cor[[tx]][names(x)]), na.rm = TRUE))
    x_means <- sapply(binned_ts, mean)

    plot.hexbin.as.plot(distal.herit[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Variance Explained", ylab = "Max Correlation with Trait",
        main = tissue.names[tx])
    points(x_means, bin_means, col = "lightblue", type = "l", lwd = 3)
}
```

#### Variance Explained - Linear

The following figures show the same data, but with linear fit models.

```{r lin_fig, fig.width = 8, fig.height = 8}
par(mfrow = c(2,2))
for(tx in 1:length(trait.cor)){
    distal.herit <- tx.distal.herit[[tx]][,"residual_variance"]*tx.distal.herit[[tx]][,"residual_heritability"]
    common.tx <- intersect(names(trait.cor[[tx]]), names(distal.herit))

    plot.with.model(distal.herit[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Variance Explained", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], report = "cor.test")
}
```

## Distal Tails {.tabset .tabset-fade .tabset-pills}

The plots with distal variance explained vs. trait relevance
have prominant tails with high variance explained but very low
trait relevance. What are these genes?

```{r tails, fig.width = 10, fig.height = 4.5, results = "asis"}

tail.boundaries <- matrix(c(0.5, 0.3, 0.7, 0.2, 0.5, 0.3, 0.55, 0.2), ncol = 2, byrow = TRUE)
rownames(tail.boundaries) <- tissue.names
colnames(tail.boundaries) <- c("Variance Explained", "Trait Relatedness")

tail.enrich <- tail.genes <- vector(mode = "list", length = length(tissue.names))
names(tail.enrich) <- names(tail.genes) <- tissue.names

#pdf("~/Desktop/tails.pdf", width = 8, height = 5)
for(tx in 1:length(trait.cor)){
    cat("###", tissue.names[tx], "\n")

    distal.herit <- tx.distal.herit[[tx]][,"residual_variance"]*tx.distal.herit[[tx]][,"residual_heritability"]
    common.tx <- intersect(names(trait.cor[[tx]]), names(distal.herit))

    tail.idx <- intersect(which(distal.herit[common.tx] > tail.boundaries[tx,1]), 
        which(abs(trait.cor[[tx]][common.tx]) < tail.boundaries[tx,2]))
    tail.genes[[tx]] <- common.tx[tail.idx]
    tail.enrich[[tx]] <- gost(tail.genes[[tx]], organism = "mmusculus", 
        sources = c("GO", "REACTOME", "HP", "CORuM"))


    pt.col <- rep("black", length(common.tx))
    pt.col[tail.idx] <- "salmon"

    layout(matrix(c(1,2), ncol = 2), widths = c(0.8, 1))

    par(mar = c(4,4,4,4))
    plot.with.model(distal.herit[common.tx], abs(trait.cor[[tx]][common.tx]),
        xlab = "Variance Explained", ylab = "Max Correlation with Trait",
        main = tissue.names[tx], report = "cor.test", col = pt.col)

    plot.enrichment(tail.enrich[[tx]], order.by = "p_value", num.terms = 20, max.term.size = 3000,
        plot.label = paste("Tail Enrichment for", tissue.names[tx]), text.size = 0.7)

    cat("\n\n")
}
#dev.off()
```

### All Tissues

All the tails are enriched for ribosomal genes.

```{r all_enrich, fig.height = 7, fig.width = 5}
plot.enrichment.group(tail.enrich, plot.label = "All Tail Enrichments", max.term.size = 3000,
    n.terms = 20)
```

### All Genes

The individual gene names in each tail are shown below

```{r tail_genes, fig.width = 3, fig.height = 15}
all.tail <- Reduce("union", tail.genes)

tail.table <- matrix(0, nrow = length(all.tail), ncol = length(tissue.names))
rownames(tail.table) <- all.tail
colnames(tail.table) <- tissue.names

for(tx in 1:length(tissue.names)){
    tail.table[tail.genes[[tx]],tx] <- 1
}

gene.info  <- cbind(Reduce("c", lapply(gene.tables, function(x) x$symbol)), 
    Reduce("c", lapply(gene.tables, function(x) x$gene.id)))
gene.names  <- gene.info[match(all.tail, gene.info[,2]),1]
rownames(tail.table) <- gene.names

#pdf("~/Desktop/tail.genes.pdf", width = 3, height = 15)
pheatmap(tail.table, cluster_cols = FALSE, cex = 0.7)
#dev.off()


#gene.order <- order(rownames(tail.table))
#pheatmap(tail.table[gene.order,], cluster_rows = FALSE, cluster_cols = FALSE, cex = 0.7)

#par(mfrow = c(2,2))
#for(tx in 1:length(tissue.names)){
#    var.test <- apply(adj.expr[[tx]], 2, var)
#    not.tail <- setdiff(colnames(adj.expr[[tx]]), tail.genes[[tx]])
#    boxplot(list("Tail Genes" = log10(var.test[tail.genes[[tx]]]), 
#        "Other Genes" = log10(var.test[not.tail])), main = tissue.names[tx])
#}
```

## Transcription Networks {.tabset .tabset-fade .tabset-pills}

How well does imputation recapitulate the correlation structure
in the transcriptome? Transcripts tend to be highly correlated
in real life, and we know that this correlation structure is 
important. However, alleles in the genome are inherited randomly
and independently. Thus, we expect that the imputed transcriptome
completely mis-specifies the transcript correlation structure. 

For now we are just sampling the correlation matrix instead of
calculating it exhaustively. If this is a useful analysis, we'll
do it properly.

What a strange relationship these two things happen to have!
The imputed transcripts are mostly uncorrelated with each
other whereas the measured transcripts are much more highly
correlated, so we get this strange plus shape in the plot
of the two measures against each other. There is an 
ever-so-slightly positively correlation between the 
correlation structures, but for the most part, the imputation
completely mis-specifies the correlation structure.

The blue lines below show the line of best fit. The red lines show
y = x.

```{r transcript_correlations, fig.width = 5, fig.height = 5, warning = FALSE, results = "asis"}
sample.size = 100
n.samples = 100

measured.cor <- imputed.cor <- vector(mode = "list", length = length(tissue.names))
names(measured.cor) <- names(imputed.cor) <- tissue.names
for(tx in 1:length(tissue.names)){  
    measured <- scaled.expr[[tx]]
    imputed <- tx.local.imp[[tx]]

    common.tx <- intersect(colnames(measured), colnames(imputed))
    common.ind <- intersect(rownames(measured), rownames(imputed))

    measured.cor.v <- imp.cor.v <- NULL
    for(sp in 1:n.samples){
        tx.sample <- sample(common.tx, sample.size)
        sampled.measured.cor <- cor(measured[,tx.sample], use = "pairwise.complete.obs")
        sampled.imp.cor <- cor(imputed[,tx.sample], use = "pairwise.complete.obs")
        measured.cor.v <- c(measured.cor.v, sampled.measured.cor[upper.tri(sampled.measured.cor, diag = FALSE)])
        imp.cor.v <- c(imp.cor.v, sampled.imp.cor[upper.tri(sampled.imp.cor, diag = FALSE)])
    }
    measured.cor[[tx]] <- measured.cor.v
    imputed.cor[[tx]] <- imp.cor.v
}


for(tx in 1:length(tissue.names)){
    cat("###", tissue.names[tx], "\n")
    plot_with_marginal_hist(x = measured.cor[[tx]], y = imputed.cor[[tx]], main = tissue.names[tx],
    xlab = "Measured Correlation", ylab = "Imputed Correlation", report = "cor.test")
    cat("\n\n")
}

#hist(Kg[upper.tri(Kg, diag = FALSE)], breaks = 100, main = "Kinship coefficient", xlab = "Kinship coefficient")
#hist(abs(unlist(imputed.cor)), breaks = 100, main = "Correlation of Imputed Transcripts", xlab = "Correlation coefficient")
```


## TWAS {.tabset .tabset-fade .tabset-pills}

We correlated each imputed transcript with the traits.


```{r max_cor_fun}
#this function finds the maximum absolute value of 
#a vector of correlations, and returns the correlation
#with its original sign.

max_cor <- function(cor_v){
    max.idx <- which.max(abs(cor_v))
    return(cor_v[max.idx])
}

```

```{r twas}
clin.pheno <- readRDS(here("Results", "DO", "Data", "Clinical_Phenotypes_Adjusted.RDS")) #generated by 1b.Trait_Selection.Rmd
trait.names <- colnames(clin.pheno)

imp.cor.file <- file.path(transcriptome.data.dir, "TWAS_imputed_expr_trait_cor.RDS")

if(!file.exists(imp.cor.file)){
    tx.imp.cor <- vector(mode = "list", length = length(tx.local.imp))
    names(tx.imp.cor) <- tissue.names

    for(tx in 1:length(tx.local.imp)){
        tx.imp <- tx.local.imp[[tx]]
        common.ind <- intersect(rownames(tx.imp), rownames(clin.pheno))
        imp.cor <- apply(tx.imp[common.ind,], 2, 
            function(x) apply(clin.pheno[common.ind,], 2, 
            function(y) cor(x, y, use = "pairwise.complete.obs")))
        #hist(imp.max.cor)
        tx.imp.cor[[tx]] <- imp.cor
    }
    saveRDS(tx.imp.cor, imp.cor.file)
}else{
    tx.imp.cor <- readRDS(imp.cor.file)
}

```

In TWAS we care about the association of the imputed gene 
expression with traits. To assess the significance of imputed 
gene expression with traits here, we do permutations. We permute 
the labels on the phenotypes and recalculate transcript-trait 
associations.

```{r perm}
nperm = 10

cor.perm.file <- file.path(transcriptome.data.dir, "TWAS_imputed_expr_trait_cor_perm.RDS")

if(!file.exists(cor.perm.file)){
    perm.imp.cor <- vector(mode = "list", length = length(tx.local.imp))
    names(perm.imp.cor) <- tissue.names

    for(tx in 1:length(tx.local.imp)){
        tx.imp <- tx.local.imp[[tx]]
        common.ind <- intersect(rownames(tx.imp), rownames(clin.pheno))
        perm.list <- vector(mode = "list", length = nperm)
        for(p in 1:nperm){
            print(p)
            perm.ind <- sample(common.ind)
            perm.list[[p]] <- apply(tx.imp[common.ind,], 2, 
                function(x) apply(clin.pheno[perm.ind,], 2, 
                function(y) cor(x, y, use = "pairwise.complete.obs")))
            #hist(perm.list[[p]])
        }
        perm.imp.cor[[tx]] <- perm.list
    }
    saveRDS(perm.imp.cor, cor.perm.file)
}else{
    perm.imp.cor <- readRDS(cor.perm.file)
}
```

```{r plot_comparison, results = "asis", fig.width = 10, fig.height = 7, warning = FALSE}
#plot_comparison = FALSE
plot_comparison = TRUE
#calculate sigificant correlations for imputed transcripts
#pdf("~/Desktop/twas.pdf", width = 10, height = 7)
sig.tx <- vector(mode = "list", length = length(tissue.names))
names(sig.tx) <- tissue.names
for(tx in 1:length(tx.local.imp)){
    cat("###", tissue.names[tx], "{.tabset .tabset-fade .tabset-pills}\n")
    tx.obs.cor <- tx.imp.cor[[tx]]
    tx.perm.cor <- perm.imp.cor[[tx]]

    tr.sig <- vector(mode = "list", length = nrow(tx.obs.cor))
    names(tr.sig) <- rownames(tx.obs.cor)
    #look at each trait independently
    for(tr in 1:nrow(tx.obs.cor)){
        cat("####", rownames(tx.obs.cor)[tr], "\n")
        tr.obs.cor <- tx.obs.cor[tr,]
        tr.null.cor <- unlist(lapply(tx.perm.cor, function(x) x[tr,]))

        permv  <- abs(unlist(tr.null.cor))
        obsv <- abs(unlist(tr.obs.cor))

        if(plot_comparison){
            max.cor <- 0.3
            null.dens <- density(permv[which(is.finite(permv))], from = 0, to = max.cor)
            obs.dens <- density(obsv[which(is.finite(obsv))], from = 0, to = max.cor)
            xlim <- c(0, max(c(null.dens$x, obs.dens$x)))
            
            layout.mat <- matrix(c(rep(c(1,2), each = 7), rep(3, 6), c(1:20+4)), nrow = 2, byrow = TRUE)
            layout.mat <- cbind(c(0,4), layout.mat, c(0,0))
            
            layout(layout.mat)
            par(mar = c(4,4,4,4))
            plot(null.dens, type = "l", lwd = 2, xlim = xlim, main = "TWAS correlations")
            points(obs.dens, type = "l", lwd = 2, col = "red")
            legend("topright", lty = 1, col = c("black", "red"), legend = c("null", "observed"))
            
            #range.col <- colors.from.values(null.dens$x, use.pheatmap.colors = TRUE)
            #plot(null.dens$y, obs.dens$y, xlab = "Null density", 
            #    ylab = "Observed density", col = range.col, pch = 16, 
            #    main = "Densities Compared")
            #abline(0,1)
            #add_color_bar(null.dens$x, position = "bottomright", use.pheatmap.colors = TRUE,
            #    margin.buffer = 0.05, label.buffer = 0.03, legend.title = "Cor")
            
            mtext("Correlation", side = 3)
            mtext(paste(tissue.names[tx], "-", trait.names[tr]), 
                side = 3, outer = TRUE, line = -1.5)
        }

        emp.p <- calc.emp.p(obsv, permv)
        
        if(plot_comparison){
            qqunif.plot(emp.p)
        }

        #Manhattan plot
        tx.idx <- match(names(emp.p), gene.tables[[tx]]$gene.id)
        tx.table <- cbind(gene.tables[[tx]][tx.idx,c(1,2,4,5)], "Empirical_p" = emp.p)
        tx.table$chr[which(tx.table$chr == "X")] <- 20
        sorted.table <- sort.by.then.by(tx.table, c(3,4), c("n", "n"))

        neg.log.p <- -log10(sorted.table[,"Empirical_p"])
        max.p <- max(neg.log.p[which(is.finite(neg.log.p))])
        neg.log.p[which(!is.finite(neg.log.p))] <- max.p + 1

        sig.table <- sorted.table[which(neg.log.p >= max.p),]
        tr.sig[[tr]] <- sig.table

        if(plot_comparison){
            par(mar = c(4,0,4,0))
            if(nrow(sig.table) > 0){
                sig.transcripts <- sig.table$symbol                
                sig.par <- words_to_paragraph(sig.transcripts, line.len = 4)
                plot.text(sig.par)
            }else{
                plot.text("No significant transcripts.")
            }


            ymax <- max(neg.log.p)
            par(mar = c(4,2,4,0))
            plot.new()
            plot.window(xlim = c(0,1), ylim = c(0, ymax))
            axis(2, line = -2)
            mtext("-log10(p)", side = 2, line = 0.5)
            
            for(ch in 1:20){
                chr.idx <- which(sorted.table$chr == ch)
                par(mar = c(4,0,4,0))
                plot(neg.log.p[chr.idx], type = "h", ylim = c(0, ymax), axes = FALSE,
                    xlab = "", ylab = "")
                mtext(ch, side = 1, line = 0)
                abline(h = max.p, col = "red")
            }
        }
        cat("\n\n")
    } #stop looping through traits
    sig.tx[[tx]] <- tr.sig
    cat("\n\n")
} #stop looking through tissues
#dev.off()

#combine results for traits in each tissue
twas.tissue <- vector(mode = "list", length = length(sig.tx))
names(twas.tissue) <- tissue.names
for(tx in 1:length(twas.tissue)){
    tx.results <- sig.tx[[tx]]
    trait.results <- lapply(1:length(tx.results), function(x) rep(trait.names[x], nrow(tx.results[[x]])))
    tx.table <- cbind(Reduce("rbind", tx.results), "trait" = unlist(trait.results))
    twas.tissue[[tx]] <- tx.table
}

tissue.label <- lapply(1:length(twas.tissue), function(x) rep(tissue.names[x], nrow(twas.tissue[[x]])))
twas.table <- cbind(Reduce("rbind", twas.tissue), "Tissue" = unlist(tissue.label))
write.table(twas.table, file.path(imputed.results.dir, "TWAS_results.csv"), sep = ",", 
    quote = FALSE, row.names = FALSE)
#barplot(sort(table(twas.table[,"symbol"])), las = 2)
#mhead(sort(table(twas.table[,"symbol"]), decreasing = TRUE))
```


The table of significant TWAS hits is below.

```{r twas_table}
datatable(twas.table)
```

```{r test, eval = FALSE}
par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    imp.tx <- sapply(strsplit(names(max.imp.cor[[tx]]), ".", fixed = TRUE), function(x) x[1])
    names(max.imp.cor[[tx]]) <- imp.tx
    common.tx <- intersect(imp.tx, names(trait.cor[[tx]]))
    
    plot.with.model(trait.cor[[tx]][common.tx], abs(max.imp.cor[[tx]][common.tx]),
        xlab = "Trait Correlation", ylab = "Imputed Trait Correlation",
        main = tissue.names[tx], report = "cor.test")
}

par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    imp.tx <- sapply(strsplit(names(max.imp.cor[[tx]]), ".", fixed = TRUE), function(x) x[1])
    names(max.imp.cor[[tx]]) <- imp.tx
    
    local.var <- tx.distal.herit[[tx]][,"local_var"]
    common.tx <- intersect(names(max.imp.cor[[tx]]), names(local.var))
    plot.with.model(local.var[common.tx], abs(max.imp.cor[[tx]][common.tx]),
        xlab = "Local Variance Explained", ylab = "Imputed Trait Correlation",
        main = tissue.names[tx], report = "cor.test")
}

par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    imp.tx <- sapply(strsplit(names(max.imp.cor[[tx]]), ".", fixed = TRUE), function(x) x[1])
    names(max.imp.cor[[tx]]) <- imp.tx
    distal.var <- tx.distal.herit[[tx]][,"residual_variance"]
    common.tx <- intersect(names(max.imp.cor[[tx]]), names(distal.var))
    plot.with.model(distal.var[common.tx], abs(max.imp.cor[[tx]][common.tx]),
        xlab = "Distal Variance Explained", ylab = "Imputed Trait Correlation",
        main = tissue.names[tx], report = "cor.test")
}    

```

### Predictive TWAS
We have a result that we are not the sum of our eQTLs because loading
times imputed transcript does not predict trait. But what it loading
isn't the right number to use here. What about just correlation of the
imputed transcript with the trait, as we do in TWAS? That might be a
more direct test of this question. Or is it totally circular?

The barplots below show the correlations between this eQTL-predicted
trait and the actual trait. Weirdly, adipose has the lowest correlations.

```{r pred_twas, fig.width = 12, fig.height = 7}
all.imp.cor <- vector(mode = "list", length = length(tissue.names))
names(all.imp.cor) <- tissue.names
for(tx in 1:length(tissue.names)){
    ind.expr <- scaled.expr[[tx]]
    imp.cor <- tx.imp.cor[[tx]]
    common.tx <- intersect(colnames(ind.expr), colnames(imp.cor))
    pred <- sapply(1:nrow(imp.cor), function(x) colMeans(apply(ind.expr[,common.tx], 1, function(y) imp.cor[x,common.tx]*y), na.rm = TRUE))
    colnames(pred) <- rownames(imp.cor)
    common.ind <- intersect(rownames(pred), rownames(clin.pheno))
    pred.cor <- sapply(1:ncol(clin.pheno), function(x) cor(clin.pheno[common.ind,x], pred[common.ind,x], use = "pairwise.complete.obs"))
    names(pred.cor) <- trait.names
    barplot(sort(pred.cor), las = 2, main = tissue.names[tx])
    abline(h = 0.5, col = "red")
    all.imp.cor[[tx]] <- pred.cor
}
```

The following boxplot compares the correlations across tissues.

```{r cor_box, fig.height = 5, fig.width = 7}
boxplot(all.imp.cor)
```