---
title: "eQTL"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
bibliography: hdm.bib 
---

```{r param}
rm(list = ls()) #clear out previous R environment
remove.pseudogenes = TRUE #if TRUE, removes pseudogenes and predicted genes 
is.interactive = FALSE
overwrite.results = FALSE #set to TRUE to overwrite previous results
```

## Introduction
In this workflow, we recalculate all the eQTLs so we can compare 
the variance explained by each haplotype for local and distal eQTL.

This workflow depends on results from 
1a.Tissue_Expression.Rmd, 1b.Trait_Selection.Rmd, and 
2a.Kinship_Expression_Traits.Rmd.

```{r here, message = FALSE, warning = FALSE}
library(here)
eqtl.results.dir <- here("Results", "DO", "eQTL") #result generated by this workflow
if(!file.exists(eqtl.results.dir)){dir.create(eqtl.results.dir, recursive = TRUE)}
data.results.dir <- here("Results", "DO", "Data") #a results folder that holds parsed data that we don't want to mix in with downloaded data
transcriptome.data.dir <- here("Results", "DO", "Transcriptomes") #directory holding results from 1a.Tissue_Expression.Rmd
```

Load code and libraries.

```{r load_code, message = FALSE, warning = FALSE}
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(i in 1:length(all.fun)){source(all.fun[i])}

all.packages <- c("qtl2", "gprofiler2", "pheatmap", "knitr", "DT", "hexbin", 
    "RColorBrewer", "vioplot")

load_libraries(all.packages)
```

Read in data. We use expression and traits that have
already been adjusted for covariates and filtered for
pseudogenes in previous workflows. See code for details.


```{r read_transcriptomes}
#generated by 1a.Tissue_Expression.Rmd
#these transcripts are normalized and adjusted
#for covariates, but are not mean centered or 
#scaled
adj.expr <- readRDS(file.path(data.results.dir, "Tissue_Expression_Adjusted.RDS")) 
tissue.names <- names(adj.expr)
tissue.cols <- as.matrix(read.delim(here("Data", "general", "tissue_colors.txt"), header = FALSE, row.names = 1))[,1]

#mean center and standardize gene expression so all variances are 1
scaled.expr <- lapply(adj.expr, function(x) apply(x, 2, scale))
for(tx in 1:length(scaled.expr)){
    dimnames(scaled.expr[[tx]]) <- dimnames(adj.expr[[tx]])
}

load(here("Data", "DO", "QTLViewer_Geno_V10.Rdata")) #ensembl.version, genoprobs, K, map, markers
covar <- readRDS(file.path(data.results.dir, "Clinical_Phenotype_Covariates.RDS")) #generated by 1b.Trait_Selection.Rmd
gene.tables <- readRDS(file.path(data.results.dir, "Gene_Tables.RDS")) #generated by 1a.Tissue_Expression.Rmd

```


## Scan eQTLs

Use qtl2. Because the
transcripts are all scaled to have a variance of 1, the variance
of the fitted values is the variance explained by the local genotype.


```{r kin}
kin.file <- file.path(data.results.dir, "overall.kinship.RDS")
if(!file.exists(kin.file)){
    Kg = calc_kinship(genoprobs, "overall")
    saveRDS(Kg, kin.file)
}else{
    Kg <- readRDS(kin.file)
}
```

```{r extract_local_distal}
label_local_distal <- function(peak.table, chr, pos, dist.thresh = 4){
    
    pos.labels <- rep(NA, nrow(peak.table))
    for(i in 1:nrow(peak.table)){
        if(peak.table[i,"chr"] != chr){
            pos.labels[i] <- "distal"
        }else{
            marker.dist <- abs(pos - peak.table[i,"pos"])
            if(marker.dist <= dist.thresh){
                pos.labels[i] <- "local"
            }else{
                pos.labels[i] <- "distal"
            }
        }
    }
    return(pos.labels)
}

get_coef <- function(peak.table, genoprobs, map, tr.pheno){
    coef.table <- matrix(NA, nrow = nrow(peak.table), ncol = 8)
    colnames(coef.table) <- LETTERS[1:8]
    for(i in 1:nrow(peak.table)){
        marker.geno <- pull_genoprobpos(genoprobs, map = map, chr = peak.table[i,"chr"], pos = peak.table[i,"pos"])
        coef.scan <- fit1(marker.geno, tr.pheno, kinship = Kg)
        coef.table[i,] <- coef.scan$coef[LETTERS[1:8]]
    }
    result <- cbind(peak.table[,2:ncol(peak.table)], coef.table)
    return(result)
}

```

```{r eqtl}
save.every = 100
lodthresh = 8
dist.thresh = 4

for(tx in 1:length(tissue.names)){
        
    start.idx <- 1
    transcript.id <- colnames(scaled.expr[[tx]])
    
    local.eQTL.coef.file <- file.path(eqtl.results.dir, paste0("local_eQTL_", tissue.names[tx], ".RDS"))
    local.eQTL.coef.file <- file.path(eqtl.results.dir, paste0("distal_eQTL_", tissue.names[tx], ".RDS"))

    #check to see if we've already saved some results that we can use
    #we only need to use one of the files to figure this out
    if(file.exists(local.eQTL.coef.file) && !overwrite.results){
      local.eQTL.coef <- readRDS(local.eQTL.coef.file)
      distal.eQTL.coef <- readRDS(distal.eQTL.coef.file)
      local.len <- sapply(local.eQTL.coef, length)
      start.idx <- max(which(local.len > 0))
    }
      
  #start where we left off, or at the beginning if there are no data
  #or we are overwriting previous data.
  if(start.idx < length(transcript.id)){
    sink(file.path(eqtl.results.dir, paste0("progress_", tissue.names[tx], ".txt")))

    #if we are starting fresh...
    if(!file.exists(eQTL.coef.file) || overwrite.results){
      local.eQTL.coef <- distal.eQTL.coef <- vector(mode = "list", length = length(transcript.id))
      names(local.eQTL.coef) <- names(distal.eQTL.coef) <- transcript.id
    }
  

    #start where we left off, or at the beginning, if that is warranted
    for(tr in start.idx:length(transcript.id)){
        print(tr)

        tr.data <- scaled.expr[[tx]][,tr]
        one.scan <- scan1(genoprobs, tr.data)
        #plot(one.scan, map = map)

        tr.peak <- find_peaks(one.scan, map = map, threshold = lod.thresh)

        if(nrow(tr.peak) > 0){
            gene.info.idx <- which(gene.tables[[tx]]$gene.id == transcript.id[tr])
            gene.chr <- gene.tables[[tx]]$chr[gene.info.idx]
            gene.pos <- gene.tables[[tx]]$start[gene.info.idx]
            loc.dist <- label_local_distal(tr.peak, gene.chr, gene.pos, dist.thresh)
            qtl.coef <- get_coef(tr.peak, genoprobs, map, tr.data)
            qtl.coef[,1] <- rep(transcript.id[tr], nrow(qtl.coef))

            local.idx <- which(loc.dist == "local")
            if(length(local.idx) > 0){
                local.eQTL.coef[[tr]] <- qtl.coef[local.idx,,drop=FALSE]
            }else{
                local.eQTL.coef[[tr]] <- NA
            }
            distal.idx <- which(loc.dist == "distal")
            if(length(distal.idx) > 0){
                local.eQTL.coef[[tr]] <- qtl.coef[distal.idx,,drop=FALSE]
            }else{
                local.eQTL.coef[[tr]] <- NA
            }

        }else{ #if there are no eQTL put NAs for both local and distal
            local.eQTL.coef[[tr]] <- NA
            distal.eQTL.coef[[tr]] <- NA
        }

      if(tr %% save.every == 0){
        saveRDS(local.eQTL.coef, local.eQTL.coef.file)
        saveRDS(distal.eQTL.coef, distal.eQTL.coef.file)
      }
    }
    saveRDS(local.eQTL.coef, local.eQTL.coef.file)
    saveRDS(distal.eQTL.coef, distal.eQTL.coef.file)
    sink()
  }
}
```
