---
title: "CC-RIX Metformin Effects"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

The purpose of this workflow is to explore the effects of 
metformin, diet, and sex, and the interactions between 
these things on gene expression in the CC-RIX.

This workflow is outside of the DO manuscript but depends 
on running the data parsing workflows for the CC-RIX.

Here we fit a linear model to each measured transcript 
that includes sex, diet, and treatment. This will tell 
us the contribution of each of these factors to the 
measured gene expression. 


```{r get_args}
args <- commandArgs(trailingOnly = T)
tissue.name <- args[1]
cape_type <- args[2]
delete_previous <- as.logical(args[3])

if(is.na(tissue.name)){
    tissue.name = "adipose"
    cape_type = "query"
    #cape_type = "original"
    delete_previous <- FALSE
}
is.interactive <- FALSE
#is.interactive <- TRUE
```

This report shows results for `r tissue.name` tissue.

## Load data and code

```{r source_code}
library("here")
all.fun <- list.files(here("Code"), pattern = ".R", full.names = TRUE)
for(j in 1:length(all.fun)){source(all.fun[j])}
```


```{r load_libraries,  warning = FALSE, message = FALSE, error = FALSE}
#update.packages("vctrs", repos = "http://cran.us.r-project.org")

needed.packages <- c("pheatmap", "qtl2", "RColorBrewer", 
    "biomaRt", "stringr", "gprofiler2", "DT")
load_libraries(needed.packages)

#these are hard-coded so as not to interfere with the 
#HDM manuscript
load_latest_cape("~/Documents/git_repositories/cape")
#load query cape
query.fun <- list.files("~/Documents/git_repositories/query_cape", pattern = ".R", full.names = TRUE)
for(j in 1:length(query.fun)){source(query.fun[j])}
```

```{r set_dir}
data.dir <- here("Data", "CC-RIX") #where original data are stored
processed.data.dir <- here("Results", "CC-RIX", "Data") #where parsed data are stored
results.dir <- here("Results", "CC-RIX", "Gene_Expression_Effects") #where results from this workflow are stored
if(!file.exists(results.dir)){dir.create(results.dir)}
```


Get the expression for the specified tissue.


```{r read_data}
all.var <- ls()
data.loaded <- as.logical(length(which(all.var == "expr")))

if(!data.loaded){
    annot <- read.csv(file.path(data.dir, "final_sample_annotations.csv"))
    tissue.expr <- readRDS(file.path(processed.data.dir, "Expression.by.Tissue.RDS"))
    manifest <- as.matrix(read.csv(file.path(data.dir, "Original_Samples_Manifest.csv")))
    covar <- as.matrix(read.csv(file.path(processed.data.dir, "Covar.csv"), row.names = 1))
    groups <- read.csv(file.path(processed.data.dir, "Groups.csv"), row.names = 1)
    group.def <- colnames(groups)

    gene.info.file <- here("Data", "general", "mouse_gene_info.txt")
    gene.info <- read.delim(gene.info.file)
    
}
```


```{r adjust, warning = FALSE, message = FALSE, error = FALSE}
tissue.locale <- which(names(tissue.expr) == tissue.name)

tissue.info <- unique(t(sapply(rownames(tissue.expr[[tissue.locale]]), 
  function(x) get_mouse_info(x, manifest, "Barcode.1"))))

orsam.id <- names(rownames(tissue.expr[[tissue.locale]]))
orsam.idx <- match(orsam.id, rownames(covar))

#adjust for age and RNASeq Batch
rnaseq.covar <- dummy_covar(covar[orsam.idx,"RNAseq_Batch",drop=FALSE])
adj.cc.expr <- adjust(tissue.expr[[tissue.locale]], rnaseq.covar)

mouse.info <- unique(t(sapply(rownames(covar), function(x) get_mouse_info(x, manifest, "User.Inputs"))))
rownames(mouse.info) <- mouse.info[,1]
```

The metformin treatment did not start until the animals were 6 months
old. If we want to compare the effects of metformin treatment to no
treatment, we can only compare animals in the timepoint after metformin
started. We thus removed all animals from the earlier timepoint.

```{r remove_timepoint}
keep.rows <- which(mouse.info[,"Timepoint"] == "57")
keep.info <- mouse.info[keep.rows,]

matched.mats <- get.xz(adj.cc.expr, keep.info)
matched.expr <- matched.mats$X
expr.matched.info <- matched.mats$Z
saveRDS(matched.expr, file.path(processed.data.dir, paste0("Adjusted.Expression.", tissue.name, ".RDS")))
```

## Group Expression

Group the measured expression based on the CC-RIX crosses.

```{r group_expr}
cross.id <- expr.matched.info[,"User.Defined.Strain"]
u_cross <- unique(cross.id)
cross.idx <- lapply(u_cross, function(x) which(cross.id == x))
cross_expr <- lapply(cross.idx, function(x) matched.expr[x,])
names(cross_expr) <- u_cross
mean.cross.expr <- t(sapply(cross_expr, colMeans))
```

## Expression Decomposition

The following plots show the decompmosition of the gene expression
matrix. Each point represents one individual. Its strain ID is 
indicated by a number. Points are colored by Sex, Diet, or Treatment.

```{r expr_decomp, fig.width = 8, fig.height = 8}
expr.decomp <- plot.decomp(matched.expr, pc = 10, plot.results = FALSE)
xlab = paste0("PC1 (", signif(expr.decomp$var.exp[1], 2)*100, "%)")
ylab = paste0("PC2 (", signif(expr.decomp$var.exp[2], 2)*100, "%)")

if(is.interactive){quartz(width = 12, height = 12)}
par(mfrow = c(2,2))
plot(expr.decomp$u, xlab = xlab, ylab = ylab, main = "Colored by Diet", 
    cex = 1.5, col = "darkgray")
text(expr.decomp$u[,1], expr.decomp$u[,2], 
    labels = as.numeric(as.factor(expr.matched.info[,"User.Defined.Strain"])),
    col = as.numeric(as.factor(expr.matched.info[,"Diet"])), cex = 0.5)

plot(expr.decomp$u, xlab = xlab, ylab = ylab, main = "Colored by Sex", 
    cex = 1.5, col = "darkgray")
text(expr.decomp$u[,1], expr.decomp$u[,2], 
    labels = as.numeric(as.factor(expr.matched.info[,"User.Defined.Strain"])),
    col = as.numeric(as.factor(expr.matched.info[,"Sex"])), cex = 0.5)

plot(expr.decomp$u, xlab = xlab, ylab = ylab, main = "Colored by Treatment", 
    cex = 1.5, col = "darkgray")
text(expr.decomp$u[,1], expr.decomp$u[,2], 
    labels = as.numeric(as.factor(expr.matched.info[,"User.Defined.Strain"])),
    col = as.numeric(as.factor(expr.matched.info[,"Treatment"])), cex = 0.5)

```

We also checked additional PCs for correlation with metformin 
treatment

```{r pc_check, fig.height = 5, fig.width = 7}
#check other PCs to see if metformin is associated with any of them
pc.cor <- apply(expr.decomp$u, 2, function(x) cor(x, as.numeric(as.factor(expr.matched.info[,"Treatment"])), method = "spearman"))
barplot(pc.cor, names = paste("PC", 1:ncol(expr.decomp$u)), ylab = "Spearman Correation", las = 2)
max.pc <- which.max(abs(pc.cor))
```

The best correlation was `r signif(max(abs(pc.cor)), 2)`. This was with PC 
`r max.pc`, which explained `r signif(expr.decomp$var.exp[max.pc], 2)*100`\% 
of the total variance.

## Effect of treatment on individual expression

For each measured transcript, we fit a model that indludes Sex, Diet, 
Treatment, and strain, as well as the interaction between strain and 
each other cofactor.

```{r model_with_strain}

model.coef.file <- file.path(results.dir, paste0("model_strain_coef_", tissue.name, ".RDS"))
model.p.file <- file.path(results.dir, paste0("model_strain_p_", tissue.name, ".RDS"))

if(!file.exists(model.coef.file) || !file.exists(model.p.file) || delete_previous){
    all.coef <- all.p <- matrix(NA, nrow = ncol(matched.expr), ncol = 11)
    rownames(all.coef) <- rownames(all.p) <- colnames(matched.expr)

    for(i in 1:ncol(matched.expr)){
        if(is.interactive){report.progress(i, ncol(matched.expr))}

        df <- data.frame(cbind(matched.expr[,i], as.factor(expr.matched.info[,"Sex"]), 
            as.factor(expr.matched.info[,"Diet"]), as.factor(expr.matched.info[,"Treatment"]), 
            as.factor(expr.matched.info[,"User.Defined.Strain"])))
        colnames(df) <- c("measured_expression", "Sex", "Diet", "Treatment", "Strain")

        model <- lm(measured_expression~(Sex+Diet+Treatment+Strain)^2, data = df)
        #summary(model)

        model.sum <- summary(model)
        all.coef[i,] <- model.sum$coefficients[,"Estimate"]
        all.p[i,] <- model.sum$coefficients[,"Pr(>|t|)"]
    }
    colnames(all.coef) <- colnames(all.p) <- rownames(model.sum$coefficients)
    saveRDS(all.coef, model.coef.file)
    saveRDS(all.p, model.p.file)
}else{
    all.coef <- readRDS(model.coef.file)
    all.p <- readRDS(model.p.file)
}
```

```{r bin_counts, eval = FALSE}
head(expr.matched.info)

#builds a data frame out of a covariate matrix derived from
#matched.covar and creates a table of specified factors to 
#make sure there is even representation across bins.
test_counts <- function(covar.mat, to.test = c("timepoint", "treatment")){
    test.df <- data.frame("sex" = factor(covar.mat[,"Sex"], levels = c("Female", "Male")),
        "diet" = factor(covar.mat[,"Diet"], levels = c("LFD", "HFD")),
        "treatment" = factor(covar.mat[,"Treatment"], levels = c("None", "Metformin")),
        "strain" = factor(covar.mat[,"User.Defined.Strain"]),
        "timepoint" = factor(covar.mat[,"Timepoint"]))

    table(test.df[,to.test])
}
```

The following box plot shows the distribution of the coefficients
for each factor. This plot shows that sex and diet have the largest
effects on transcription, as seen above in the expression matrix 
decompositions. Interestingly, strain has an even smaller effect
than metformin treatment, which already has a pretty minimal effect.

```{r coef_dist}
par(mar = c(8, 4, 2, 2))
boxplot(abs(all.coef[,2:ncol(all.coef)]), las = 2, ylab = "|Coefficient|")
```

The following plots show the p value distributions for each set of 
factors. Overall, treatment really doesn't have any effects on 
individual transcripts.

```{r qqunif, fig.width = 12, fig.height = 9}
layout.mat <- get.layout.mat((ncol(all.p)-1))
layout(layout.mat)
for(i in 2:ncol(all.p)){
    qqunif.plot(all.p[,i], plot.label = colnames(all.p)[i])
}
```

## Effects summary

The following plots show the beta coefficients plotted against
the -log(fdr) for all independent variables in the model. Each
dot is a transcript

```{r effect_summary, fig.width = 12, fig.height = 9}
fdr.level = 0.1
fdr <- apply(all.p, 2, function(x) p.adjust(x, "fdr"))
layout(layout.mat)
for(i in 2:ncol(all.coef)){
    plot(all.coef[,i], -log(fdr[,i]), main = colnames(all.coef)[i], 
    xlab = "Beta coefficient", ylab = "-log(p value)")
    abline(h = -log(fdr.level))
    #abline(h = -log(0.05/nrow(all.coef)))
}
```

## Number of Significant Effects

The following plot shows the number of significant effects per
category.

```{r num_sig}
sig.transcripts <- apply(fdr, 2, function(x) rownames(all.coef)[which(x <= fdr.level)])
sig.up <- lapply(1:ncol(fdr), function(x) rownames(all.coef)[intersect(which(fdr[,x] <= fdr.level), which(all.coef[,x] > 0))])
sig.down <- lapply(1:ncol(fdr), function(x) rownames(all.coef)[intersect(which(fdr[,x] <= fdr.level), which(all.coef[,x] < 0))])
names(sig.up) <- names(sig.down) <- colnames(fdr)

num.sig <- sapply(sig.transcripts, length)
num.up <- sapply(sig.up, length)
num.down <- sapply(sig.down, length)
par(mar = c(4, 14, 2, 2))
barplot_with_num(num.sig[2:length(num.sig)], las = 2, horiz = TRUE, text.gap = 0.15, text.shift = 0.07)
```

```{r example}
plot_int <- function(x.factor, int.factor, response, xlab = "", ylab = "", 
    trace.label = "", main = ""){

    df <- data.frame(cbind(response, x.factor, int.factor))
    colnames(df) <- c("response", "x", "y")
    model <- lm(response~x*y, data = df)
    #summary(model)
	p <- signif(summary(model)$coefficients[4,4], 2)

    group.x <- levels(x.factor)
    group.int <- levels(int.factor)
    
    comp.ids <- rbind(cbind(rep(group.x, each = length(group.int)), group.int))
    group.names <- apply(comp.ids, 1, function(x) paste(x, collapse = "_"))
    comp.idx <- apply(comp.ids, 1, function(x) intersect(which(x.factor == x[1]), which(int.factor == x[2])))

    group.vals <- lapply(comp.idx, function(x) response[x])
    group.means <- sapply(group.vals, mean)
    group.se  <- sapply(group.vals, function(x) sd(x)/sqrt(length(x)))
    names(group.vals) <- names(group.means) <- names(group.se) <- group.names

    ylim <- c(min(group.means-group.se), max(group.means+group.se))

    error.x <- rep(1:length(group.x), each = length(group.int))

    interaction.plot(x.factor, int.factor, response, 
        main = paste0(main, " (p = ", signif(p, 2), ")"), ylim = ylim,
        xlab = xlab, ylab = ylab, trace.label = trace.label)
    segments(error.x, group.means-group.se, error.x, group.means+group.se)
    
}


plot_example <- function(transcript.id){
    
    gene.name <- gene.info[which(gene.info[,1] == transcript.id),"external_gene_name"]
    sex.diet.covar <- dummy_covar(expr.matched.info[,c("Sex", "Diet")])

    par(mfrow = c(1,3), mar = c(4,4,6,4))

    model <- lm(cc.measured[,transcript.id]~as.factor(expr.matched.info[,"Sex"]))
    f <- summary(model)$fstatistic
	p <- signif(pf(f[1],f[2],f[3],lower.tail=F), 2)
    plot(cc.measured[,transcript.id]~as.factor(expr.matched.info[,"Sex"]),
    xlab = "Sex", ylab = "Measured Expression", main = paste("Effect of Sex\np =", p))

    model <- lm(cc.measured[,transcript.id]~as.factor(expr.matched.info[,"Diet"]))
    f <- summary(model)$fstatistic
	p <- signif(pf(f[1],f[2],f[3],lower.tail=F), 2)
    plot(cc.measured[,transcript.id]~as.factor(expr.matched.info[,"Diet"]),
    xlab = "Diet", ylab = "Measured Expression", main = paste("Effect of Diet\np =", p))

    model <- lm(cc.measured[,transcript.id]~as.factor(expr.matched.info[,"Treatment"]))
    f <- summary(model)$fstatistic
	p <- signif(pf(f[1],f[2],f[3],lower.tail=F), 2)
    plot(cc.measured[,transcript.id]~as.factor(expr.matched.info[,"Treatment"]),
    xlab = "Treatment", ylab = "Measured Expression", main = paste("Effect of Treatment\np =", p))

    int.group <- unique(expr.matched.info[,"Treatment"])
    treat1.locale <- which(expr.matched.info[,"Treatment"] == int.group[1])
    treat2.locale <- which(expr.matched.info[,"Treatment"] == int.group[2])
    ylim <- c(min(cc.measured[,transcript.id]), max(cc.measured[treat1.locale,transcript.id]))

    mar = c(4,4,2,4)

    #adjust the measured expression for sex and diet
    adj.measured <- adjust(cc.measured[,transcript.id,drop=FALSE], sex.diet.covar)

    mtext(gene.name, side = 3, outer = TRUE, line = -1.5, font = 2)

}
```

## Responders and non-responders overall {.tabset .tabset-fade .tabset-pills}

Work by Jeff Harder and Kendrick Key has identified two strains
that are representative of responders and non-responders:
CC43x42 (non-responder) and CC11x5 (non-responder).

We want to look at the adipose expression in these strains
with and without metformin to see if their response,
or lack thereof can be detected in the transcriptome.

The plots below show the decomposition of the gene expression
matrix for each strain individually. We adjusted the expression
for sex and timepoint before decomposing. The dot plot shows the
individual mice in PC space. The plotting symbol shows diet, and
the color shows treatment.

The box plots show the overall effects of diet and metformin 
treatment on the first two PCs of expression. This can tell 
us the effect of metformin on transcription overall.

The second row of plots shows interaction effects between 
diet and metformin on overall gene expression. The line
plots show the interaction. The bar plots show the effect
size of each factor. The number above each bar shows the
p value of that coefficient.

Metformin does actually have a significant effect on PC1
in the responder, but not in the non-responder when diet
is taken into account. This suggests that metformin has
relatively broad effects on adipose transcription overall 
in the responder.


```{r, responder_comparison, fig.width = 12, fig.height = 7, results = "asis"}

strain.comp  <- c("non-responder" = "CC043 X CC042", "responder" = "CC011 X CC005")

plot_model <- function(model.result, text.shift = 0.05, plot.label = ""){
    effects <- coef(model.result)[2:length(coef(model.result))]
    num.effects <- length(effects)
    par(mar = c(8,4,4,2))
    a <- barplot(effects, names = rownames(anova(model.result))[1:num.effects],
        ylim  = c(min(effects), max(abs(effects), na.rm = TRUE)*1.1),
        ylab = "Effect Size", main = plot.label, las = 2)
    plot.dim <- par("usr")
    plot.height <- plot.dim[4] - plot.dim[3]
    pvals <- signif(anova(model.result)$"Pr(>F)", 2)[1:num.effects]
    text.y <- effects
    text.y[which(text.y < 0)] <- 0
    text(a[,1], y = text.y+(plot.height*text.shift), labels = pvals)
    abline(h = 0)
}

#pdf("~/Desktop/strain_metformin.pdf", width = 12, height = 7)
for(i in 1:length(strain.comp)){
    cat("###", paste(names(strain.comp)[i], strain.comp[i]), "\n")
    
    strain.idx <- which(expr.matched.info[,"User.Defined.Strain"] == strain.comp[i])
    ind.id <- expr.matched.info[strain.idx,]

    #test_counts(ind.id, c("diet", "treatment"))

    #regress out the effect of sex
    ind.expr <- matched.expr[strain.idx,]
    ind.covar <- dummy_covar(ind.id[,c("Sex"), drop=FALSE])
    covar.adj <- adjust(ind.expr, ind.covar)

    #look at effects in the two diets separately
    #u_diet <- list("LFD", "HFD", c("LFD", "HFD"))
    u_diet <- list(c("LFD", "HFD"))
    diet.idx <- lapply(u_diet, function(x) which(ind.id[,"Diet"] %in% x))
    names(diet.idx) <- sapply(u_diet, function(x) paste(x, collapse = "_"))

    for(d in 1:length(diet.idx)){
        cat("####", names(u_diet)[d], "\n")
        strain.decomp <- plot.decomp(covar.adj[diet.idx[[d]],], plot.results = FALSE)

        diet.pch <- as.numeric(as.factor(ind.id[diet.idx[[d]],"Diet"]))+15
        treatment.col <- as.numeric(as.factor(ind.id[diet.idx[[d]],"Treatment"]))
    
        #quartz(width = 10, height = 7)

        if(length(u_diet[[d]]) > 1){
            layout.mat <- matrix(c(1:8), nrow = 2, byrow = TRUE)
        }else{
            layout.mat <- matrix(c(1:4), nrow = 2, byrow = TRUE)
        }
        layout(layout.mat)
    
        plot(strain.decomp$u, col = treatment.col, pch = diet.pch,
            xlab = paste0("PC1 (", round(strain.decomp$var.exp[1]*100), "%)"), 
            ylab = paste0("PC2 (", round(strain.decomp$var.exp[2]*100), "%)"), 
            main = strain.comp[i])

        data.df <- data.frame("PC1" = strain.decomp$u[,1], "PC2" = strain.decomp$u[,2],
            "diet" = factor(ind.id[diet.idx[[d]],"Diet"], levels = c("LFD", "HFD")),
            "treatment" = factor(ind.id[diet.idx[[d]],"Treatment"], levels = c("None", "Metformin")))
    
        plot.new()
        plot.window(xlim = c(0,1), ylim = c(0,1))
        legend(x = 0, y = 0.9, legend = levels(data.df$treatment), col = c(1,2), pch = 16)
        if(length(u_diet[[d]]) > 1){
            legend(x = 0, y = 0.5, legend = levels(data.df$diet), col = "black", pch = c(16, 17))
        }

        model <- lm(PC1~treatment, data = data.df)
        model.p <- signif(anova(model)$"Pr(>F)"[1], 2)
        boxplot(PC1~treatment, data = data.df, main = paste("PC1", "\np =", model.p))
        #summary(model)
        #plot_model(model, plot.label = "PC1")
    
        model <- lm(PC2~treatment, data = data.df)
        model.p <- signif(anova(model)$"Pr(>F)"[1], 2)
        boxplot(PC2~treatment, data = data.df, main = paste("PC2", "\np =", model.p))
        #summary(model)
        #plot_model(model, plot.label = "PC2")

        if(length(u_diet[[d]]) > 1){
            model <- lm(PC1~treatment*diet, data = data.df)
            #boxplot(PC1~treatment*diet, data = data.df, main = paste("PC1", "\np =", model.p), las = 2, xlab = "")
            plot_int(x.factor = data.df[,"treatment"], int.factor = data.df[,"diet"], response = data.df[,"PC1"], 
                xlab = "treatment", ylab = "PC1", trace.label = "diet", main = "PC1")
            plot_model(model, plot.label = "PC1 Interaction")
        
            model <- lm(PC2~treatment*diet, data = data.df)
            #boxplot(PC2~treatment*diet, data = data.df, main = paste("PC2", "\np =", model.p))
            plot_int(x.factor = data.df[,"treatment"], int.factor = data.df[,"diet"], response = data.df[,"PC2"], 
                xlab = "treatment", ylab = "PC2", trace.label = "diet", main = "PC2")
            plot_model(model, plot.label = "PC2 Interaction")
        
        }

        mtext(paste0(names(strain.comp)[i], ", ", names(diet.idx)[d]), outer = TRUE, side = 3, line = -1.5)
        cat("\n\n")
    }
    cat("\n\n")
}
#dev.off()
```

## Responders and non-responders by gene {.tabset .tabset-fade .tabset-pills}

We next looked at the effect of metformin and diet on individual 
genes in these strains. The plots below show the statistics for 
tests of the effects of diet and treatment on individual genes.
The volcano plots show distributions of effect sizes and p values
for the main effects of diet and treatment as well as the interaction
between them. 

The three plots underneath show the qq plots for p values for each
of the factors in the model.

```{r strain_ind_genes, results = "asis"}

for(i in 1:length(strain.comp)){
    cat("###", paste(names(strain.comp)[i], strain.comp[i]), "\n")

    strain.idx <- which(expr.matched.info[,"User.Defined.Strain"] == strain.comp[i])
    ind.id <- expr.matched.info[strain.idx,]
    ind.expr <- matched.expr[strain.idx,]

    #adjust out the effect of sex        
    ind.covar <- dummy_covar(ind.id[,"Sex",drop=FALSE])
    covar.adj <- adjust(ind.expr, ind.covar)

    strain.result.file <- file.path(results.dir, paste0("Ind_Gene_Test_", gsub(" ", "", strain.comp[i]), ".RDS"))

    if(!file.exists(strain.result.file)){
        effect.table <- p.table <- matrix(NA, ncol = 3, nrow = ncol(covar.adj))

        for(tr in 1:ncol(covar.adj)){
            report.progress(tr, ncol(covar.adj))
            #boxplot(covar.adj[,tr]~as.factor(ind.id[,"Diet"])*as.factor(ind.id[,"Treatment"]))
            df <- data.frame("expr" = covar.adj[,tr],
                "diet" = factor(ind.id[,"Diet"], levels = c("LFD", "HFD")),
                "treatment" = factor(ind.id[,"Treatment"], levels = c("None", "Metformin")))
    
            model <- lm(expr~diet*treatment, data = df)
            effect.table[tr,] <- coef(model)[2:4]
            p.table[tr,] <- anova(model)$"Pr(>F)"[1:3]
        }
        colnames(effect.table) <- colnames(p.table) <- rownames(anova(model))[1:3]
        strain.result <- list("effects" = effect.table, "p" = p.table)
        saveRDS(strain.result, strain.result.file)
    }else{
        strain.result <- readRDS(strain.result.file)
        effect.table <- strain.result$effects
        p.table <- strain.result$p
    }

    layout.matrix <- matrix(c(0,1,0,2,3,4,5,6,7), nrow = 3, byrow = TRUE)
    layout(layout.matrix, heights = c(0.3, 1, 1))
    #layout.show(7)

    par(mar = c(0,0,0,0))    
    plot.text(paste(names(strain.comp)[i], ", ", strain.comp[i]))

    par(mar = c(4,4,4,2))
    for(p in 1:ncol(p.table)){
        plot(effect.table[,p], -log10(p.table[,p]), xlab = "Effect Size", 
            main = colnames(p.table)[p], ylab = "-log10(p)")
    }

    for(p in 1:ncol(p.table)){
        qqunif.plot(p.table[,p], plot.label = colnames(p.table)[p])
    }
    
    cat("\n\n")
}
```

## Significant genes for responders and non-responders {.tabset .tabset-fade .tabset-pills}

The bar plots below show how many transcripts were significantly affected
by each factor. There are too many genes in diet to worry about, but the
responder has a few genes affected by metformin and the interaction between
metformin and diet that we can look at.

```{r deg_count, results = "asis", fig.width = 5, fig.height = 5}
fdr.thresh = 0.1
sig.genes <- vector(mode = "list", length = length(strain.comp))
names(sig.genes) <- names(strain.comp)

for(i in 1:length(strain.comp)){
    cat("###", paste(names(strain.comp)[i], strain.comp[i]), "\n")

    strain.result.file <- file.path(results.dir, paste0("Ind_Gene_Test_", gsub(" ", "", strain.comp[i]), ".RDS"))
    strain.result <- readRDS(strain.result.file)
    effect.table <- strain.result$effects
    p.table <- strain.result$p

    p.adj <- apply(p.table, 2, function(x) p.adjust(x, "fdr"))
    sig.p <- lapply(1:ncol(p.adj), function(x) which(p.adj[,x] <= fdr.thresh))
    names(sig.p) <- colnames(p.adj)
    
    gene.id <- lapply(sig.p, function(x) colnames(adj.cc.expr)[x])
    gene.names <- lapply(gene.id, function(x) gene.info[match(x, gene.info[,"ensembl_gene_id"]),c("ensembl_gene_id", "external_gene_name", "chromosome_name", "start_position")])
    gene.effect <- lapply(sig.p, function(x) effect.table[x,])
    gene.p <- lapply(sig.p, function(x) p.table[x,])

    #put all the results in one place for each factor
    factor.tables <- vector(mode = "list", length = length(sig.p))
    names(factor.tables) <- names(sig.p)
    for(f in 1:length(factor.tables)){
        factor.tables[[f]] <- cbind(gene.names[[f]], signif(gene.effect[[f]][,f], 2), signif(gene.p[[f]][,f]))
        colnames(factor.tables[[f]])[5:6] <- c("effect", "p_value")
    }
    saveRDS(factor.tables, file.path(results.dir, paste0("Factor_Genes_", gsub(" ", "", strain.comp[i]), ".RDS")))

    num.sig <- sapply(sig.p, length)
    barplot_with_num(num.sig, main = paste(names(strain.comp)[i], "\n", strain.comp[i]))
    cat("\n\n")
}
```

## Gene Tables

The table below show the genes that were affected by metformin
in the responder.

```{r responder_genes}
i = 2
responder.table <- readRDS(file.path(results.dir, paste0("Factor_Genes_", gsub(" ", "", strain.comp[i]), ".RDS")))

f = 2
has.vals <- which(apply(responder.table[[f]][,1:4], 1, function(x) !all(is.na(x))))
datatable(responder.table[[f]][has.vals,])
```

The table below shows the genes that were affected by an interaction
between diet and treatment.

```{r diet_by_treatment}
f = 3
has.vals <- which(apply(responder.table[[f]][,1:4], 1, function(x) !all(is.na(x))))
datatable(responder.table[[f]][has.vals,])
```

## Enrichments for responders  {.tabset .tabset-fade .tabset-pills}

There were no genes significantly affected by metformin in the non-responder.
But the responder had a few. Unfortunately, the enrichments aren't particularly 
enlightening.

Diet affects mitochondrial activity, and metformin affects genes involved in
rRNA binding. Hmm...

```{r deg_enrich, results = "asis"}
for(i in 1:length(strain.comp)){
    cat("###", paste(names(strain.comp)[i], strain.comp[i]), "\n")

    strain.result.file <- file.path(results.dir, paste0("Ind_Gene_Test_", gsub(" ", "", strain.comp[i]), ".RDS"))
    strain.result <- readRDS(strain.result.file)
    effect.table <- strain.result$effects
    p.table <- strain.result$p

    p.adj <- apply(p.table, 2, function(x) p.adjust(x, "fdr"))
    sig.p <- lapply(1:ncol(p.adj), function(x) which(p.adj[,x] <= fdr.thresh))
    names(sig.p) <- colnames(p.adj)

    sig.effects <- lapply(sig.p, function(x) effect.table[x,])
    if(length(sig.effects[[2]]) > 0){
        plot.with.model(sig.effects[[2]][,"diet"], sig.effects[[2]][,"treatment"], report = "cor.test",
            xlab = "Effect of Diet", ylab = "Effect of Treatment", main = "Effect of Diet vs. Treatment")
    }else{
        plot.text("No genes were affected by treatment")
    }

    enrich <- lapply(sig.p, function(x) gost(colnames(adj.cc.expr)[x], organism = "mmusculus", 
        source = c("GO", "KEGG", "REACTOME", "CORUM", "HP")))
    plot.enrichment.group(enrich, transformation = function(x) sqrt(sqrt(x)))

    cat("\n\n")
}
```