---
title: "Correlation Networks"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction

We are curious about constraints on gene correlation networks.
We think that, in addition to cis effects, transcript levels
are constrained to a small space of correlation structures. 

Cis effects measured in any given population incorporate the
constraints of the correlation matrix for that population, but
this might not translate to another population. The cis effects
and interaction effects calculated in another population with 
different allele frequencies and combinations would be different
because of the constraints on the gene expression correlation
network.

Something along these lines could explain why we can predict 
how measured gene expression in the CC-RIX is correlated with
traits, but we can't predict the correlation between gene 
expression and traits (i.e. there is no correlation between
imputed gene expression and phenotype in the CC-RIX).

We are missing something when it comes to the gene expression
imputation. We have measured marginal h's in the DO population 
that are influenced by the j's in this population. We then assume
that these h's are the "true" h's and try to port them over to 
the CC-RIX. However, if we were to measure h's directly in the 
CC-RIX, we would get different results even if the true h's are
the same in both populations because the measured h's depend on
the j's, which could be different in the two populations. 

How do we get at both h's and j's in these populations? 
Gene correlation networks? How do the gene correlation networks
in the two populations compare? How do these networks compare
to the correlation network for the imputed transcripts?

```{r load_code}
is.interactive = FALSE
#is.interactive = TRUE
library("here")

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("propagate", "pheatmap", "igraph")
load_libraries(all.packages, personal.library = TRUE)
```

Read in gene expression for DO, CC-RIX, and imputed CC-RIX.

```{r read_data}
tissue.names <- c("adipose", "liver", "muscle")
do.expr <- readRDS(here("Data", "DO", "Tissue_Expression_Adjusted.RDS")) #generated by 1a.Tissue_Expression.Rmd
cc.expr <- readRDS(here("Data", "CC-RIX", "Expression.by.Tissue.Adjusted.RDS")) #generated by 1c.CC-RIX_Data_Parsing.Rmd
cc.imp <- lapply(tissue.names, 
  function(x) readRDS(get.files(here("Data", "imputed"), 
  want = c("expanded", x), full.names = TRUE))) #generated by Matt Mahoney
names(cc.imp) <- tissue.names
do.genetic.imp <- readRDS(here("Data", "imputed", "Adjusted_Expression_DO_imputed_genetic.RDS")) #generated by Matt Mahoney
do.local.imp <- readRDS(here("Data", "imputed", "Adjusted_Expression_DO_imputed_local.RDS")) #generated by Matt Mahoney
```

Build a transcript-by-transcript correlation network for 
each tissue in each population. Within each tissue
compare the networks across populations. The comparison
is the correlation of the edge weights in the gene
correlation network. 


```{r match_transcripts, warning = FALSE, error = FALSE, message = FALSE}

pops <- c("DO_Measured", "DO_Imputed_Genetic", "DO_Imputed_Local", 
  "CC-RIX_Measured", "CC-RIX_Imputed")
pop.pairs <- pair.matrix(1:length(pops))
pop.name.pairs <- pair.matrix(pops)
net.cor.files <- here("Results", "Transcription_Networks", 
  paste0("Network_Cor_", tissue.names,".RDS"))

#generate correlation matrices for each transcriptome
#there is one transcriptome per tissue per population
#currently, there are three tissues and 5 populations
#and each network is about 1 Gig, which means these
#correlation networks will take up 15 gigs of space!

#we also calculate the degree of each as we calculate
#the correlation networks, so we don't need to regenerate
#the full networks later
gene.deg <- vector(mode = "list", length = length(tissue.names))
names(gene.deg) <- tissue.names

for(tx in 1:length(tissue.names)){
    
    do.m.idx <- grep(tissue.names[tx], names(do.expr), ignore.case = TRUE)
    cc.m.idx <- grep(tissue.names[tx], names(cc.expr), ignore.case = TRUE)
    cc.imp.idx <- grep(tissue.names[tx], names(cc.imp), ignore.case = TRUE)
    do.local.imp.idx <- grep(tissue.names[tx], names(do.local.imp), ignore.case = TRUE)
    do.genetic.imp.idx <- grep(tissue.names[tx], names(do.genetic.imp), ignore.case = TRUE)

    #tx.expr needs to be built in the same order as pops!
    #make sure the index names match too!
    tx.expr <- list(do.expr[[do.m.idx]], 
                    do.genetic.imp[[do.genetic.imp.idx]],
                    do.local.imp[[do.local.imp.idx]],
                    cc.expr[[cc.m.idx]], 
                    cc.imp[[cc.imp.idx]])
    names(tx.expr) <- pops

    common.tx <- Reduce("intersect", lapply(tx.expr, colnames))[1:2000]
    num.tx <- length(common.tx)
    matched.expr <- lapply(tx.expr, function(x) apply(x[,common.tx], 2, rankZ))

    pop.deg <- vector(mode = "list", length = length(pops))
    names(pop.deg) <- pops

    for(pop in 1:length(matched.expr)){
      cor.filename <- here("Results", "Transcription_Networks", 
        paste0("Transcript_Cor_", tissue.names[tx], "_", 
        names(matched.expr)[pop], ".RDS"))
      if(!file.exists(cor.filename)){
        pop.cor.result <- bigcor(matched.expr[[pop]])
        #pull out the matrix
        pop.cor <- pop.cor.result[1:num.tx, 1:num.tx]
        rownames(pop.cor) <- colnames(pop.cor) <- colnames(matched.expr[[pop]])
        pop.deg[[pop]] <- colMeans(abs(pop.cor))
        #pull out the upper triangle to reduce the size
        pop.corV <- as.vector(pop.cor[upper.tri(pop.cor, diag = FALSE)])
        #save a concrete version of the reduced set of correlations
        saveRDS(pop.corV, cor.filename)
        #delete the ff version
        result.attr <- attributes(pop.cor.result)
        ff.file <- attributes(result.attr$physical)$filename
        file.remove(ff.file)
      }
    }
  gene.deg[[tx]] <- pop.deg
}

saveRDS(gene.deg, here("Results", "Transcription_Networks", "Gene_Degree.RDS"))

```

```{r pairwise_nets}
#calculate pairwise correlations for all transcription networks
#the result is one matrix per tissue containing all pairwise 
#comparisons of populations
all.net.cor <- vector(mode = "list", length = length(tissue.names))
names(all.net.cor) <- tissue.names
for(tx in 1:length(tissue.names)){
    
  if(file.exists(net.cor.files[tx])){next()}
    pop.cor <- rep(NA, nrow(pop.pairs))
    for(pp in 1:nrow(pop.pairs)){
      file1 <- here("Results", "Transcription_Networks", 
        paste0("Transcript_Cor_", tissue.names[tx], "_", 
        names(matched.expr)[pop.pairs[pp,1]], ".RDS"))
      tcor1 <- readRDS(file1)

      file2 <- here("Results", "Transcription_Networks", 
        paste0("Transcript_Cor_", tissue.names[tx], "_", 
        names(matched.expr)[pop.pairs[pp,2]], ".RDS"))
      tcor2 <- readRDS(file2)

      #cat(pops[pop.pairs[pp,1]], "vs.", pops[pop.pairs[pp,2]], "\n")
      pop.cor[pp] <- cor(tcor1, tcor2, method = "spearman")
      #plot(as.vector(all.cor[[pop.pairs[pp,1]]][1:100, 1:num.tx]), as.vector(all.cor[[pop.pairs[pp,2]]][1:100, 1:num.tx]))
    }
    net.cor <- cbind(pop.name.pairs, pop.cor)
    colnames(net.cor) <- c("Pop1", "Pop2", "correlation")

    all.net.cor[[tx]] <- net.cor
    saveRDS(net.cor, net.cor.files[tx])
}

```

```{r pairwise_degree}
#calculate pairwise correlations for all degree
deg.cor.files <- gsub("Network_Cor", "Degree_Cor", net.cor.files)
all.deg.cor <- vector(mode = "list", length = length(tissue.names))
names(all.deg.cor) <- tissue.names
for(tx in 1:length(tissue.names)){
    pop.deg <- gene.deg[[tx]]
    deg.cor <- rep(NA, nrow(pop.pairs))
    for(pp in 1:nrow(pop.pairs)){
      pop1 <- pop.pairs[pp,1]
      pop2 <- pop.pairs[pp,2]
      #cat(pops[pop.pairs[pp,1]], "vs.", pops[pop.pairs[pp,2]], "\n")
      deg.cor[pp] <- cor(pop.deg[[pop1]], pop.deg[[pop2]], method = "spearman")
      #plot(as.vector(all.cor[[pop.pairs[pp,1]]][1:100, 1:num.tx]), as.vector(all.cor[[pop.pairs[pp,2]]][1:100, 1:num.tx]))
    }
    deg.cor.mat <- cbind(pop.name.pairs, deg.cor)
    colnames(net.cor) <- c("Pop1", "Pop2", "correlation")

    saveRDS(deg.cor.mat, deg.cor.files[tx])
    all.deg.cor[[tx]] <- deg.cor.mat
}

```

We see below that the measured gene expression correlation 
networks are similar (but by no means identical) between the 
DO and CC-RIX. For any given tissue, these transcripts have 
a similar correlation structure. Thus, the transcription 
network itself is conserved across the populations.

Next, look at the DO_measured vs DO_imputed_local. There
is extremely little correlation. This means that the local
genetic effects alone are insufficient to describe the 
transcription correlation network. This seems obvious once
stated, but it means that transcripts imputed from local
genotype alone, are missing a huge amount of information
about the state of the transcriptional network. 

Contrast this low correlation with the correlation 
between the measured network and the network imputed from
the full genetic model. This later correlation is very high. 
The full genetic model does capture information about 
transcript correlations. However, here we cannot distinguish 
between direct distal genetic effects, and other indirect 
effects. For example, the full genome may make a fatter mouse 
that affects expression of genes that respond to being fat, 
independent of a direct effect of the genome on the expression 
of those genes. We can't measure the reason for the correlations. 

Similar to the case within the DO, the transcription network 
of the imputed CC-RIX transcriptome is very weakly correlated
with the CC-RIX measured transcriptome and not at all correlated
with the DO measured transcriptome.

I'm not really sure if people care about this. On the one hand
it seems trivial to say that the locally imputed transcripts
do not recapitulate the full complexity of the transcription 
network. However, that correlation structure is critically 
important to the state of the animal's transcriptome and our
interpretation of what is going on in the cells.

```{r compare_nets, fig.width = 8, fig.height = 7}
tissue.col <- categorical_pal(8)
all.net.cor <- lapply(net.cor.files, readRDS)
names(all.net.cor) <- tissue.names
comp.names <- apply(pop.name.pairs, 1, function(x) paste(x, collapse = " vs. "))

comp.mat <- sapply(all.net.cor, function(x) as.numeric(x[,3]))
comp.order <- order(rowMeans(comp.mat), decreasing = TRUE)
par(mar = c(8, 4, 4, 4))
a <- barplot(t(comp.mat[comp.order,]), beside = TRUE, 
  col = tissue.col[1:length(tissue.names)],
  ylab = "Correlation")
legend("topright", fill = tissue.col[1:length(tissue.names)], legend = tissue.names)
par(xpd = NA)
text(x = colMeans(a), y = -0.02, labels = gsub(".vs", ".vs\n", comp.names[comp.order]), 
  srt = 45, adj = 1)
par(xpd = FALSE)

#comp.test <- matrix(NA, nrow = length(pops), ncol = length(pops))
#rownames(comp.test) <- colnames(comp.test) <- pops
#comp.test[upper.tri(comp.test)] <- comp.mat[,1]
#pheatmap(comp.test, cluster_rows = FALSE, cluster_cols = FALSE, display_numbers = TRUE)
```

```{r eval = FALSE}
all.deg.cor.mat <- sapply(all.deg.cor, function(x) as.numeric(x[,3]))
par(mar = c(8, 4, 4, 4))
barplot(t(all.deg.cor.mat[comp.order,]), beside = TRUE, 
  col = tissue.col[1:length(tissue.names)],ylab = "Degree Correlation")
legend("topright", fill = tissue.col[1:length(tissue.names)], legend = tissue.names)
par(xpd = NA)
text(x = colMeans(a), y = -0.02, labels = gsub(".vs", ".vs\n", comp.names[comp.order]), 
  srt = 45, adj = 1)
par(xpd = FALSE)

```


## Network implications

We have now constructed 

* What is the relationship between degree in the correlation 
  network and local eQTL LOD score?
We might expect that genes with a high degree in the network 
would be less likely to have large local eQTL, because a large 
eQTL would be disruptive
to a large part of the network.

* What is the relationship between degree in the correlation
  network and mediated genetic effect?
We might expect that genes with large mediated genetic 
effects would be more central (higher degree) than genes
with low mediated genetic effects.

The plots below show the correlation between degree
in the correlation network and local LOD score. There
is a strong negative corrlation. That is, eQTL with 
strong local effects mostly influence peripheral genes.


```{r cor_vs_allele, fig.width = 9, fig.height = 3, message = FALSE, error = FALSE, warning = FALSE}
tissue.data.file <- lapply(tissue.names, 
  function(x) get.files(here("Data", "DO"), want = x, dont.want = "Gene", full.names = TRUE))
tissue.lod <- lapply(tissue.data.file, function(x) readRDS(x)$lod.peaks$additive)
names(tissue.lod) <- tissue.names

gene.deg <- vector(mode = "list", length = length(tissue.names))
names(gene.deg) <- tissue.names

par(mfrow = c(1,3))
pop <- 1 #look only at correlation network for DO measured transcription
for(tx in 1:length(tissue.names)){
  cor.filename <- here("Results", "Transcription_Networks", 
        paste0("Transcript_Cor_", tissue.names[tx], "_", 
        names(matched.expr)[pop], ".RDS"))
  tissue.cor <- readRDS(cor.filename)
  gene.deg[[tx]] <- apply(tissue.cor[,1:num.tx], 2, function(x) mean(abs(x)))
  matched.lod <- tissue.lod[[tx]]$lod[match(names(gene.deg[[tx]]), 
    tissue.lod[[tx]]$gene.id)]
  plot.with.model(matched.lod, gene.deg[[tx]], report = "cor.test", 
    xlab = "Local LOD", ylab = "Degree", main = tissue.names[tx])
}
saveRDS(gene.deg, here("Results", "Transcription_Networks", "Gene_Degree_DO_Measured.RDS"))
```

The following figures show the relationship between
mediated genetic effect calculated in 3.High_Dimensional_Mediation.Rmd
and degree in the gene correlation network. There also seems 
to be a negative correlation here, although, it is much
weaker. This suggests that genes with high MGE are also 
peripheral to the network. LOD score is one component of MGE, 
so that could be driving the negative effect. Maybe we should 
look at loading, rather than MGE here. 

Also, there is no correlation in adipose tissue between MGE
and gene degree. What does that mean?

```{r mge_v_degree, fig.width = 9, fig.height = 3}
mge.results.dir <- here("Results", "High_Dim_Med", "tissue_together-local_imp-_-complete_mediation")
pop.mge <- readRDS(file.path(mge.results.dir, "Population_MGE_Loadings.RDS"))

par(mfrow = c(1,3))
for(tx in 1:length(tissue.names)){
  tx.idx <- grep(tissue.names[tx], names(pop.mge), ignore.case = TRUE)
  common.tx <- intersect(names(pop.mge[[tx.idx]]), names(gene.deg[[tx]]))
  mge.idx <- match(common.tx, names(pop.mge[[tx.idx]]))
  deg.idx <- match(common.tx, names(gene.deg[[tx]]))
  plot.with.model(abs(pop.mge[[tx.idx]][mge.idx]), gene.deg[[tx]][deg.idx],
    ylab = "Degree", xlab = "|MGE|", main = tissue.names[tx], report = "cor.test")
}
```