---
title: "Gene Set Validation"
author: Anna L Tyler
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: hide
    collapsed: no
    toc: yes
    toc_float: yes
---

## Introduction

The purpose of this workflow is to take in an arbitrary
set of genes that might be associated with a GO term, or
a particular process and to look at expression of these
genes in their relation to mouse phenotype.

The 3b.High_Dimensional_Mediation.Rmd workflow writes out
all the gene sets from the clustered TOA analysis into 
Results/Gene_Sets. Gene sets can also be made by hand
and placed in this folder.

The clusters generated in 3b.High_Dimensional_Mediation.Rmd
are based on loading patterns across the four tissues. Clusters
do not necessarily indicate that the contained genes are 
co-expressed in any given tissue.

```{r load_code}
rm(list = ls())

#is.interactive = TRUE
library("here")

exp.name = "tissue_together-_-complete_mediation-germline_kinship"
prev.results.dir <- here("Results", "DO", "High_Dim_Med_no_MGE", exp.name)

#tissue.cols <- c("orange", "#8dd3c7", "tan", "brown") #colors that are not related to the DO/CC colors
tissue.cols <- as.matrix(read.delim(here("Data", "general", "tissue_colors.txt"), header = FALSE, row.names = 1))[,1]

all.fun <- list.files(here("Code"), full.names = TRUE, pattern = ".R")
for(i in 1:length(all.fun)){source(all.fun[i])}
```

```{r read_gene set}
# gene set

args <- commandArgs(trailingOnly=T)
gene.set.name <- args[1]

if(is.na(gene.set.name) || length(gene.set.name) == 0){
    gene.set.name <- "Autophagy_-_animal_cluster.txt"
    #gene.set.name <- "Apelin_signaling_pathway_cluster.txt"
    #gene.set.name <- "Lep-Lepr_quaternary_protein_cluster.txt"
    #gene.set.name <- "Riboflavin_metabolism_cluster.txt" #just one gene
    #gene.set.name <- "endomembrane_system_cluster.txt" #islet-specific cluster
}

toa.cluster.dir <- file.path(prev.results.dir, "TOA_Clusters")
cluster.name <- gsub("_", " ", gsub(".txt", "", gene.set.name), "_")
gene.set <- read.table(file.path(toa.cluster.dir, gene.set.name)) 
```

```{r load_libraries, message = FALSE, warning = FALSE, error = FALSE}
all.packages <- c("pheatmap", "gprofiler2", "RColorBrewer", "GEOquery", "limma", "vioplot", "grid")

all.paths <- .libPaths()
personal.path <- grep("atyler", all.paths)
load_libraries(all.packages, lib.loc = all.paths[personal.path])
```


```{r read_data}

model_scores <- readRDS(file.path(prev.results.dir, "Model_Scores_0.RDS"))

transcript_loadings <- readRDS(file.path(prev.results.dir, "Loadings_Transcripts_0.RDS"))
tissue.names <- names(transcript_loadings)

adj.expr <- readRDS(here("Results", "DO", "Data", "Tissue_Expression_Adjusted.RDS")) #generated by 1a.Tissue_Expression.Rmd

mus.hum <- read.delim(here("Data", "general", "human.mouse.orthologs.txt"))
gene.info <- read.delim(here("Data", "general", "mouse_gene_info.txt"))
```

This report shows results for the `r cluster.name`

## Gene Set Loadings Across Tissues

The plot below shows the relative loadings of this gene 
set across tissues.

```{r tissue_loadings}
tx.loadings <- lapply(transcript_loadings, function(x) x[intersect(rownames(x), gene.set[,1]),,drop=FALSE])
boxplot(tx.loadings, main = cluster.name, ylab = "Loading", col = tissue.cols)
abline(h = 0)
```

## Gene Set Phenotype Prediction

We will not do phenotype prediction because the information 
in the transcriptome is so redundant that any random 
selection of genes can predict phenotoype.

```{r set_predict, fig.width = 8, fig.height = 8}

set_expr <- lapply(adj.expr, function(x) x[,intersect(colnames(x), gene.set[,1]),drop=FALSE])

scaled.expr <- lapply(set_expr, function(x) apply(x, 2, scale))
for(tx in 1:length(scaled.expr)){
    dimnames(scaled.expr[[tx]]) <- dimnames(set_expr[[tx]])
}
common.ind <- Reduce("intersect", lapply(scaled.expr, rownames))

cluster_expr <- lapply(scaled.expr, function(x) x[,intersect(colnames(x), gene.set[,1]),drop=FALSE])


par(mfrow = c(2,2))
for(tx in 1:length(scaled.expr)){
    loaded_expr <- apply(cluster_expr[[tx]], 1, function(x) x*transcript_loadings[[tx]][intersect(rownames(transcript_loadings[[tx]]), gene.set[,1]),,drop=FALSE])
    #pheatmap(loaded_expr)
    if(length(dim(loaded_expr)) > 0){
        pred_expr <- colMeans(loaded_expr)
    }else{
        pred_expr <- loaded_expr #if we are only looking at one gene
    }
    plot.with.model(pred_expr[common.ind], model_scores[[tx]][common.ind,"Outcome"], 
        main = paste(tissue.names[tx], "\n", ncol(cluster_expr[[tx]]), "Genes"), 
        xlab = "Cluster Prediction", ylab = "Phenotypic Outcome")
}
```

## Gene Set Expression {.tabset .tabset-fade .tabset-pills}

The following plot shows the relative gene expression of the gene
cluster across all individuals ordered by their phenotype score. 

```{r cluster_expr, fig.width = 8, fig.height = 8, results = "asis"}
pheno  <- model_scores[[1]][common.ind,"Outcome"]
geno <- model_scores[[1]][common.ind,"Causal"]
ordered.ind <- names(pheno)[order(pheno)]

for(tx in 1:length(tissue.names)){
    cat("###", tissue.names[tx], "\n")

    layout(matrix(c(1,2), nrow = 2), heights = c(0.5, 1))
    par(mar = c(0,4,4,4))
    plot(pheno[ordered.ind], ylab = "Phenotype Score", axes = FALSE,
        xlab = "")
    
    axis(2)
    par(mar = c(4,4,0,4))
    
    if(ncol(cluster_expr[[tx]]) > 1){
        boxplot(t(cluster_expr[[tx]][ordered.ind,]), axes = FALSE, ylab = "Expression")
        expr.mean <- rowMeans(cluster_expr[[tx]][ordered.ind,])
    }else{
        plot(1:length(ordered.ind), t(cluster_expr[[tx]][ordered.ind,]), axes = FALSE, xlab = "",
            ylab = "Expression", pch = 16)
        expr.mean <- cluster_expr[[tx]][ordered.ind,]
    }

    axis(2)

    smooth.fit <- loess(expr.mean[ordered.ind]~c(1:length(ordered.ind)))
    smooth.pred <- predict(smooth.fit)
    points(1:length(expr.mean), smooth.pred, type = "l", col = "red", lwd = 3)

    mtext(paste(cluster.name, "expression", tissue.names[tx]), side = 3, outer = TRUE, line = -2.5)
    cat("\n\n")
}
```

## Gene Set PC plots

Another way to look at this is with PC plots. The decomposition
is of the scaled gene expression matrix. Points are
individuals, and they are colored by phenotype score.

```{r expr_pc, fig.width = 8, fig.height = 8}

pheno.col <- colors.from.values(pheno[ordered.ind], use.pheatmap.colors = TRUE)

par(mfrow = c(2,2))
for(tx in 1:length(tissue.names)){
    if(ncol(cluster_expr[[tx]]) > 1){
        plot.decomp(cluster_expr[[tx]][ordered.ind,], cols = pheno.col, main = tissue.names[tx])
    }
}
```

## Individual Gene Correlations with Phenotype Score {.tabset .tabset-fade .tabset-pills}

We can correlate the individual genes with both genotype
and phenotype scores

```{r individual_gene_cor}
#check to see if individual genes are smoothly associated with phenotype
pheno.cor.mat <- geno.cor.mat <- matrix(NA, nrow = length(tissue.names), ncol = length(gene.set[,1]))
colnames(pheno.cor.mat) <- colnames(geno.cor.mat) <- gene.set[,1]
rownames(pheno.cor.mat) <- rownames(geno.cor.mat) <- tissue.names

for(tx in 1:length(tissue.names)){
    pheno.cor <- apply(cluster_expr[[tx]][ordered.ind,], 2, 
        function(x) cor(x, pheno[ordered.ind], use = "pairwise.complete.obs"))
    pheno.cor.mat[tx,names(pheno.cor)] <- pheno.cor

    geno.cor <- apply(cluster_expr[[tx]][ordered.ind,], 2, 
        function(x) cor(x, geno[ordered.ind], use = "pairwise.complete.obs"))
    geno.cor.mat[tx, names(geno.cor)] <- geno.cor
}

gene.names <- gene.info[match(gene.set[,1], gene.info[,1]), "external_gene_name"]
na.idx <- which(is.na(gene.names))
if(length(na.idx) > 0){
    gene.names[na.idx] <- gene.set[na.idx,1]
}
colnames(pheno.cor.mat) <- colnames(geno.cor.mat) <- gene.names
```

### Correlation with Phenotype Score

The following heatmap shows the correlations between individual 
gene expression and the *phenotype score*.

```{r pheno_cor, fig.width = 5, fig.height = 8}
pheatmap(t(pheno.cor.mat), display_numbers = TRUE, cluster_cols = FALSE)
```

### Correlation with Genotype Score

The following heatmap shows the correlations between individual 
gene expression and the *genotype score*.

```{r geno_cor, fig.width = 5, fig.height = 8}
pheatmap(t(geno.cor.mat), display_numbers = TRUE, cluster_cols = FALSE)
```

## Genotype Score Correlation vs. Phenotype Score Correlation

The following plots show gene-by-gene the comparison
between their correlations with genotype and phenotype.

```{r comp, fig.width = 9, fig.height = 7}

all.geno.cor <- as.vector(geno.cor.mat)
all.pheno.cor <- as.vector(pheno.cor.mat)
pt.col <- rep(tissue.cols, ncol(geno.cor.mat))
pt.name <- rep(colnames(geno.cor.mat), each = nrow(geno.cor.mat))

xmin <- min(all.geno.cor, na.rm = TRUE)*1.05
xmax <- max(all.geno.cor, na.rm = TRUE)*1.05
ymin <- min(all.pheno.cor, na.rm = TRUE)*1.05
ymax <- max(all.pheno.cor, na.rm = TRUE)*1.05

layout(matrix(c(1,2), ncol = 2), widths = c(1, 0.3))
par(mar = c(4,4,4,0))
plot(all.geno.cor, all.pheno.cor, col = pt.col, pch = 16, xlab = "Correlation with Genotype Score",
    ylab = "Correlation with Phenotype Score", xlim = c(xmin, xmax), ylim = c(ymin, ymax))
text(all.geno.cor, all.pheno.cor, labels = pt.name, cex = 0.7, pos = 4, col = pt.col)
abline(h = 0, v = 0)
par(mar = c(4,0,4,0))
plot.new()
plot.window(xlim = c(0,1), ylim = c(0,1))
legend(0, 1, col = tissue.cols, pch = 16, legend = tissue.names, horiz = FALSE)

```



